{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "25a3eb7b",
   "metadata": {},
   "source": [
    "### Load libraries and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a9a52127",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘proxy’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    as.dist, dist\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    as.matrix\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.11.1\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.14\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:circlize’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:igraph’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: splines\n",
      "\n",
      "\n",
      "Attaching package: ‘pheatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    pheatmap\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Loading required package: R.oo\n",
      "\n",
      "Loading required package: R.methodsS3\n",
      "\n",
      "R.methodsS3 v1.8.1 (2020-08-26 16:20:06 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "R.oo v1.24.0 (2020-08-26 16:11:58 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    hierarchy\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "R.utils v2.11.0 (2021-09-26 08:30:02 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    draw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, inherits, isOpen, nullfile, parse,\n",
      "    warnings\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load libraries\n",
    "library(Seurat)\n",
    "library(stringr)\n",
    "library(viridis)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "library(cluster)\n",
    "library(data.table)\n",
    "library(foreach)\n",
    "library(doParallel)\n",
    "library(proxy)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(igraph)\n",
    "library(qvalue)\n",
    "library(dplyr)\n",
    "library(viridis)\n",
    "library(VGAM)\n",
    "library(forcats)\n",
    "library(grDevices)\n",
    "library(graphics)\n",
    "library(RColorBrewer)\n",
    "library(pheatmap)\n",
    "library(Cairo)\n",
    "library(reshape2)\n",
    "library(R.utils)\n",
    "library(Rcpp)\n",
    "library(parallelDist)\n",
    "set.seed(seed = 42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54558b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load basic functions\n",
    "createEmptyDf = function( nrow, ncol, colnames = c() ){\n",
    "  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )\n",
    "}\n",
    "\n",
    "tableread_fast = function(i, header=TRUE, quote=\"\", sep=\",\"){\n",
    "  tmp = fread(i, header=header, sep=sep, quote=quote, nThread=32)\n",
    "  tmp = as.data.frame(tmp)\n",
    "  return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce268f3f",
   "metadata": {},
   "source": [
    "### Download dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2778a137",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:47: unexpected symbol\n1: ##Download dataset\n2: system(\"wget --load-cookies /tmp/cookies.txt \"https\n                                                 ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:47: unexpected symbol\n1: ##Download dataset\n2: system(\"wget --load-cookies /tmp/cookies.txt \"https\n                                                 ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "#Please download data from following http and put them in your working directory\n",
    "\n",
    "#SCT_raw.data.tar.gz\n",
    "#https://drive.google.com/file/d/1xyffNUUdoH8qD2O9ua6yMTqVAwFJNrRm/view?usp=sharing\n",
    "\n",
    "#scTCR_raw.data.tar.gz\n",
    "#https://drive.google.com/file/d/1UOgi6re2WdIfX6-e7MnRrCJI1RjeNcC7/view?usp=sharing\n",
    "\n",
    "#Bulk TCRseq data with AIM status\n",
    "#https://drive.google.com/file/d/18egQKUu4Htoqg9TkuRXh6Kn_-z3xySP4/view?usp=sharing\n",
    "\n",
    "###Unzip rawdata\n",
    "gunzip(\"SCT_raw.data.tar.gz\")\n",
    "gunzip(\"scTCR_raw.data.tar.gz\")\n",
    "gunzip(\"WGCNA_AIMcombined_selected.tar.gz\")\n",
    "untar(\"SCT_raw.data.tar\")\n",
    "untar(\"scTCR_raw.data.tar\")\n",
    "untar(\"WGCNA_AIMcombined_selected.tar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7221f3f",
   "metadata": {},
   "source": [
    "### Create Seurat object and filter genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "73dd4b1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(out.dir):\n",
      "“'scTCR_raw.data' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'RNA'</li><li>'ADT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'RNA'\n",
       "\\item 'ADT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'RNA'\n",
       "2. 'ADT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"RNA\" \"ADT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###\"ID28\" in single-cell dataset corresponds to \"ID027\" in TCR repertoire dataset ###\n",
    "\n",
    "##load data of ID02\n",
    "#Put gene list in the target panel in the working directory (\"BD_genes_human\")\n",
    "\n",
    "###Input layer\n",
    "set.seed(seed = 42)\n",
    "\n",
    "sample.name <- \"ID02\"\n",
    "out.dir <- \"scTCR_raw.data\"\n",
    "\n",
    "################################Processing layer##########################################\n",
    "dir.create(out.dir)\n",
    "\n",
    "##Load targeted gene expression data\n",
    "input_name <- str_c(\"raw.data/\", sample.name, \"Target_results/matrix/\", sep = \"\") %>% \n",
    "                str_c(\"matrix_inflection_\", sample.name, \"Target.txt\", sep = \"\")\n",
    "compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "gunzip(compressed.name)\n",
    "#Create Seurat object\n",
    "matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "row.names(matrix) <- matrix$V1\n",
    "matrix <- dplyr::select(matrix, -V1)\n",
    "seu1 <- CreateSeuratObject(counts=matrix, project = \"seu\", min.cells = 3, min.features = 10)\n",
    "#Output cell barcode information for scTCR analysis\n",
    "name.output <- str_c(out.dir, sample.name, sep = \"/\") %>% str_c(\"BC_target.txt\", sep = \"_\")\n",
    "BC <- row.names(seu1@meta.data)\n",
    "write.table(BC, name.output, row.names = FALSE, quote = FALSE, col.names = FALSE)\n",
    "\n",
    "##Load protein expression data\n",
    "input_name <- str_c(\"raw.data/\", sample.name, \"Abseq_results/matrix/\", sep = \"\") %>% \n",
    "                str_c(\"Hashtag_top1M_\", sample.name, \"Abseq.txt\", sep = \"\")\n",
    "compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "#gunzip(compressed.name)\n",
    "#Extract cells in SCT seurat object\n",
    "Abseq.matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "Abseq.matrix <- dplyr::filter(Abseq.matrix, CellBC %in% BC)\n",
    "row.names(Abseq.matrix) <- Abseq.matrix$CellBC\n",
    "Abseq.matrix <- dplyr::select(Abseq.matrix, -CellBC)\n",
    "Abseq.matrix <- t(Abseq.matrix)\n",
    "all.equal(rownames(seu1@meta.data), colnames(Abseq.matrix)) #Certify\n",
    "\n",
    "##Combine ADT data into SCT seurat object\n",
    "adt_assay <- CreateAssayObject(counts = Abseq.matrix)\n",
    "seu1[[\"ADT\"]] <- adt_assay\n",
    "Assays(seu1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f07daa53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(out.dir):\n",
      "“'scTCR_raw.data' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'RNA'</li><li>'ADT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'RNA'\n",
       "\\item 'ADT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'RNA'\n",
       "2. 'ADT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"RNA\" \"ADT\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in CheckDuplicateCellNames(object.list = objects):\n",
      "“Some cell names are duplicated across objects provided. Renaming to enforce unique cell names.”\n"
     ]
    }
   ],
   "source": [
    "##load data of ID02\n",
    "#Put gene list in the target panel in the working directory (\"BD_genes_human\")\n",
    "\n",
    "###Input layer\n",
    "set.seed(seed = 42)\n",
    "\n",
    "sample.name <- \"ID28\"\n",
    "out.dir <- \"scTCR_raw.data\"\n",
    "\n",
    "################################Processing layer##########################################\n",
    "dir.create(out.dir)\n",
    "\n",
    "##Load targeted gene expression data\n",
    "input_name <- str_c(\"raw.data/\", sample.name, \"Target_results/matrix/\", sep = \"\") %>% \n",
    "                str_c(\"matrix_inflection_\", sample.name, \"Target.txt\", sep = \"\")\n",
    "compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "gunzip(compressed.name)\n",
    "#Create Seurat object\n",
    "matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "row.names(matrix) <- matrix$V1\n",
    "matrix <- dplyr::select(matrix, -V1)\n",
    "seu2 <- CreateSeuratObject(counts=matrix, project = \"seu\", min.cells = 3, min.features = 10)\n",
    "#Output cell barcode information for scTCR analysis\n",
    "name.output <- str_c(out.dir, sample.name, sep = \"/\") %>% str_c(\"BC_target.txt\", sep = \"_\")\n",
    "BC <- row.names(seu2@meta.data)\n",
    "write.table(BC, name.output, row.names = FALSE, quote = FALSE, col.names = FALSE)\n",
    "\n",
    "##Load protein expression data\n",
    "input_name <- str_c(\"raw.data/\", sample.name, \"Abseq_results/matrix/\", sep = \"\") %>% \n",
    "                str_c(\"Hashtag_top1M_\", sample.name, \"Abseq.txt\", sep = \"\")\n",
    "compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "gunzip(compressed.name)\n",
    "#Extract cells in SCT seurat object\n",
    "Abseq.matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "Abseq.matrix <- dplyr::filter(Abseq.matrix, CellBC %in% BC)\n",
    "row.names(Abseq.matrix) <- Abseq.matrix$CellBC\n",
    "Abseq.matrix <- dplyr::select(Abseq.matrix, -CellBC)\n",
    "Abseq.matrix <- t(Abseq.matrix)\n",
    "all.equal(rownames(seu2@meta.data), colnames(Abseq.matrix)) #Certify\n",
    "\n",
    "##Combine ADT data into SCT seurat object\n",
    "adt_assay <- CreateAssayObject(counts = Abseq.matrix)\n",
    "seu2[[\"ADT\"]] <- adt_assay\n",
    "Assays(seu2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3f68114b",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Merge data of ID02 and ID28\n",
    "\n",
    "#Input layer\n",
    "sample.name <- \"COVID_merged\"\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "seu <- merge(x=seu1, y = seu2, add.cell.ids = c(\"ID02\", \"ID28\"))\n",
    "\n",
    "name.output <- str_c(sample.name, \"rda\", sep = \".\")\n",
    "save(seu, name.output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "34ae7625",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'Seurat_plots' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scale for 'x' is already present. Adding another scale for 'x', which will\n",
      "replace the existing scale.\n",
      "\n",
      "Picking joint bandwidth of 0.0268\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Picking joint bandwidth of 5.33\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "###Gene filtering\n",
    "\n",
    "##Input layer\n",
    "dir.name=(\"Seurat_plots/\")\n",
    "sample.name <- \"COVID_merged\"\n",
    "gene.list.name <- \"BD_genes_human.txt\"\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "dir.create(dir.name)\n",
    "\n",
    "##load gene list of BD target panel\n",
    "BD_genes <- read.table(gene.list.name, header = TRUE)\n",
    "BD_genes <- as.vector(BD_genes$Genesymbol)\n",
    "\n",
    "##Chose BD target gene list for downstream analysis\n",
    "res1 = seu@assays$RNA@counts\n",
    "res1 =res1[res1@Dimnames[[1]] %in% BD_genes,]\n",
    "ngenes <- length(res1@Dimnames[[1]])\n",
    "seu@assays$RNA@counts <- res1\n",
    "seu@assays$RNA@data <- res1\n",
    "\n",
    "#Scatter plot for gene/read count\n",
    "file.name=paste(dir.name, sample.name, \"Reads.Genes.png\", sep='')\n",
    "png(file.name, width = 512, height = 400)\n",
    "FeatureScatter(object = seu, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\", pt.size = 0.3) +\n",
    "  geom_hline(yintercept=10) +\n",
    "  scale_x_log10()\n",
    "dev.off()\n",
    "\n",
    "#RidgePlot for gene/read count, \n",
    "file.name=paste(dir.name, sample.name, \"nCount_RNA_log.png\", sep='')\n",
    "png(file.name, width = 512, height = 400)\n",
    "RidgePlot(object = seu, features = \"nCount_RNA\", group.by=\"orig.ident\", ncol = 1) +\n",
    "  scale_x_log10()\n",
    "dev.off()\n",
    "\n",
    "file.name=paste(dir.name, sample.name, \"nGenes.png\", sep='')\n",
    "png(file.name, width = 512, height = 400)\n",
    "RidgePlot(object = seu, features = \"nFeature_RNA\", group.by=\"orig.ident\", ncol = 1)\n",
    "dev.off()\n",
    "\n",
    "#save Seurat object\n",
    "name.output <- str_c(sample.name, \"rda\", sep = \".\")\n",
    "save(seu, file=name.output)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93270836",
   "metadata": {},
   "source": [
    "### 2nd analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "67dda37b",
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"COVID_merged.rda\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed3cd178",
   "metadata": {},
   "source": [
    "### Exclude doublet using CD4/CD8 protein expression by ADT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "810f9da0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'S5A' already exists”\n",
      "Normalizing across cells\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Check doublet using CD4 and CD8 ADTs\n",
    "#CD4 and CD8 double positive cells are regarded as doublet\n",
    "\n",
    "###Input layer\n",
    "fea1 <- \"CD4\"\n",
    "fea2 <- \"CD8\"\n",
    "#Threshold for positive/negative\n",
    "xthres <- 3 #for fea1\n",
    "ythres1 <- 2.5 #for fea2\n",
    "ythres2 <- 1.5 #for fea2\n",
    "\n",
    "dir.name=\"S5A\"\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "DefaultAssay(seu) <- \"ADT\"\n",
    "dir.create(dir.name)\n",
    "\n",
    "#Normalize data\n",
    "seu <- NormalizeData(seu, normalization.method = \"CLR\", margin = 2, assay = \"ADT\")\n",
    "# Variable feature: All ADT Features\n",
    "VariableFeatures(seu) <- rownames(seu[[\"ADT\"]])\n",
    "#Scaling data \n",
    "seu = ScaleData(object = seu)\n",
    "\n",
    "# Draw ADT scatter plots (like biaxial plots for FACS). Note that you can even 'gate' cells if\n",
    "# desired by using HoverLocator and FeatureLocator\n",
    "file.name = str_c(dir.name, \"FACSPlot\", sep='/') %>% str_c(fea1, fea2, \"png\", sep = \".\")\n",
    "png(file.name, width = 360, height = 360)\n",
    "p <- FeatureScatter(seu, feature1 = fea1, feature2 = fea2)\n",
    "p <- p + geom_hline(yintercept = ythres1, size=0.5) +\n",
    "        geom_hline(yintercept = ythres2, size=0.5) +         \n",
    "        geom_vline(xintercept = xthres, size=0.5)\n",
    "plot(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570a18e9",
   "metadata": {},
   "source": [
    "### Separate CD4 and CD8 cells and dimentional reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "1e8ad719",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'Seurat_plots_2nd' already exists”\n",
      "Regressing out nCount_RNA\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 4 features requested have zero variance (running reduction without them): AQP9, CCL17, IL33, EPX”\n",
      "PC_ 1 \n",
      "Positive:  GZMH, GZMB, FCGR3A, GNLY, NKG7, LGALS1, LAIR2, ITGAM, KLRF1, IFNG \n",
      "\t   B3GAT1, CCL5, ZNF683, CCL4, CX3CR1, KLRC3, CST7, TBX21, PRF1, CCL3 \n",
      "\t   CTSW, HLA-DPA1, CD63, ITGB2, GZMA, CD300A, HLA-DRA, APOBEC3G, CD244, LYN \n",
      "Negative:  IL7R, GZMK, TCF7, KLRB1, PIK3IP1, DUSP1, MYC, DUSP2, LTB, JUNB \n",
      "\t   BTG1, CXCR4, ZBTB16, FOSB, TNFRSF25, CD27, CD28, JUN, CD69, DPP4 \n",
      "\t   CCR7, LEF1, RORC, CD44, HLA-A, CD48, IL18RAP, TRAT1, PASK, FTH1 \n",
      "PC_ 2 \n",
      "Positive:  TYMS, TOP2A, LEF1, MCM4, UBE2C, PASK, HLA-DRA, MKI67, AURKB, SELL \n",
      "\t   MCM2, CXCR3, CTLA4, ICOS, LGALS9, HMMR, CD8B, PTTG2, CCR7, LGALS1 \n",
      "\t   YBX3, BTLA, CD27, PCNA, ZNF683, RGS1, HLA-DMA, TRIB2, ADA, CD79A \n",
      "Negative:  KLRB1, CST7, NKG7, BTG1, GZMA, GNLY, CXCR4, DUSP1, KLRG1, ARL4C \n",
      "\t   PRF1, ZBTB16, KLRF1, RUNX3, CTSW, PRDM1, IKZF2, NCR3, IL7R, IL18RAP \n",
      "\t   CD63, CCL5, XBP1, PIK3AP1, RORC, CD3E, IFNGR1, GZMK, IL2RB, DUSP2 \n",
      "PC_ 3 \n",
      "Positive:  LEF1, CD8B, RGS1, CCR7, SELL, YBX3, KLRK1, CD37, TRAC, BTG1 \n",
      "\t   CD5, CD8A, CD69, PIK3IP1, LCK, PASK, S100A10, FOSB, CD3E, CD3D \n",
      "\t   CD27, TCF7, JUN, PTPRC, KLRC3, HLA-DQB1, JUNB, CD44, BIN2, DUSP1 \n",
      "Negative:  KLRB1, TYMS, ZBTB16, MCM4, UBE2C, CCL5, IL18RAP, TOP2A, MCM2, AURKB \n",
      "\t   NCR3, RORC, MKI67, HMMR, NKG7, DPP4, GZMK, TNFRSF25, PRF1, GZMA \n",
      "\t   PTTG2, PMCH, PCNA, IL18R1, CCR2, PDIA4, CXCR6, KLRG1, GNLY, PIK3AP1 \n",
      "PC_ 4 \n",
      "Positive:  ZNF683, CD52, LTB, IFITM2, TNFRSF25, LGALS1, LAIR2, ITGAM, LEF1, NCR3 \n",
      "\t   IL18RAP, ITGAX, GZMB, TRBC2, CD79B, TSPAN32, TRIB2, ZBTB16, PASK, KLRB1 \n",
      "\t   IL7R, IFITM3, CD79A, RORC, CD6, IL32, TGFB1, CD7, DPP4, B3GAT1 \n",
      "Negative:  TYMS, TOP2A, UBE2C, MCM4, MCM2, AURKB, HMMR, PCNA, MKI67, HLA-DQB1 \n",
      "\t   KLRC3, KLRF1, CST7, GZMK, CD74, PTTG2, GZMA, CXCR4, CCL3, HLA-DPA1 \n",
      "\t   RGS1, CCL4, TIGIT, IL2RB, CXCR3, HLA-DMA, RUNX3, FYN, YBX3, FOSB \n",
      "PC_ 5 \n",
      "Positive:  CCL5, GZMK, CST7, CCL4, DUSP2, CD27, RGS1, CHI3L2, HLA-DMA, CD74 \n",
      "\t   NKG7, HLA-DQB1, CCL3, HLA-DRA, GZMA, CCR5, CXCR3, MS4A1, TGFBI, CXCR5 \n",
      "\t   CBLB, TIGIT, EGR1, PDCD1, CD38, FASLG, IFNG, IFNGR1, VNN2, CTLA4 \n",
      "Negative:  CD52, IFITM2, TYMS, UBE2C, MCM2, TOP2A, MCM4, S100A10, LGALS1, GNLY \n",
      "\t   PDIA6, ZNF683, LEF1, PCNA, AURKB, HMMR, PDIA4, MKI67, XBP1, ITGAM \n",
      "\t   CD63, LGALS3, LTB, TRAC, RPN2, CTSW, GZMB, IL7R, PTTG2, HLA-A \n",
      "\n",
      "Normalizing across cells\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“You're computing too large a percentage of total singular values, use a standard svd instead.”\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“did not converge--results might be invalid!; try increasing work or maxit”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "PC_ 1 \n",
      "Positive:  HLA-DR, CD8, CD45RA, CD278, CD11c, CD62L, IgM, CD272, CD16, Tim3 \n",
      "\t   CD137, CD183, CD3, IgD, CD14 \n",
      "Negative:  CD161, CD127, CD196, CD28, CXCR6, CD27, CD25, CD4, CD56, CCR7 \n",
      "\t   CXCR5, CD134, GITR, CD19, CD279 \n",
      "PC_ 2 \n",
      "Positive:  CD27, CD183, CD28, CD279, CCR7, CD278, CD272, CD8, CD25, CXCR5 \n",
      "\t   HLA-DR, CD137, CD134, CD127, CD62L \n",
      "Negative:  CD45RA, CD56, CD161, CD196, IgM, IgD, CD19, CD4, CD11c, CD3 \n",
      "\t   CXCR6, CD16, GITR, CD14, Tim3 \n",
      "PC_ 3 \n",
      "Positive:  CD272, CCR7, CD25, CD3, CD4, CD8, CD62L, CD45RA, CD127, CXCR5 \n",
      "\t   CD183, CD28, CD19, CD134, CD56 \n",
      "Negative:  CD279, CD137, CD16, CD11c, CXCR6, HLA-DR, Tim3, CD14, CD278, CD161 \n",
      "\t   CD27, IgM, CD196, IgD, GITR \n",
      "PC_ 4 \n",
      "Positive:  IgD, CD19, IgM, CXCR5, CCR7, CD25, GITR, CD134, CD28, CD183 \n",
      "\t   CD127, CD137, CD196, CD27, CD272 \n",
      "Negative:  CD4, CD3, CD11c, CD14, CD62L, CD278, Tim3, CD279, CD16, HLA-DR \n",
      "\t   CD56, CXCR6, CD45RA, CD8, CD161 \n",
      "PC_ 5 \n",
      "Positive:  CD3, CXCR5, CD279, CXCR6, CD8, CD183, CD4, CD19, CD45RA, CD27 \n",
      "\t   HLA-DR, CD56, IgD, IgM, CD28 \n",
      "Negative:  CD25, CD14, CD134, GITR, CD16, CD11c, CD272, Tim3, CD62L, CD196 \n",
      "\t   CCR7, CD127, CD278, CD161, CD137 \n",
      "\n",
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: PC_), setting key to 'apca_'”\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 12533\n",
      "Number of edges: 399877\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9033\n",
      "Number of communities: 19\n",
      "Elapsed time: 8 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6 singletons identified. 13 final clusters.\n",
      "\n",
      "Warning message:\n",
      "“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session”\n",
      "14:36:46 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "14:36:46 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "14:36:47 Initializing from normalized Laplacian + noise\n",
      "\n",
      "14:36:47 Commencing optimization for 200 epochs, with 389854 positive edges\n",
      "\n",
      "14:36:52 Optimization finished\n",
      "\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 11\n",
      "\n",
      "Calculating cluster 12\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 11\n",
      "\n",
      "Calculating cluster 12\n",
      "\n",
      "Regressing out nCount_RNA\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 4 features requested have zero variance (running reduction without them): CD1C, CXCL10, CLEC10A, IGLC3”\n",
      "PC_ 1 \n",
      "Positive:  GZMA, CCL5, NKG7, GZMH, CST7, GZMK, PRF1, GNLY, KLRG1, GZMB \n",
      "\t   CTSW, ITGAM, TBX21, IFNG, IL18RAP, HLA-DPA1, RUNX3, CCL4, HLA-DRA, EOMES \n",
      "\t   CD300A, CX3CR1, LGALS1, PRDM1, LGALS3, APOBEC3G, B3GAT1, TNF, CCR5, CCL3 \n",
      "Negative:  LEF1, SELL, TCF7, CCR7, CD27, PIK3IP1, PASK, TRIB2, LTB, TXK \n",
      "\t   JUNB, CXCR5, CD7, LGALS9, CD5, FOXP1, CD37, CD69, ICOS, BIRC3 \n",
      "\t   LAT, TRAT1, CD28, CD44, CD48, CD6, LAP3, TRAC, LCK, FYN \n",
      "PC_ 2 \n",
      "Positive:  FOXP3, HLA-DRA, CTLA4, DUSP4, IL32, LGALS1, HLA-DQB1, IKZF2, HLA-DMA, LAIR2 \n",
      "\t   HLA-DPA1, LAP3, IL2RA, F5, TOP2A, PTTG2, LGALS9, TIGIT, TYMS, CD52 \n",
      "\t   NINJ2, CD79B, CD74, TRIB2, MKI67, PASK, ANXA5, MCM4, UBE2C, AURKB \n",
      "Negative:  IL7R, BTG1, JUN, DUSP1, CXCR4, FOSB, ARL4C, GZMK, TCF7, KLRB1 \n",
      "\t   CCL5, CD48, XBP1, PIK3IP1, DPP4, DUSP2, RUNX3, CD69, GZMA, CTSW \n",
      "\t   STAT4, TRAT1, CD44, CD3E, KLRG1, CD6, IFNGR1, GNLY, FYN, EGR1 \n",
      "PC_ 3 \n",
      "Positive:  ITGA4, PASK, TCF7, GZMH, IFNG, NKG7, CHI3L2, CXCR5, ADA, CCR7 \n",
      "\t   GZMB, TYMS, TOP2A, TXK, ZAP70, CCL5, TBX21, IL23R, CD200, ZNF683 \n",
      "\t   CD7, MCM2, MCM4, CD6, CX3CR1, UBE2C, CD300A, KLRB1, BCL2A1, B3GAT1 \n",
      "Negative:  FOXP3, LGALS3, DUSP4, CXCR4, HLA-DQB1, PRDM1, LGALS1, IL2RA, IKZF2, DUSP1 \n",
      "\t   RGS1, S100A10, BTG1, CD44, HLA-DPA1, JUN, PDIA6, F5, SELPLG, FOSB \n",
      "\t   ARL4C, ANXA5, IL2RB, CCR10, RUNX3, TRAC, CD63, HLA-DRA, NINJ2, RORC \n",
      "PC_ 4 \n",
      "Positive:  UBE2C, TYMS, MKI67, MCM4, MCM2, FOSB, AURKB, HMMR, PCNA, DUSP1 \n",
      "\t   TOP2A, CXCR4, JUN, IL7R, TMEM97, JUNB, PTTG2, EGR1, DPP4, RNASE6 \n",
      "\t   PMCH, LGALS3, MZB1, EGR3, TNFSF13B, S100A9, RORC, PDIA4, BTG1, TNF \n",
      "Negative:  GZMH, LAIR2, NKG7, CD74, ITGB2, GZMB, LCK, CD27, CD3D, CST7 \n",
      "\t   LAT, BIN2, CD2, GAPDH, CTLA4, SELPLG, CD37, LGALS9, CCL4, CD3E \n",
      "\t   ITGAM, HLA-DPA1, CD52, SELL, CD3G, GNAI2, TRIB2, TBX21, CD5, CD4 \n",
      "PC_ 5 \n",
      "Positive:  FOXP3, IKZF2, GZMK, CTLA4, TIGIT, CD27, IL2RB, TXK, IL2RA, IL18R1 \n",
      "\t   CST7, IL18RAP, CXCR6, CBLB, FYN, LAIR2, NRP1, CXCR5, CCL5, PIK3IP1 \n",
      "\t   IGHM, NCR3, PRF1, CCR7, CD8A, F5, EGR1, EOMES, GNLY, IL23R \n",
      "Negative:  LGALS1, S100A10, CD52, IFITM2, ANXA5, CD63, GAPDH, CD4, ITGB2, CD6 \n",
      "\t   GNAI2, TGFB1, ITGAM, PTGDR2, TYMS, TNFSF10, RPN2, PCNA, TRAT1, IKZF1 \n",
      "\t   CD3D, ARL4C, MCM4, HLA-DPA1, RNASE6, IL32, LAT, IFNG, CD3E, ADA \n",
      "\n",
      "Normalizing across cells\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“You're computing too large a percentage of total singular values, use a standard svd instead.”\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“did not converge--results might be invalid!; try increasing work or maxit”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (30). Setting to 30.”\n",
      "PC_ 1 \n",
      "Positive:  CCR7, CD272, CD27, CD3, CD62L, CXCR5, CD278, CD28, CD183, CD4 \n",
      "\t   Tim3, CD279, IgD, CD19, CD127 \n",
      "Negative:  CD196, GITR, CXCR6, CD11c, CD8, CD16, CD161, CD25, CD56, CD14 \n",
      "\t   HLA-DR, CD137, IgM, CD134, CD45RA \n",
      "PC_ 2 \n",
      "Positive:  CD127, CD4, CD161, CD28, CD196, CCR7, CD16, CD3, CXCR5, CD8 \n",
      "\t   CD183, CXCR6, CD11c, CD272, CD27 \n",
      "Negative:  HLA-DR, CD62L, CD278, CD137, GITR, CD279, CD25, IgD, IgM, Tim3 \n",
      "\t   CD45RA, CD134, CD19, CD14, CD56 \n",
      "PC_ 3 \n",
      "Positive:  CD25, GITR, CD134, CD28, CD137, CD196, CD27, CD4, CD62L, CD3 \n",
      "\t   CCR7, CXCR5, CD127, CD161, CD278 \n",
      "Negative:  CD45RA, CD183, IgD, CD19, CD279, IgM, CD11c, CD56, CXCR6, CD272 \n",
      "\t   CD14, HLA-DR, CD8, Tim3, CD16 \n",
      "PC_ 4 \n",
      "Positive:  CXCR5, CD25, CCR7, CD45RA, CD27, CD62L, IgM, IgD, CD19, CD8 \n",
      "\t   CD56, CD137, CD127, CD16, Tim3 \n",
      "Negative:  CD279, CD278, CD28, CD4, CD3, HLA-DR, CD183, CD272, CXCR6, CD134 \n",
      "\t   GITR, CD11c, CD196, CD14, CD161 \n",
      "PC_ 5 \n",
      "Positive:  CD4, CD25, CD28, CD3, CD272, CD8, HLA-DR, CD62L, CD11c, IgM \n",
      "\t   CD19, CD127, CD16, CD45RA, IgD \n",
      "Negative:  CD161, CD279, CXCR5, CD134, CD27, CD196, Tim3, GITR, CD183, CD278 \n",
      "\t   CXCR6, CD56, CCR7, CD137, CD14 \n",
      "\n",
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: PC_), setting key to 'apca_'”\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 17588\n",
      "Number of edges: 445472\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8769\n",
      "Number of communities: 22\n",
      "Elapsed time: 12 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10 singletons identified. 12 final clusters.\n",
      "\n",
      "14:45:36 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "14:45:36 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "14:45:37 Initializing from normalized Laplacian + noise\n",
      "\n",
      "14:45:37 Commencing optimization for 200 epochs, with 539736 positive edges\n",
      "\n",
      "14:45:44 Optimization finished\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 11\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Warning message in FindMarkers.default(object = data.use, slot = data.slot, counts = counts, :\n",
      "“No features pass logfc.threshold threshold; returning empty data.frame”\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 11\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Exclude doublet /normalizing RNA and Abseq data / PCA and JackStraw for RNA data\n",
    "\n",
    "##Input layer\n",
    "dir.name=(\"Seurat_plots_2nd/\")\n",
    "sample.CD4 <- \"COVID_merged_CD4\"\n",
    "sample.CD8 <- \"COVID_merged_CD8\"\n",
    "resol <- 0.5 # resolution for clustering\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "dir.create(dir.name)\n",
    "##Separate CD4 and CD8 cells\n",
    "DefaultAssay(seu) <- \"ADT\"\n",
    "seu.CD8 <- subset(seu, subset = CD4 < 3 & CD8 > 1.5)\n",
    "seu.CD4 <- subset(seu, subset = CD4 > 3 & CD8 < 2.5)\n",
    "\n",
    "seus <- list(seu.CD8, seu.CD4)\n",
    "sample.names <- c(sample.CD8, sample.CD4)\n",
    "\n",
    "##Process Seurat object (CD4 and CD8) with iteration\n",
    "for(i in 1:length(seus)){\n",
    "    #load seurat object\n",
    "    seu <- seus[[i]]\n",
    "    sample.name <- sample.names[i]\n",
    "    \n",
    "    ##For RNA\n",
    "    DefaultAssay(seu) <- \"RNA\"\n",
    "    #Normalizing and scaling data\n",
    "    seu = NormalizeData(object = seu, scale.factor=1000000)\n",
    "    VariableFeatures(seu) <- rownames(seu[[\"RNA\"]]) #Use all genes in target panel for Variable features\n",
    "    ngenes <- length(x = seu@assays$RNA@var.features)\n",
    "    all.genes <- rownames(seu)\n",
    "    seu = ScaleData(object = seu, vars.to.regress = c(\"nCount_RNA\"), features = all.genes) \n",
    "    #Perform PCA\n",
    "    seu = RunPCA(object = seu, features = seu@assays$RNA@var.features, npcs = 100)\n",
    "    #JackStraw\n",
    "    seu = JackStraw(object = seu, num.replicate = 100, dims = 50)\n",
    "    seu <- ScoreJackStraw(object = seu, dims = 1:50, score.thresh = 0.05)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='') %>% str_c(\"Jackstraw.png\", sep='.')\n",
    "    png(file.name, width = 1250, height = 500)\n",
    "    JackStrawPlot(object = seu, dims = 1:50)\n",
    "    dev.off()\n",
    "    #Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "    #Extract PCs which fulfill the pvalue threshold\n",
    "    tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "    tmp1 = tmp[tmp$Score>0.05,1]\n",
    "    dims= c(1:(min(tmp1)-1))\n",
    "    dims\n",
    "\n",
    "    ##For Abseq\n",
    "    DefaultAssay(seu) <- 'ADT'\n",
    "    #Normalizing and scaling data\n",
    "    seu <- NormalizeData(seu, normalization.method = \"CLR\", margin = 2, assay = \"ADT\")\n",
    "    VariableFeatures(seu) <- rownames(seu[[\"ADT\"]])\n",
    "    seu = ScaleData(object = seu) \n",
    "    #Perform PCA\n",
    "    seu = RunPCA(object = seu, reduction.name = 'apca')\n",
    "    \n",
    "    #Clustering Weighted Nearest Neighbor\n",
    "    seu <- FindMultiModalNeighbors(seu, reduction.list = list(\"pca\", \"apca\"), \n",
    "                                   dims.list = list(dims, dims))\n",
    "    seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resol)\n",
    "\n",
    "    #Dimentional reduction\n",
    "    seu <- RunUMAP(seu, nn.name = \"weighted.nn\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "    #Visualization using UMAP\n",
    "    p1 = DimPlot(object = seu, reduction = \"wnn.umap\", label = TRUE, label.size = 10, pt.size = 0.5) +\n",
    "                theme(axis.title.x = element_text(size=10, family = \"Arial\"), \n",
    "                        axis.title.y = element_text(size=10, family = \"Arial\"), \n",
    "                        axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"), \n",
    "                        axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "                theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "    p2 = DimPlot(object = seu, reduction = \"wnn.umap\", label = FALSE, label.size = 10, pt.size = 0.5, group.by = \"orig.ident\") +\n",
    "                theme(axis.title.x = element_text(size=10, family = \"Arial\"),\n",
    "                        axis.title.y = element_text(size=10, family = \"Arial\"),\n",
    "                        axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"),\n",
    "                        axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "                theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "    legend1 <- cowplot::get_legend(p1)\n",
    "    legend2 <- cowplot::get_legend(p2)\n",
    "    p1 = p1 + theme(legend.position = 'none')\n",
    "    p2 = p2 + theme(legend.position = 'none')\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"wnn.umap_reso\", resol, \".png\", sep='')\n",
    "    save_plot(file = file.name, plot_grid(p1, legend1, p2, legend2, ncol=2, nrow=2), device=\"png\", \n",
    "              units=\"in\", dpi = 600, base_width = 10, base_height = 10, limitsize=FALSE)\n",
    "\n",
    "    ###Extract marker genes\n",
    "    ##For RNA\n",
    "    DefaultAssay(seu) <- 'RNA'\n",
    "    seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "    #Create heatmap with top10 marker genes\n",
    "    top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "    top10 = as.data.frame(top10)\n",
    "    top10  = top10 [!duplicated(top10$gene),]\n",
    "    top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    top10 = as.data.frame(top10)\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.marker_res\", resol, \".png\", sep='')\n",
    "    p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "    ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "           width = 20, height = 20, limitsize=FALSE)\n",
    "    #Output marker gene table\n",
    "    seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "    seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    seu.markers = as.data.frame(seu.markers)\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "    fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "    save(seu.markers, file=file.name)\n",
    "\n",
    "    ##For ADT\n",
    "    DefaultAssay(seu) <- 'ADT'\n",
    "    seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "    #Create heatmap with top10 marker genes\n",
    "    top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "    top10 = as.data.frame(top10)\n",
    "    top10  = top10 [!duplicated(top10$gene),]\n",
    "    top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    top10 = as.data.frame(top10)\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.marker_res\", resol, \".png\", sep='')\n",
    "    p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "    ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "           width = 20, height = 20, limitsize=FALSE)\n",
    "    #Output marker gene table\n",
    "    seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "    seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    seu.markers = as.data.frame(seu.markers)\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "    fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "    file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "    save(seu.markers, file=file.name)\n",
    "\n",
    "    #Output Seurat object in initial analysis\n",
    "    file.name=paste(sample.name, \".wnn.res\", resol, \".rda\", sep='')\n",
    "    save(seu, file=file.name) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "918f54cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'MarkerGene.Plot_2nd' already exists”\n"
     ]
    }
   ],
   "source": [
    "#Check marker genes for lineage using RNA expression\n",
    "sample.names <- c(\"COVID_merged_CD4.wnn.res0.5\", \"COVID_merged_CD8.wnn.res0.5\")\n",
    "tmp_rna = c(\"TRBC2\", \"TRDC\", \"CD14\", \"MS4A1\", \"KLRC1\", \"CHI3L1\")\n",
    "tmp_adt = c(\"CD3\", \"CD19\", \"CD16\", \"CD14\", \"CD56\", \"CD161\")\n",
    "dir.name=\"MarkerGene.Plot_2nd\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "for(i in sample.names){\n",
    "    load(str_c(i, \"rda\", sep = \".\"))\n",
    "    \n",
    "    ##Plotting RNA expression\n",
    "    DefaultAssay(seu) <- \"RNA\"\n",
    "    #Scatter plot\n",
    "    tmp_name <- paste(i, \"Scatter.RNA.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- FeaturePlot(seu, features = tmp_rna, ncol = 6, order = TRUE,\n",
    "                     reduction = \"wnn.umap\", dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.2)\n",
    "    plot(p)\n",
    "    dev.off()    \n",
    "    #Violin plot\n",
    "    tmp_name <- paste(i, \"Violin.RNA.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- VlnPlot(seu, features = tmp_rna, ncol = 6, pt.size = 0.1)\n",
    "    plot(p)\n",
    "    dev.off()\n",
    "    \n",
    "    ##Plotting Protein expression\n",
    "    DefaultAssay(seu) <- \"ADT\"\n",
    "    #Scatter plot\n",
    "    tmp_name <- paste(i, \"Scatter.ADT.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- FeaturePlot(seu, features = tmp_adt, ncol = 6, order = TRUE,\n",
    "                     reduction = \"wnn.umap\", dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.2)\n",
    "    plot(p)\n",
    "    dev.off()    \n",
    "    #Violin plot\n",
    "    tmp_name <- paste(i, \"Violin.ADT.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- VlnPlot(seu, features = tmp_adt, ncol = 6, pt.size = 0.1)\n",
    "    plot(p)\n",
    "    dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "753f88a4",
   "metadata": {},
   "source": [
    "### 3rd analysis for CD8+ T cells with removing contaminant non-abT cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "20c073d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"COVID_merged_CD8.wnn.res0.5.rda\")\n",
    "\n",
    "###Remove contaminant cluster and repeat Seurat pipelines, define cluster resolution\n",
    "\n",
    "###Input layer \n",
    "sample.name <- \"COVID_merged_CD8_3rd\"\n",
    "dir.name <- \"Seurat_plots_3rd\"\n",
    "\n",
    "#Choose subcluster for analyse\n",
    "#Exclude cluster 16 (contaminant of myeoid cells)\n",
    "DefaultAssay(seu) <- \"RNA\"\n",
    "seu <- subset(seu, idents = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11))\n",
    "\n",
    "#For defining clustering resolution\n",
    "resolutions = c(0.1, 0.2, 0.3, 0.4, 0.5)\n",
    "resols.name = \"minute\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "\n",
    "##For Abseq, Normalize ~ PCA\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Normalizing and scaling data\n",
    "seu <- NormalizeData(seu, normalization.method = \"CLR\", margin = 2, assay = \"ADT\")\n",
    "#Exclude markers from Variable features that were used in CD4/CD8 separation and contami detection\n",
    "ADT_list <- rownames(seu[[\"ADT\"]])\n",
    "VariableFeatures(seu) <- ADT_list[-which(ADT_list %in% c(\"CD4\", \"CD8\", \"CD19\", \"IgD\", \"IgM\"))] \n",
    "seu = ScaleData(object = seu) \n",
    "#Perform PCA：\n",
    "seu = RunPCA(object = seu, reduction.name = 'apca')\n",
    "\n",
    "##For RNA\n",
    "DefaultAssay(seu) <- \"RNA\"\n",
    "#Normalizing and scaling data\n",
    "seu = NormalizeData(object = seu, scale.factor=1000000)\n",
    "VariableFeatures(seu) <- rownames(seu[[\"RNA\"]]) #Use all genes in target panel for Variable features\n",
    "ngenes <- length(x = seu@assays$RNA@var.features)\n",
    "all.genes <- rownames(seu)\n",
    "seu = ScaleData(object = seu, vars.to.regress = c(\"nCount_RNA\"), features = all.genes) \n",
    "#Perform PCA\n",
    "seu = RunPCA(object = seu, features = seu@assays$RNA@var.features, npcs = 100)\n",
    "#JackStraw\n",
    "seu = JackStraw(object = seu, num.replicate = 100, dims = 50)\n",
    "seu <- ScoreJackStraw(object = seu, dims = 1:50, score.thresh = 0.05)\n",
    "file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Jackstraw.png\", sep='.')\n",
    "png(file.name, width = 1250, height = 500)\n",
    "JackStrawPlot(object = seu, dims = 1:50)\n",
    "dev.off()\n",
    "#Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "#Extract PCs which fulfill the pvalue threshold\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "dims\n",
    "\n",
    "#Clustering Weighted Nearest Neighbor\n",
    "seu <- FindMultiModalNeighbors(\n",
    "  seu, reduction.list = list(\"pca\", \"apca\"), \n",
    "  dims.list = list(dims, dims))\n",
    "#Dimentional reduction\n",
    "seu <- RunUMAP(seu, nn.name = \"weighted.nn\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "#Calculate Shilouhette score\n",
    "distance_matrix = parDist(seu@reductions$wnn.umap@cell.embeddings, method = \"euclidean\", threads=4)\n",
    "silhouette_score = NULL\n",
    "silhouette_score_mean = list()\n",
    "\n",
    "for (i in c(1:length(resolutions))){\n",
    "  #Perform clustering\n",
    "  seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resolutions[i])\n",
    "  silhouette = silhouette(as.numeric(seu@active.ident), dist = distance_matrix)\n",
    "  silhouette = silhouette[,3]\n",
    "  silhouette_score = c(silhouette_score, mean(silhouette))\n",
    "  names(silhouette_score)[i]=resolutions[i]\n",
    "  res <- vector()\n",
    "  for (j in c(0:length(unique(seu@active.ident))-1)){\n",
    "    tmp = mean(silhouette[seu@active.ident %in% j]) \n",
    "    res = c(res, tmp)\n",
    "  }\n",
    "  silhouette_score_mean[[i]]=res\n",
    "  names(silhouette)=rownames(seu@meta.data)\n",
    "  silhouette.name = sprintf(\"silhouette_score.res.%s\", resolutions[i])\n",
    "  seu =  AddMetaData(object = seu, metadata = silhouette, col.name = silhouette.name)\n",
    "}\n",
    "\n",
    "max_silhouette = names(silhouette_score[max(silhouette_score)])\n",
    "\n",
    "x = data.frame(resolution = resolutions,\n",
    "               mean_silhouette_score = silhouette_score)\n",
    "\n",
    "p_silhouette = ggplot(x, aes(x = resolution, y = mean_silhouette_score)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  ggtitle(\"Mean silhouette scores of clustering\") +\n",
    "  theme_linedraw() + \n",
    "  theme(plot.title=element_text(hjust = 0.5), text=element_text(size=12)) + \n",
    "  theme(axis.text.x=element_text(size=12), axis.text.y=element_text(size=12)) \n",
    "file.name=paste(sample.name, resols.name, \"_silhouette.png\", sep='')\n",
    "file.name=paste(dir.name, file.name, sep='/')\n",
    "ggsave(file = file.name, plot = p_silhouette, device=\"png\", units=\"in\", dpi = 300,\n",
    "       width = 4, height = 3, limitsize=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d7ada517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 12392\n",
      "Number of edges: 397052\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9593\n",
      "Number of communities: 12\n",
      "Elapsed time: 10 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6 singletons identified. 6 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    }
   ],
   "source": [
    "#Clustering using RNA and ADTinformation \n",
    "\n",
    "##Input layer\n",
    "dir.name=(\"Seurat_plots_3rd/\")\n",
    "sample.name <- \"COVID_merged_CD8_3rd\"\n",
    "resol <- 0.1 # resolution for clustering\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "#Clustering Weighted Nearest Neighbor\n",
    "seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resol)\n",
    "\n",
    "#Visualization using UMAP\n",
    "p1 = DimPlot(object = seu, reduction = \"wnn.umap\", label = TRUE, label.size = 10, pt.size = 0.5) +\n",
    "  theme(axis.title.x = element_text(size=10, family = \"Arial\"), \n",
    "        axis.title.y = element_text(size=10, family = \"Arial\"), \n",
    "        axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"), \n",
    "        axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "p2 = DimPlot(object = seu, reduction = \"wnn.umap\", label = FALSE, label.size = 10, pt.size = 0.5, group.by = \"orig.ident\") +\n",
    "  theme(axis.title.x = element_text(size=10, family = \"Arial\"),\n",
    "        axis.title.y = element_text(size=10, family = \"Arial\"),\n",
    "        axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"),\n",
    "        axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "\n",
    "legend1 <- cowplot::get_legend(p1)\n",
    "legend2 <- cowplot::get_legend(p2)\n",
    "p1 = p1 + theme(legend.position = 'none')\n",
    "p2 = p2 + theme(legend.position = 'none')\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"wnn.umap_reso\", resol, \".png\", sep='')\n",
    "save_plot(file = file.name, plot_grid(p1, legend1, p2, legend2, ncol=2, nrow=2), device=\"png\", \n",
    "          units=\"in\", dpi = 600, base_width = 10, base_height = 10, limitsize=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "fb0b6d78",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Extract marker genes for RNA and ADT\n",
    "\n",
    "###Input layer\n",
    "dir.name <- \"Seurat_plots_3rd\"\n",
    "sample.name <- \"COVID_merged_CD8_3rd\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "\n",
    "##For RNA\n",
    "DefaultAssay(seu) <- 'RNA'\n",
    "seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "#Create heatmap with top10 marker genes\n",
    "top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "top10 = as.data.frame(top10)\n",
    "top10  = top10 [!duplicated(top10$gene),]\n",
    "top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "top10 = as.data.frame(top10)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.marker_res\", resol, \".png\", sep='')\n",
    "p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "       width = 20, height = 20, limitsize=FALSE)\n",
    "#Output marker gene table\n",
    "seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "seu.markers = as.data.frame(seu.markers)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "save(seu.markers, file=file.name)\n",
    "\n",
    "##For ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "#Create heatmap with top10 marker genes\n",
    "top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "top10 = as.data.frame(top10)\n",
    "top10  = top10 [!duplicated(top10$gene),]\n",
    "top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "top10 = as.data.frame(top10)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.marker_res\", resol, \".png\", sep='')\n",
    "p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "       width = 20, height = 20, limitsize=FALSE)\n",
    "#Output marker gene table\n",
    "seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "seu.markers = as.data.frame(seu.markers)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "save(seu.markers, file=file.name)\n",
    "\n",
    "#Output Seurat object in initial analysis\n",
    "file.name=paste(sample.name, \".wnn.res\", resol, \".rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17a82a65",
   "metadata": {},
   "source": [
    "### 3rd analysis for CD4+ T cells with removing contaminant non-abT cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "45b7c096",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'Seurat_plots_3rd' already exists”\n",
      "Normalizing across cells\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“You're computing too large a percentage of total singular values, use a standard svd instead.”\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“did not converge--results might be invalid!; try increasing work or maxit”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (25). Setting to 25.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (25). Setting to 25.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (25). Setting to 25.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (25). Setting to 25.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (25). Setting to 25.”\n",
      "PC_ 1 \n",
      "Positive:  CCR7, CD272, CD27, CD3, CXCR5, CD62L, CD278, CD183, CD28, Tim3 \n",
      "\t   CD279, CD127 \n",
      "Negative:  GITR, CD196, CXCR6, CD11c, CD25, CD16, CD161, HLA-DR, CD14, CD56 \n",
      "\t   CD137, CD134 \n",
      "PC_ 2 \n",
      "Positive:  CD127, CD161, CD196, CD28, CD16, CXCR5, CCR7, CXCR6, CD183, CD11c \n",
      "\t   CD3, CD56 \n",
      "Negative:  HLA-DR, CD62L, CD278, CD137, GITR, CD279, CD25, Tim3, CD45RA, CD134 \n",
      "\t   CD272, CD14 \n",
      "PC_ 3 \n",
      "Positive:  CD25, GITR, CD134, CD137, CD27, CD28, CD196, CD62L, CCR7, CXCR5 \n",
      "\t   CD127, CD161 \n",
      "Negative:  CD279, CD183, CD45RA, HLA-DR, CXCR6, CD11c, CD278, CD272, CD14, CD56 \n",
      "\t   Tim3, CD16 \n",
      "PC_ 4 \n",
      "Positive:  CD45RA, CD62L, CD25, CD56, CD11c, CCR7, CXCR5, CD16, CD14, CD137 \n",
      "\t   Tim3, CD27 \n",
      "Negative:  CD279, CD28, CD134, CD278, GITR, CD196, CD3, CD161, CD272, CXCR6 \n",
      "\t   CD183, HLA-DR \n",
      "PC_ 5 \n",
      "Positive:  CD28, CD11c, CD3, CD16, CD14, CD127, CD272, HLA-DR, CD25, CD137 \n",
      "\t   CD62L, CXCR6 \n",
      "Negative:  CXCR5, CD161, CD279, CD45RA, CD134, CD56, CD27, GITR, Tim3, CD196 \n",
      "\t   CCR7, CD183 \n",
      "\n",
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: PC_) setting key to original value 'apca_'”\n",
      "Regressing out nCount_RNA\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 4 features requested have zero variance (running reduction without them): CD1C, CXCL10, CLEC10A, IGLC3”\n",
      "PC_ 1 \n",
      "Positive:  GZMA, CCL5, NKG7, GZMH, CST7, GZMK, PRF1, GNLY, KLRG1, GZMB \n",
      "\t   CTSW, ITGAM, TBX21, IFNG, HLA-DPA1, CCL4, HLA-DRA, IL18RAP, RUNX3, CX3CR1 \n",
      "\t   EOMES, LGALS1, CD300A, PRDM1, LGALS3, APOBEC3G, B3GAT1, TNF, FCGR3A, ZNF683 \n",
      "Negative:  LEF1, TCF7, SELL, CCR7, CD27, PIK3IP1, PASK, TRIB2, LTB, TXK \n",
      "\t   JUNB, CXCR5, CD7, LGALS9, CD5, FOXP1, CD37, CD69, ICOS, BIRC3 \n",
      "\t   LAT, TRAT1, CD44, CD28, CD48, CD6, TRAC, LCK, FYN, HLA-A \n",
      "PC_ 2 \n",
      "Positive:  FOXP3, HLA-DRA, CTLA4, DUSP4, IL32, LGALS1, HLA-DQB1, IKZF2, HLA-DMA, HLA-DPA1 \n",
      "\t   LAIR2, LAP3, IL2RA, F5, TOP2A, LGALS9, TYMS, TIGIT, PTTG2, NINJ2 \n",
      "\t   CD79B, CD52, ANXA5, MKI67, CD74, TRIB2, MCM4, LGALS3, PASK, AURKB \n",
      "Negative:  IL7R, JUN, BTG1, DUSP1, CXCR4, FOSB, ARL4C, GZMK, TCF7, CCL5 \n",
      "\t   KLRB1, CD48, XBP1, PIK3IP1, DPP4, DUSP2, RUNX3, CD69, GZMA, CTSW \n",
      "\t   STAT4, KLRG1, TRAT1, CD44, CD6, CD3E, GNLY, IFNGR1, FYN, BCL11B \n",
      "PC_ 3 \n",
      "Positive:  ITGA4, PASK, TCF7, GZMH, IFNG, NKG7, ADA, CHI3L2, CXCR5, TYMS \n",
      "\t   TOP2A, CCR7, GZMB, ZAP70, TXK, ZNF683, MCM4, CD200, CCL5, IL23R \n",
      "\t   MCM2, TBX21, UBE2C, CD7, CD6, CX3CR1, CD300A, KLRB1, BCL2A1, AURKB \n",
      "Negative:  FOXP3, LGALS3, DUSP4, CXCR4, HLA-DQB1, PRDM1, IKZF2, IL2RA, DUSP1, LGALS1 \n",
      "\t   RGS1, S100A10, BTG1, CD44, HLA-DPA1, JUN, PDIA6, F5, SELPLG, FOSB \n",
      "\t   IL2RB, ARL4C, ANXA5, TRAC, CCR10, RUNX3, CD63, HLA-DRA, NINJ2, FOXO1 \n",
      "PC_ 4 \n",
      "Positive:  UBE2C, TYMS, MKI67, MCM4, MCM2, FOSB, AURKB, HMMR, PCNA, CXCR4 \n",
      "\t   DUSP1, TOP2A, JUN, IL7R, TMEM97, PTTG2, EGR1, JUNB, DPP4, RNASE6 \n",
      "\t   PMCH, LGALS3, EGR3, MZB1, BTG1, PDIA4, RORC, S100A9, TNFSF13B, MYC \n",
      "Negative:  GZMH, LAIR2, CD74, NKG7, ITGB2, GZMB, CD27, LCK, CD3D, CST7 \n",
      "\t   LAT, GAPDH, BIN2, LGALS9, CD2, CTLA4, SELPLG, CD37, CCL4, CD52 \n",
      "\t   ITGAM, CD3E, HLA-DPA1, CD3G, SELL, GNAI2, TRIB2, TBX21, CD4, CD5 \n",
      "PC_ 5 \n",
      "Positive:  FOXP3, IKZF2, TIGIT, CTLA4, CD27, GZMK, TXK, FOSB, IL2RA, IL2RB \n",
      "\t   LAIR2, CXCR5, CST7, CXCR6, IL18R1, RGS1, NRP1, CBLB, FYN, CXCR4 \n",
      "\t   PIK3IP1, JUN, DUSP1, CCR7, EGR1, IL18RAP, CCL5, IGHM, PRF1, JUNB \n",
      "Negative:  LGALS1, S100A10, CD52, IFITM2, ANXA5, TYMS, CD4, GAPDH, CD63, ITGB2 \n",
      "\t   CD6, MCM4, TNFSF10, GNAI2, RPN2, UBE2C, PCNA, TGFB1, PTGDR2, MCM2 \n",
      "\t   CD3D, TOP2A, MKI67, IL32, TRAT1, RNASE6, TRAC, IKZF1, PDIA4, CD3E \n",
      "\n",
      "Warning message:\n",
      "“Removed 15774 rows containing missing values (geom_point).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li><li>10</li><li>11</li><li>12</li><li>13</li><li>14</li><li>15</li><li>16</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 4\n",
       "\\item 5\n",
       "\\item 6\n",
       "\\item 7\n",
       "\\item 8\n",
       "\\item 9\n",
       "\\item 10\n",
       "\\item 11\n",
       "\\item 12\n",
       "\\item 13\n",
       "\\item 14\n",
       "\\item 15\n",
       "\\item 16\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 3\n",
       "4. 4\n",
       "5. 5\n",
       "6. 6\n",
       "7. 7\n",
       "8. 8\n",
       "9. 9\n",
       "10. 10\n",
       "11. 11\n",
       "12. 12\n",
       "13. 13\n",
       "14. 14\n",
       "15. 15\n",
       "16. 16\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "00:29:22 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "00:29:22 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "00:29:23 Initializing from normalized Laplacian + noise\n",
      "\n",
      "00:29:23 Commencing optimization for 200 epochs, with 512498 positive edges\n",
      "\n",
      "00:29:30 Optimization finished\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9395\n",
      "Number of communities: 15\n",
      "Elapsed time: 17 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 4 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9275\n",
      "Number of communities: 17\n",
      "Elapsed time: 14 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 6 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9190\n",
      "Number of communities: 18\n",
      "Elapsed time: 14 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 7 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9044\n",
      "Number of communities: 21\n",
      "Elapsed time: 13 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 10 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8930\n",
      "Number of communities: 22\n",
      "Elapsed time: 13 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 11 final clusters.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.8823\n",
      "Number of communities: 24\n",
      "Elapsed time: 12 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 13 final clusters.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "load(\"COVID_merged_CD4.wnn.res0.5.rda\")\n",
    "\n",
    "###Remove contaminant cluster and repeat Seurat pipelines, define cluster resolution\n",
    "\n",
    "###Input layer \n",
    "sample.name <- \"COVID_merged_CD4_3rd\"\n",
    "dir.name <- \"Seurat_plots_3rd\"\n",
    "\n",
    "#Choose subcluster for analyse\n",
    "#Exclude cluster 16 (contaminant of myeoid cells)\n",
    "DefaultAssay(seu) <- \"RNA\"\n",
    "seu <- subset(seu, idents = c(0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11))\n",
    "\n",
    "#For defining clustering resolution\n",
    "resolutions = c(0.1, 0.15, 0.2, 0.3, 0.4, 0.5)\n",
    "resols.name = \"minute\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "\n",
    "##For Abseq, Normalize ~ PCA\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Normalizing and scaling data\n",
    "seu <- NormalizeData(seu, normalization.method = \"CLR\", margin = 2, assay = \"ADT\")\n",
    "#Exclude markers from Variable features that were used in CD4/CD8 separation and contami detection\n",
    "ADT_list <- rownames(seu[[\"ADT\"]])\n",
    "VariableFeatures(seu) <- ADT_list[-which(ADT_list %in% c(\"CD4\", \"CD8\", \"CD19\", \"IgD\", \"IgM\"))] \n",
    "seu = ScaleData(object = seu) \n",
    "#Perform PCA：\n",
    "seu = RunPCA(object = seu, reduction.name = 'apca')\n",
    "\n",
    "##For RNA\n",
    "DefaultAssay(seu) <- \"RNA\"\n",
    "#Normalizing and scaling data\n",
    "seu = NormalizeData(object = seu, scale.factor=1000000)\n",
    "VariableFeatures(seu) <- rownames(seu[[\"RNA\"]]) #Use all genes in target panel for Variable features\n",
    "ngenes <- length(x = seu@assays$RNA@var.features)\n",
    "all.genes <- rownames(seu)\n",
    "seu = ScaleData(object = seu, vars.to.regress = c(\"nCount_RNA\"), features = all.genes) \n",
    "#Perform PCA\n",
    "seu = RunPCA(object = seu, features = seu@assays$RNA@var.features, npcs = 100)\n",
    "#JackStraw\n",
    "seu = JackStraw(object = seu, num.replicate = 100, dims = 50)\n",
    "seu <- ScoreJackStraw(object = seu, dims = 1:50, score.thresh = 0.05)\n",
    "file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Jackstraw.png\", sep='.')\n",
    "png(file.name, width = 1250, height = 500)\n",
    "JackStrawPlot(object = seu, dims = 1:50)\n",
    "dev.off()\n",
    "#Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "#Extract PCs which fulfill the pvalue threshold\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "dims\n",
    "\n",
    "#Clustering Weighted Nearest Neighbor\n",
    "seu <- FindMultiModalNeighbors(\n",
    "  seu, reduction.list = list(\"pca\", \"apca\"), \n",
    "  dims.list = list(dims, dims))\n",
    "#Dimentional reduction\n",
    "seu <- RunUMAP(seu, nn.name = \"weighted.nn\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "#Calculate Shilouhette score\n",
    "distance_matrix = parDist(seu@reductions$wnn.umap@cell.embeddings, method = \"euclidean\", threads=4)\n",
    "silhouette_score = NULL\n",
    "silhouette_score_mean = list()\n",
    "\n",
    "for (i in c(1:length(resolutions))){\n",
    "  #Perform clustering\n",
    "  seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resolutions[i])\n",
    "  silhouette = silhouette(as.numeric(seu@active.ident), dist = distance_matrix)\n",
    "  silhouette = silhouette[,3]\n",
    "  silhouette_score = c(silhouette_score, mean(silhouette))\n",
    "  names(silhouette_score)[i]=resolutions[i]\n",
    "  res <- vector()\n",
    "  for (j in c(0:length(unique(seu@active.ident))-1)){\n",
    "    tmp = mean(silhouette[seu@active.ident %in% j]) \n",
    "    res = c(res, tmp)\n",
    "  }\n",
    "  silhouette_score_mean[[i]]=res\n",
    "  names(silhouette)=rownames(seu@meta.data)\n",
    "  silhouette.name = sprintf(\"silhouette_score.res.%s\", resolutions[i])\n",
    "  seu =  AddMetaData(object = seu, metadata = silhouette, col.name = silhouette.name)\n",
    "}\n",
    "\n",
    "max_silhouette = names(silhouette_score[max(silhouette_score)])\n",
    "\n",
    "x = data.frame(resolution = resolutions,\n",
    "               mean_silhouette_score = silhouette_score)\n",
    "\n",
    "p_silhouette = ggplot(x, aes(x = resolution, y = mean_silhouette_score)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  ggtitle(\"Mean silhouette scores of clustering\") +\n",
    "  theme_linedraw() + \n",
    "  theme(plot.title=element_text(hjust = 0.5), text=element_text(size=12)) + \n",
    "  theme(axis.text.x=element_text(size=12), axis.text.y=element_text(size=12)) \n",
    "file.name=paste(sample.name, resols.name, \"_silhouette.png\", sep='')\n",
    "file.name=paste(dir.name, file.name, sep='/')\n",
    "ggsave(file = file.name, plot = p_silhouette, device=\"png\", units=\"in\", dpi = 300,\n",
    "       width = 4, height = 3, limitsize=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "5a6c372d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 16640\n",
      "Number of edges: 454953\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9395\n",
      "Number of communities: 15\n",
      "Elapsed time: 17 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11 singletons identified. 4 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    }
   ],
   "source": [
    "#Clustering using RNA and ADTinformation \n",
    "\n",
    "##Input layer\n",
    "dir.name=(\"Seurat_plots_3rd/\")\n",
    "sample.name <- \"COVID_merged_CD4_3rd\"\n",
    "resol <- 0.1 # resolution for clustering\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "#Clustering Weighted Nearest Neighbor\n",
    "seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resol)\n",
    "\n",
    "#Visualization using UMAP\n",
    "p1 = DimPlot(object = seu, reduction = \"wnn.umap\", label = TRUE, label.size = 10, pt.size = 0.5) +\n",
    "  theme(axis.title.x = element_text(size=10, family = \"Arial\"), \n",
    "        axis.title.y = element_text(size=10, family = \"Arial\"), \n",
    "        axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"), \n",
    "        axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "p2 = DimPlot(object = seu, reduction = \"wnn.umap\", label = FALSE, label.size = 10, pt.size = 0.5, group.by = \"orig.ident\") +\n",
    "  theme(axis.title.x = element_text(size=10, family = \"Arial\"),\n",
    "        axis.title.y = element_text(size=10, family = \"Arial\"),\n",
    "        axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"),\n",
    "        axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "  theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "\n",
    "legend1 <- cowplot::get_legend(p1)\n",
    "legend2 <- cowplot::get_legend(p2)\n",
    "p1 = p1 + theme(legend.position = 'none')\n",
    "p2 = p2 + theme(legend.position = 'none')\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"wnn.umap_reso\", resol, \".png\", sep='')\n",
    "save_plot(file = file.name, plot_grid(p1, legend1, p2, legend2, ncol=2, nrow=2), device=\"png\", \n",
    "          units=\"in\", dpi = 600, base_width = 10, base_height = 10, limitsize=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e1220bdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'Seurat_plots_3rd' already exists”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Extract marker genes for RNA and ADT\n",
    "\n",
    "###Input layer\n",
    "dir.name <- \"Seurat_plots_3rd\"\n",
    "sample.name <- \"COVID_merged_CD4_3rd\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "\n",
    "##For RNA\n",
    "DefaultAssay(seu) <- 'RNA'\n",
    "seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "#Create heatmap with top10 marker genes\n",
    "top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "top10 = as.data.frame(top10)\n",
    "top10  = top10 [!duplicated(top10$gene),]\n",
    "top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "top10 = as.data.frame(top10)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.marker_res\", resol, \".png\", sep='')\n",
    "p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "       width = 20, height = 20, limitsize=FALSE)\n",
    "#Output marker gene table\n",
    "seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "seu.markers = as.data.frame(seu.markers)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"RNA.ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "save(seu.markers, file=file.name)\n",
    "\n",
    "##For ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "#Create heatmap with top10 marker genes\n",
    "top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "top10 = as.data.frame(top10)\n",
    "top10  = top10 [!duplicated(top10$gene),]\n",
    "top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "top10 = as.data.frame(top10)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.marker_res\", resol, \".png\", sep='')\n",
    "p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "       width = 20, height = 20, limitsize=FALSE)\n",
    "#Output marker gene table\n",
    "seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "seu.markers = as.data.frame(seu.markers)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "file.name <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(\"ADT.ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "save(seu.markers, file=file.name)\n",
    "\n",
    "#Output Seurat object in initial analysis\n",
    "file.name=paste(sample.name, \".wnn.res\", resol, \".rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4e7cdcd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check marker genes for lineage (CD4 and CD8 dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "06e08399",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check marker genes for lineage using RNA expression\n",
    "sample.names <- c(\"COVID_merged_CD4_3rd.wnn.res0.1\", \"COVID_merged_CD8_3rd.wnn.res0.1\")\n",
    "tmp_rna = c(\"TRBC2\", \"TRDC\", \"CD14\", \"MS4A1\", \"KLRC1\", \"CHI3L1\")\n",
    "tmp_adt = c(\"CD3\", \"CD19\", \"CD16\", \"CD14\", \"CD56\", \"CD161\")\n",
    "dir.name=\"MarkerGene.Plot_3rd\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.create(dir.name)\n",
    "for(i in sample.names){\n",
    "    load(str_c(i, \"rda\", sep = \".\"))\n",
    "    \n",
    "    ##Plotting RNA expression\n",
    "    DefaultAssay(seu) <- \"RNA\"\n",
    "    #Scatter plot\n",
    "    tmp_name <- paste(i, \"Scatter.RNA.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- FeaturePlot(seu, features = tmp_rna, ncol = 6, order = TRUE,\n",
    "                     reduction = \"wnn.umap\", dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.2)\n",
    "    plot(p)\n",
    "    dev.off()    \n",
    "    #Violin plot\n",
    "    tmp_name <- paste(i, \"Violin.RNA.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- VlnPlot(seu, features = tmp_rna, ncol = 6, pt.size = 0.1)\n",
    "    plot(p)\n",
    "    dev.off()\n",
    "    \n",
    "    ##Plotting Protein expression\n",
    "    DefaultAssay(seu) <- \"ADT\"\n",
    "    #Scatter plot\n",
    "    tmp_name <- paste(i, \"Scatter.ADT.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- FeaturePlot(seu, features = tmp_adt, ncol = 6, order = TRUE,\n",
    "                     reduction = \"wnn.umap\", dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.2)\n",
    "    plot(p)\n",
    "    dev.off()    \n",
    "    #Violin plot\n",
    "    tmp_name <- paste(i, \"Violin.ADT.png\", sep = \".\")\n",
    "    file.name = paste(dir.name, tmp_name, sep='/')\n",
    "    png(file.name, width = 1536, height = 240)\n",
    "    p <- VlnPlot(seu, features = tmp_adt, ncol = 6, pt.size = 0.1)\n",
    "    plot(p)\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eae48d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Integrationa of scTCRseq results into SCT Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "23d75fb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "#Convert MiXCR output into VDJtools format\n",
    "Convert <- function(file.name, name.TRA){\n",
    "    name.input <- str_c(name.TRA, file.name, sep = \"/\")\n",
    "    data <- tableread_fast(name.input, header = TRUE, sep = '\\t')\n",
    "\n",
    "    #Extract TCR information from mixcr output\n",
    "    count.total<-data$cloneCount / data$cloneFraction\n",
    "    freq<-data$cloneFraction\n",
    "    cdr3nt<-data$nSeqImputedCDR3\n",
    "    cdr3aa<-data$aaSeqImputedCDR3\n",
    "    v<-str_sub(data$bestVHit, end=-4)\n",
    "    d<-str_sub(data$bestDHit, end=-4)\n",
    "    j<-str_sub(data$bestJHit, end=-4)\n",
    "    data3 <- rbind(count.total,freq,cdr3nt,cdr3aa,v,d,j)\n",
    "    data3 <- as.data.frame(t(data3))\n",
    "    names(data3) <- c(\"count.total\",\"freq\",\"cdr3nt\",\"cdr3aa\",\"v\",\"d\",\"j\")\n",
    "\n",
    "    #Extract largest clone in cell barcode and append cell barcode information\n",
    "    name_out <- str_split(file.name, \"_\")\n",
    "    CB <- name_out[[1]][[4]]\n",
    "    CB_out <- str_split(CB, \"\\\\.\")\n",
    "    CB <- CB_out[[1]][[1]]\n",
    "    d_out <- data3[1,]\n",
    "    d_out$CB <- CB\n",
    "\n",
    "    return(d_out)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "b214bfe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "#Create histogram was showing the total number of TCR reads detected in each cell (top left) \n",
    "#and the percentage of TCR reads that were the most common among them (bottom left).\n",
    "#Scatter plot representing these two metrices of cells were also generated\n",
    "Histogram <- function(d, file.name, sample.name, dir.output){\n",
    "\n",
    "  d_count_total <- as.numeric(d$count.total)\n",
    "  \n",
    "  #Count histogram\n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"histogram.count.tiff\", sep = '.') \n",
    "  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)\n",
    "  p <- ggplot(NULL, aes(x=d_count_total)) +\n",
    "    geom_histogram(binwidth=0.1) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    labs(x = \"Read counts of TCR\") +\n",
    "    theme(\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) +\n",
    "    scale_x_continuous(trans=scales::log2_trans(),\n",
    "                   breaks=scales::trans_breaks(\"log2\",function(x) 2^x),\n",
    "                   labels=scales::trans_format(\"log2\",scales::math_format(2^.x)))\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  #Frequency histogram  \n",
    "  d_freq <- as.numeric(d$freq)\n",
    "  med_f <- 0.3\n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"histogram.freq.tiff\", sep = '.')\n",
    "  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)\n",
    "  p <- ggplot(NULL, aes(x=d_freq)) +\n",
    "    geom_histogram(binwidth=0.05) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    labs(x = \"Proportion of the largest TCR read\") +\n",
    "    theme(\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) \n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  df <- data.frame(d_count_total, d_freq)\n",
    "  \n",
    "  #Scatter plot    \n",
    "  total <- nrow(subset(df, df$d_freq >= 0))\n",
    "  lb <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_total < 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  rb <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_total < 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  lt <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_total >= 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  rt <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_total >= 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  \n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"Scatter.tiff\", sep = '.')                     \n",
    "  tiff(image.file, width=1.2*ppi, height=1.2*ppi, res=ppi)\n",
    "  p <- ggplot(d, aes(x=d_freq, y=d_count_total)) +  \n",
    "    stat_bin2d(bins=60) +\n",
    "    scale_fill_gradient(low=\"lightblue\", high=\"red\") +\n",
    "    theme_bw(base_size = 6) +\n",
    "    geom_vline(aes(xintercept = 0.6), size=0.25, colour=\"black\") +\n",
    "    geom_hline(aes(yintercept = 2^5), size=0.25, colour=\"black\") +\n",
    "    labs(x = \"Proportion of the largest TCR read\", y = \"Read counts of TCR\") +\n",
    "    theme(\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) +\n",
    "    guides(fill=FALSE) +\n",
    "    scale_y_continuous(trans=scales::log2_trans(),\n",
    "                   breaks=scales::trans_breaks(\"log2\",function(x) 2^x),\n",
    "                   labels=scales::trans_format(\"log2\",scales::math_format(2^.x)))\n",
    "    p <- p + #xlim(0,1) +\n",
    "    annotate(\"text\", x=-Inf, y=0, hjust=-0.1, vjust=-0.4, label=lb,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=Inf, y=0, hjust=1.1, vjust=-0.4, label=rb,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=-Inf, y=Inf, hjust=-0.1, vjust=1.3, label=lt,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=Inf, y=Inf, hjust=1.1, vjust=1.3, label=rt,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) \n",
    "  print(p)\n",
    "  dev.off()  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "c27ed0bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.output):\n",
      "“'scTCR_processing' already exists”\n",
      "Warning message:\n",
      "“Removed 2538 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 2538 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 2538 rows containing non-finite values (stat_bin2d).”\n",
      "Warning message:\n",
      "“Removed 2646 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 2646 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 2646 rows containing non-finite values (stat_bin2d).”\n",
      "Warning message:\n",
      "“Removed 1893 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 1893 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 1893 rows containing non-finite values (stat_bin2d).”\n",
      "Warning message:\n",
      "“Removed 1454 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“Removed 1454 rows containing non-finite values (stat_bin).”\n",
      "Warning message:\n",
      "“`guides(<scale> = FALSE)` is deprecated. Please use `guides(<scale> = \"none\")` instead.”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Transformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“Removed 1454 rows containing non-finite values (stat_bin2d).”\n"
     ]
    }
   ],
   "source": [
    "#Extract TCR sequence for each cel barcode\n",
    "#iteration process for dataset (CD4/CD8 Seurat object) and clonotype (Ant/Bnt)\n",
    "set.seed(seed = 42)\n",
    "seurat.names <- c(\"COVID_merged_CD4_3rd.wnn.res0.1\", \"COVID_merged_CD8_3rd.wnn.res0.1\")\n",
    "sample.list <- c(\"ID02\", \"ID28\")\n",
    "dir.input.all <- \"scTCR_rawdata\"\n",
    "cores <- 12\n",
    "dir.output <- \"scTCR_processing\"\n",
    "#Threshold for valid cell barcode with single TCR sequence\n",
    "count_th <- 5 #read count threshold for all TCR sequences per cell barcode\n",
    "freq_th <- 0.6 #proportion threshold for largest TCR\n",
    "clonotype <- \"nt\" #Definition of clones. nt: nucleotide sequence / aa: amino acid sequnece \n",
    "hashtag <- FALSE #Whether hashtags are used or not.\n",
    "batch <- TRUE #Whether multiple chips are used or not.\n",
    "\n",
    "################################ Processing layer #################################################\n",
    "dir.create(dir.output)\n",
    "\n",
    "\n",
    "combined.tables <- data.frame()\n",
    "for(sample.name in sample.list){\n",
    "    #extract directory name\n",
    "    dir.input <- str_c(dir.input.all, sample.name, sep = \"/\")\n",
    "    \n",
    "    #Unzip files\n",
    "    files <- list.files(dir.input, \".tar.bz2\")\n",
    "    for(file in files){\n",
    "        name.input <- str_c(dir.input, file, sep = \"/\")\n",
    "        bunzip2(name.input, remove=FALSE)\n",
    "        name.input <- str_remove(name.input, pattern = \".bz2\")\n",
    "        untar(name.input)\n",
    "        file.remove(name.input)\n",
    "    }\n",
    "    \n",
    "    #Convert MiXCR output into VDJtools format\n",
    "    name.TRA <- str_c(sample.name, \"TCR_TRAC1_mixcr\", sep = \"\")\n",
    "    files  <- list.files(name.TRA, pattern=\".txt\")\n",
    "    cl <- makeCluster(cores)\n",
    "    registerDoParallel(cl)\n",
    "    TRA.table <- invisible(foreach(file.name = files,\n",
    "            .combine = rbind, .packages=c(\"ggplot2\", \"extrafont\", \"stringr\", \"dplyr\", \"data.table\")) %dopar% {Convert(file.name, name.TRA)})\n",
    "    stopCluster(cl)\n",
    "    unlink(name.TRA, recursive=TRUE)\n",
    "\n",
    "    name.TRB <- str_c(sample.name, \"TCR_TRBC1_mixcr\", sep = \"\")\n",
    "    files  <- list.files(name.TRB, pattern=\".txt\")\n",
    "    cl <- makeCluster(cores)\n",
    "    registerDoParallel(cl)\n",
    "    TRB.table <- invisible(foreach(file.name = files,\n",
    "            .combine = rbind, .packages=c(\"ggplot2\", \"extrafont\", \"stringr\", \"dplyr\", \"data.table\")) %dopar% {Convert(file.name, name.TRB)})\n",
    "    stopCluster(cl)\n",
    "    unlink(name.TRB, recursive=TRUE)\n",
    "\n",
    "    #Output histogram and scatter plot for summarizing scTCR status\n",
    "    Histogram(TRA.table, \"TRA\", sample.name, dir.output)\n",
    "    Histogram(TRB.table, \"TRB\", sample.name, dir.output)\n",
    "\n",
    "    #thresholding\n",
    "    TRA.table$count.total <- as.numeric(TRA.table$count.total)\n",
    "    TRA.table$freq <- as.numeric(TRA.table$freq)\n",
    "    TRB.table$count.total <- as.numeric(TRB.table$count.total)\n",
    "    TRB.table$freq <- as.numeric(TRB.table$freq)\n",
    "    TCRa_th <- subset(TRA.table, count.total >= 2^(count_th) & freq >= freq_th)\n",
    "    TCRb_th <- subset(TRB.table, count.total >= 2^(count_th) & freq >= freq_th)\n",
    "\n",
    "    #Paring TCRa and TCRb by cell barcode\n",
    "    names(TCRa_th) <- c(\"count.total.A\",\"freq.A\",\"cdr3nt.A\",\"cdr3aa.A\",\"v.A\",\"d.A\",\"j.A\", \"CB\")\n",
    "    names(TCRb_th) <- c(\"count.total.B\",\"freq.B\",\"cdr3nt.B\",\"cdr3aa.B\",\"v.B\",\"d.B\",\"j.B\", \"CB\")\n",
    "    combined <- merge(TCRa_th, TCRb_th, all=T, by =\"CB\")\n",
    "    combined$CB <- str_c(sample.name, combined$CB, sep = \"_\")\n",
    "    #Exclude cells in which neither TCRa nor TCRb sequence were detected\n",
    "    combined <- subset(combined, combined$cdr3nt.A != \"UD\" | combined$cdr3nt.B != \"UD\")\n",
    "\n",
    "    #Generate clone id for further analysis\n",
    "    #Definition of clones can be changed between \"ABnt\", \"ABaa\", \"Bnt\", and \"Baa\"\n",
    "    if(clonotype==\"nt\"){\n",
    "        combined$clone.id.TCRa <- str_c(combined$cdr3nt.A, combined$v.A, combined$j.A, sep=\"_\")\n",
    "        combined$clone.id.TCRb <- str_c(combined$cdr3nt.B, combined$v.B, combined$j.B, sep=\"_\")\n",
    "    }\n",
    "    if(clonotype==\"aa\"){\n",
    "        combined$clone.id.TCRa <- str_c(combined$cdr3aa.A, combined$v.A, combined$j.A, sep=\"_\")\n",
    "        combined$clone.id.TCRb <- str_c(combined$cdr3aa.B, combined$v.B, combined$j.B, sep=\"_\")\n",
    "    }\n",
    "    combined$clone.id.TCRab <- str_c(combined$clone.id.TCRa, combined$clone.id.TCRb, sep = \"_\")\n",
    "    name.output <- str_c(dir.output, sample.name, sep = \"/\") %>% str_c(clonotype, \"table\", \"count_th\", count_th, \"freq_th\", freq_th, clonotype, \"csv\", sep = \".\")\n",
    "    write.csv(combined, name.output, row.names = FALSE)\n",
    "    \n",
    "    combined.tables <- rbind(combined.tables, combined)\n",
    "}\n",
    "\n",
    "##Integrate to the Seurat object; iteration for dataset (CD4/CD8 Seurat object))\n",
    "for(seurat.name in seurat.names){\n",
    "    #load Seurat object\n",
    "    name.input <- str_c(seurat.name, \"rda\", sep = \".\")\n",
    "    load(name.input)\n",
    "    \n",
    "    #Extract meta.data and cell BC information\n",
    "    meta.data <- seu@meta.data\n",
    "    meta.data$names <- row.names(meta.data)\n",
    "    if(hashtag == TRUE){\n",
    "        if(batch == FALSE){\n",
    "            CB.info <- str_split(meta.data$names, pattern = \"_\", simplify = TRUE)\n",
    "            meta.data$CB <- CB.info[,ncol(CB.info)]\n",
    "        } else {\n",
    "            CB.info <- str_split(meta.data$names, pattern = \"_\", simplify = TRUE)\n",
    "            meta.data$CB <- str_c(CB.info[,1], CB.info[,3], sep = \"_\")\n",
    "        }\n",
    "    } else {\n",
    "            meta.data$CB <- meta.data$names\n",
    "    }\n",
    "\n",
    "    #Merge to Seurat object of SCT data\n",
    "    meta.data <- merge(combined.tables, meta.data, all.y = T, by =\"CB\")\n",
    "    clone.ids <- c(\"clone.id.TCRa\", \"clone.id.TCRb\", \"clone.id.TCRab\")\n",
    "    for(clone.id in clone.ids){\n",
    "        clone.id.list <- dplyr::select(meta.data, clone.id)\n",
    "        row.names(clone.id.list)=as.character(meta.data$names)\n",
    "        seu <- AddMetaData(object = seu, metadata = clone.id.list, col.name = clone.id)\n",
    "    }\n",
    "    \n",
    "    #Output Seurat object\n",
    "    name.output <- str_c(seurat.name, \"scTCRmerged\", clonotype, \"rda\", sep = \".\")\n",
    "    save(seu, file=name.output)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f42bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Integrating WGCNA and AIM status of clones into scTCRseq results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "39ebb124",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "\n",
    "##Make clone table by assembling clones from scTCR results \n",
    "CloneSummary <- function(seu, dir.name, sample.name, clonotype, cell){\n",
    "    #seu: Seurat object for analyze\n",
    "    #dir.name: directory name for output\n",
    "    dir.create(dir.name)\n",
    "\n",
    "    #load metadata\n",
    "    meta.data <- as.data.frame(seu@meta.data)\n",
    "    meta.data$CellBC <- row.names(meta.data)\n",
    "    #extract cells only in the assigned sample\n",
    "    meta.data <- dplyr::filter(meta.data, str_detect(CellBC, sample.name))\n",
    "    \n",
    "    #Define clone by TCRa / TCRb/ TCRa&b\n",
    "    if(clonotype == \"TCRa\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRa \n",
    "    }\n",
    "    if(clonotype == \"TCRb\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRb \n",
    "    }\n",
    "    if(clonotype == \"TCRab\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRab \n",
    "    } \n",
    "\n",
    "    #Summarize cluster distribution of each clone\n",
    "    tmp_out = table(meta.data$seurat_clusters, meta.data$clone.id)\n",
    "    tmp_out2 <- as.data.frame(tmp_out, row.names = NULL,\n",
    "                  responseName = \"Freq\", stringsAsFactors = TRUE,\n",
    "                  sep = \"\", base = list(LETTERS))\n",
    "    names(tmp_out2) <- c(\"Clust\", \"names\", \"Freq\")\n",
    "    tmp_out2 <- dcast(tmp_out2, names ~ Clust)\n",
    "\n",
    "    #Summarize the count and frequency of clones \n",
    "    tmp = meta.data %>% group_by(`clone.id`) %>%\n",
    "      dplyr::summarise(count = n()) %>%\n",
    "      dplyr::arrange(desc(count))\n",
    "    tmp = as.data.frame(tmp)\n",
    "    tmp <- tmp[!is.na(tmp$clone.id), ] #NAである細胞を除く\n",
    "    tmp[,3]=tmp[,2]/sum(tmp[,2])\n",
    "    colnames(tmp)=c(\"ntSeq_TRA_TRB_freq\", \"CloneCount\", \"CloneFreq\")\n",
    "    tmp <- tmp[order(tmp$ntSeq_TRA_TRB_freq),]\n",
    "    temp_count <- as.vector(tmp$CloneCount)\n",
    "    temp_freq <- as.vector(tmp$CloneFreq)\n",
    "    #Combine to the cluster distribution\n",
    "    tmp_out3 <- cbind(temp_count, temp_freq, tmp_out2)\n",
    "\n",
    "    #Assign ranks to clones\n",
    "    tmp_out3 <- tmp_out3[order(tmp_out3$temp_freq, decreasing=T),]\n",
    "    rank <-  1:nrow(tmp_out3)\n",
    "    tmp_out3 <- cbind(rank, tmp_out3)\n",
    "    tmp_out3 <- data.frame(tmp_out3)\n",
    "    tmp_out3$rank <- paste(\"Top\",tmp_out3$rank, sep=\"\")\n",
    "    \n",
    "    #Output\n",
    "    name.out <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(cell, clonotype, \"clone_within_cluster.txt\", sep = \".\")\n",
    "    write.table(tmp_out3, name.out, row.names=F, col.names=T, sep=\"\\t\", quote=F) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "888be63b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'scTCR_analysis' already exists”\n",
      "Warning message in dir.create(dir.name):\n",
      "“'scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Make clone table by assembling clones from scTCR results \n",
    "\n",
    "dir.name <- \"scTCR_analysis\"\n",
    "sample.list <- c(\"ID02\", \"ID28\")\n",
    "seurat.names <- c(\"COVID_merged_CD4_3rd.wnn.res0.1.scTCRmerged.nt\", \"COVID_merged_CD8_3rd.wnn.res0.1.scTCRmerged.nt\")\n",
    "\n",
    "##################### Processing layer #############################################################\n",
    "dir.create(dir.name)\n",
    "for(seurat.name in seurat.names){\n",
    "    #load Seurat object\n",
    "    name.input <- str_c(seurat.name, \"rda\", sep = \".\")\n",
    "    load(name.input)\n",
    "    cell <- str_split(seurat.name, pattern = \"_\", simplify = TRUE)[3]\n",
    "    \n",
    "    for(sample.name in sample.list){\n",
    "    #Make clone table by assembling clones from scTCR results \n",
    "    CloneSummary(seu, dir.name, sample.name, \"TCRb\", cell)\n",
    "    } \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "b3acadba",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Merge SCT metadata with WGCNA pattern / AIM positivity\n",
    "\n",
    "#Input layer\n",
    "dir.input.scTCR <- \"scTCR_analysis\"\n",
    "dir.input.wgcna <- \"WGCNA_AIMcombined\"\n",
    "dir.output <- \"scTCR_analysis\"\n",
    "#Information to merge into scTCR clone table\n",
    "extract <- c(\"P1\", \"P2\", \"P3\", \"P4\", \"class.x\", \"class.y\", \"cell\")\n",
    "\n",
    "Tcells <- c(\"CD4\", \"CD8\")\n",
    "sample.sct.names <- c(\"ID02\", \"ID28\")\n",
    "sample.wgcna.names <- c(\"002\", \"027\")\n",
    "\n",
    "##################### Processing layer #############################################################\n",
    "for(Tcell in Tcells){\n",
    "    for(i in 1:length(sample.wgcna.names)){\n",
    "    #load data\n",
    "    sample.sct <- str_c(dir.input.scTCR, sample.sct.names[i], sep = \"/\") %>% str_c(Tcell, \"TCRb.clone_within_cluster.txt\", sep = \".\")\n",
    "    sct <- read.table(sample.sct, header = TRUE)\n",
    "    sample.wgcna <- str_c(dir.input.wgcna, \"COVIDAIM\", sep = \"/\") %>%\n",
    "        str_c(Tcell, sample.wgcna.names[i], \"TCR.th.4.wgcna.aim.csv\", sep = \"_\") \n",
    "    wgcna <- read.csv(sample.wgcna, header = TRUE)\n",
    "  \n",
    "  #Merge wgcna and sct dataset\n",
    "  d_output <- merge(sct, wgcna, by.x = \"names\", by.y = \"query\", all.x = T)\n",
    "  \n",
    "  #Output\n",
    "  name.out <- str_c(dir.output, sample.wgcna.names[i], sep = \"/\") %>% str_c(Tcell, \"clone_within_cluster.WGCNA.AIM.csv\", sep = \".\")\n",
    "  write.csv(d_output, name.out, row.names = FALSE)\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd3582cc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65aa7f27",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "619f53fa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
