{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "25a3eb7b",
   "metadata": {},
   "source": [
    "### Load libraries and functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9a52127",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,\n",
      "which was just loaded, will retire in October 2023.\n",
      "Please refer to R-spatial evolution reports for details, especially\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html.\n",
      "It may be desirable to make the sf package available;\n",
      "package maintainers should consider adding sf to Suggests:.\n",
      "The sp package is now running under evolution status 2\n",
      "     (status 2 uses the sf package in place of rgdal)\n",
      "\n",
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "Loading required package: iterators\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘proxy’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    as.dist, dist\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    as.matrix\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.11.1\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.14\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:circlize’:\n",
      "\n",
      "    degree\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:igraph’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: splines\n",
      "\n",
      "\n",
      "Attaching package: ‘pheatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    pheatmap\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "Loading required package: R.oo\n",
      "\n",
      "Loading required package: R.methodsS3\n",
      "\n",
      "R.methodsS3 v1.8.1 (2020-08-26 16:20:06 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "R.oo v1.24.0 (2020-08-26 16:11:58 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    hierarchy\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "R.utils v2.11.0 (2021-09-26 08:30:02 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    draw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, inherits, isOpen, nullfile, parse,\n",
      "    warnings\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘scales’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:viridis’:\n",
      "\n",
      "    viridis_pal\n",
      "\n",
      "\n",
      "Loading required package: tidyr\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.utils’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:VGAM’:\n",
      "\n",
      "    fill\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:igraph’:\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Load libraries\n",
    "library(Seurat)\n",
    "library(stringr)\n",
    "library(viridis)\n",
    "library(ggplot2)\n",
    "library(cowplot)\n",
    "library(cluster)\n",
    "library(data.table)\n",
    "library(foreach)\n",
    "library(doParallel)\n",
    "library(proxy)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(igraph)\n",
    "library(qvalue)\n",
    "library(dplyr)\n",
    "library(viridis)\n",
    "library(VGAM)\n",
    "library(forcats)\n",
    "library(grDevices)\n",
    "library(graphics)\n",
    "library(RColorBrewer)\n",
    "library(pheatmap)\n",
    "library(Cairo)\n",
    "library(reshape2)\n",
    "library(R.utils)\n",
    "library(Rcpp)\n",
    "library(parallelDist)\n",
    "library(ggsci)\n",
    "library(scales)\n",
    "set.seed(seed = 42)\n",
    "library(ADTnorm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54558b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load basic functions\n",
    "createEmptyDf = function( nrow, ncol, colnames = c() ){\n",
    "  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )\n",
    "}\n",
    "\n",
    "tableread_fast = function(i, header=TRUE, quote=\"\", sep=\",\"){\n",
    "  tmp = fread(i, header=header, sep=sep, quote=quote, nThread=32)\n",
    "  tmp = as.data.frame(tmp)\n",
    "  return(tmp)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7221f3f",
   "metadata": {},
   "source": [
    "### Create Seurat object and merge datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cb7bc92a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seu.list <- list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3c46710a",
   "metadata": {},
   "outputs": [],
   "source": [
    "##load data of First batch: TP3 ID2, ID27\n",
    "#Put gene list in the target panel in the working directory (\"BD_genes_human\")\n",
    "\n",
    "###Input layer\n",
    "set.seed(seed = 42)\n",
    "\n",
    "dir.input <- \"data/original/SCT\"\n",
    "sample.names <- c(\"ID2\", \"ID27\")\n",
    "\n",
    "################################Processing layer##########################################\n",
    "for(sample.name in sample.names){\n",
    "    ##Load targeted gene expression data\n",
    "    input_name <- str_c(dir.input, \"matrix_inflection_\", sep = \"/\") %>% str_c(sample.name, \"Target.txt\")\n",
    "    compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "    gunzip(compressed.name)\n",
    "    #Create Seurat object\n",
    "    matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "    row.names(matrix) <- matrix$V1\n",
    "    matrix <- dplyr::select(matrix, -V1)\n",
    "    seu1 <- CreateSeuratObject(counts=matrix, project = \"seu\", min.cells = 3, min.features = 10)\n",
    "    BC <- row.names(seu1@meta.data)\n",
    "    gzip(input_name)\n",
    "\n",
    "    ##Load protein expression data\n",
    "    input_name <- str_replace(input_name,\"Target\", \"Abseq\") %>% str_replace(\"matrix_inflection\", \"Hashtag_top1M\")\n",
    "    compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "    gunzip(compressed.name)\n",
    "    #Extract cells in SCT seurat object\n",
    "    Abseq.matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "    Abseq.matrix <- dplyr::filter(Abseq.matrix, CellBC %in% BC)\n",
    "    row.names(Abseq.matrix) <- Abseq.matrix$CellBC\n",
    "    Abseq.matrix <- dplyr::select(Abseq.matrix, -CellBC)\n",
    "    Abseq.matrix <- t(Abseq.matrix)\n",
    "    gzip(input_name)\n",
    "\n",
    "    #Certify\n",
    "    all.equal(rownames(seu1@meta.data), colnames(Abseq.matrix))\n",
    "\n",
    "    ##Combine ADT data into SCT seurat object\n",
    "    adt_assay <- CreateAssayObject(counts = Abseq.matrix)\n",
    "    seu1[[\"ADT\"]] <- adt_assay\n",
    "    Assays(seu1)\n",
    "\n",
    "    seu1@meta.data$orig.ident <- str_c(\"TP3\", sample.name, sep = \"_\")\n",
    "    seu.list <- c(seu.list, list(seu1))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a0c9183d",
   "metadata": {},
   "outputs": [],
   "source": [
    "##load data of Second batch: TP3/TP8, ID11, 20, 23, 33, 37\n",
    "#Put gene list in the target panel in the working directory (\"BD_genes_human\")\n",
    "\n",
    "###Input layer\n",
    "set.seed(seed = 42)\n",
    "\n",
    "dir.input <- \"data/original/SCT\"\n",
    "sample.names <- c(\"TP38Target1\", \"TP38Target2\")\n",
    "\n",
    "################################Processing layer##########################################\n",
    "for(sample.name in sample.names){\n",
    "    input_name <- str_c(dir.input, \"matrix_inflection_demulti_\", sep = \"/\") %>% str_c(sample.name, \".txt\")\n",
    "    compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "    gunzip(compressed.name)\n",
    "    #Create Seurat object\n",
    "    matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "    row.names(matrix) <- matrix$V1\n",
    "    matrix <- dplyr::select(matrix, -V1)\n",
    "    seu1 <- CreateSeuratObject(counts=matrix, project = \"seu\", min.cells = 3, min.features = 10)\n",
    "    BC <- str_replace(row.names(seu1@meta.data), \"not_detected\", \"notdetected\") %>% str_split(pattern = \"_\", simplify = TRUE)\n",
    "    BC <- BC[,3]\n",
    "    seu1@assays$RNA@data@Dimnames[[2]] <- BC  \n",
    "    row.names(seu1@meta.data) <- BC  \n",
    "    gzip(input_name)\n",
    "\n",
    "    ##Load protein expression data\n",
    "    input_name <- str_replace(input_name,\"Target\", \"Abseq\") %>% str_replace(\"matrix_inflection_demulti\", \"Hashtag_top1M\")\n",
    "    compressed.name <- str_c(input_name, \".gz\", sep = \"\")\n",
    "    gunzip(compressed.name)\n",
    "    #Extract cells in SCT seurat object\n",
    "    Abseq.matrix <- tableread_fast(input_name, header = TRUE, quote=\"\", sep=\"\\t\")\n",
    "    Abseq.matrix <- dplyr::filter(Abseq.matrix, CellBC %in% BC)\n",
    "    row.names(Abseq.matrix) <- Abseq.matrix$CellBC\n",
    "    Abseq.matrix <- dplyr::select(Abseq.matrix, -CellBC)\n",
    "    Abseq.matrix <- t(Abseq.matrix)\n",
    "    gzip(input_name)\n",
    "    \n",
    "    #Certify\n",
    "    all.equal(seu1@assays$RNA@data@Dimnames[[2]], colnames(Abseq.matrix))\n",
    "\n",
    "    ##Combine ADT data into SCT seurat object\n",
    "    adt_assay <- CreateAssayObject(counts = Abseq.matrix)\n",
    "    seu1[[\"ADT\"]] <- adt_assay\n",
    "    Assays(seu1)\n",
    "\n",
    "    #transfer Sampletag to Sample ID\n",
    "    seu1@meta.data$orig.ident <- str_replace_all(seu1@meta.data$orig.ident, pattern = c(\"humanSampleTag10\" = \"TP8_ID35\"))\n",
    "    seu1@meta.data$orig.ident <- str_replace_all(seu1@meta.data$orig.ident, pattern = c(\"humanSampleTag1\" = \"TP3_ID10\",\n",
    "                                                                                       \"humanSampleTag2\" = \"TP3_ID19\",\n",
    "                                                                                       \"humanSampleTag3\" = \"TP3_ID22\",\n",
    "                                                                                       \"humanSampleTag4\" = \"TP3_ID31\",\n",
    "                                                                                       \"humanSampleTag5\" = \"TP3_ID35\",\n",
    "                                                                                       \"humanSampleTag6\" = \"TP8_ID10\",\n",
    "                                                                                       \"humanSampleTag7\" = \"TP8_ID19\",\n",
    "                                                                                       \"humanSampleTag8\" = \"TP8_ID22\",\n",
    "                                                                                       \"humanSampleTag9\" = \"TP8_ID31\"))\n",
    "    seu.list <- c(seu.list, list(seu1))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f68114b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.SCT, recursive = TRUE):\n",
      "“'result/intermediate/3_SCT' already exists”\n"
     ]
    }
   ],
   "source": [
    "##Merge data\n",
    "\n",
    "#Input layer\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "sample.name <- \"COVID_merged\"\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "dir.create(dir.SCT, recursive = TRUE)\n",
    "seu <- merge(x=seu.list[[1]], y = c(seu.list[[2]], seu.list[[3]], seu.list[[4]]),\n",
    "             add.cell.ids = c(\"Batch1-ID2\", \"Batch1-ID27\", \"Batch2-1\", \"Batch2-2\"))\n",
    "\n",
    "#Append batch name and cell barcodes\n",
    "meta <- str_split(row.names(seu@meta.data), pattern = \"_\", simplify = TRUE)\n",
    "seu@meta.data$batch <- meta[,1]\n",
    "seu@meta.data$BC <- meta[,2]\n",
    "\n",
    "name.output <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"rda\", sep = \".\")\n",
    "save(seu, file = name.output)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fa17a51",
   "metadata": {},
   "source": [
    "# Correct Batch Effect / 1st Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2682bc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing across cells\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Exclude CD4+ cells and doublet\n",
    "\n",
    "##Input layer\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "dir.name=\"Seurat_plots\"\n",
    "sample.name <- \"COVID_CD8only\"\n",
    "gene.list.name <- \"metadata/BD_genes_human.txt\"\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "##load gene list of BD target panel\n",
    "BD_genes <- read.table(gene.list.name, header = TRUE)\n",
    "BD_genes <- as.vector(BD_genes$Genesymbol)\n",
    "\n",
    "##Chose BD target gene list for downstream analysis\n",
    "res1 = seu@assays$RNA@counts\n",
    "res1 =res1[res1@Dimnames[[1]] %in% BD_genes,]\n",
    "ngenes <- length(res1@Dimnames[[1]])\n",
    "seu@assays$RNA@counts <- res1\n",
    "seu@assays$RNA@data <- res1\n",
    "\n",
    "#Normalize Abseq data\n",
    "DefaultAssay(seu) <- \"ADT\"\n",
    "seu <- NormalizeData(seu, normalization.method = \"CLR\", margin = 2, assay = \"ADT\")\n",
    "# Variable feature: All ADT Features\n",
    "VariableFeatures(seu) <- rownames(seu[[\"ADT\"]])\n",
    "#Scaling data \n",
    "seu = ScaleData(object = seu)\n",
    "\n",
    "##Exclude CD4+ cells and doublet\n",
    "DefaultAssay(seu) <- \"ADT\"\n",
    "seu <- subset(seu, subset = CD4 < 3 & CD8 > 1.5)\n",
    "##Exclude doublets\n",
    "#str(seu)\n",
    "seu <- subset(seu, subset = orig.ident %in% c(\"TP3_ID2\", \"TP3_ID27\", \"TP8_ID19\", \"TP3_ID31\", \"TP8_ID31\",\n",
    "                                                  \"TP8_ID10\", \"TP3_ID19\", \"TP8_ID35\", \"TP8_ID22\", \"TP3_ID10\",\n",
    "                                                  \"TP3_ID35\", \"TP3_ID22\"))\n",
    "\n",
    "#Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"rda\", sep = \".\")\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "341842f0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/Seurat_plots' already exists”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ADA'</li><li>'ADGRE1'</li><li>'ADGRG3'</li><li>'AIM2'</li><li>'ALAS2'</li><li>'ANXA5'</li><li>'AOC3'</li><li>'APOBEC3G'</li><li>'APOE'</li><li>'AQP9'</li><li>'ARG1'</li><li>'ARL4C'</li><li>'ATF6B'</li><li>'AURKB'</li><li>'AZU1'</li><li>'B3GAT1'</li><li>'BACH2'</li><li>'BAX'</li><li>'BCL11B'</li><li>'BCL2'</li><li>'BCL2A1'</li><li>'BCL6'</li><li>'BIN2'</li><li>'BIRC3'</li><li>'BLK'</li><li>'BLNK'</li><li>'BPI'</li><li>'BTG1'</li><li>'BTLA'</li><li>'C1QA'</li><li>'CASP5'</li><li>'CBLB'</li><li>'CCL1'</li><li>'CCL17'</li><li>'CCL19'</li><li>'CCL20'</li><li>'CCL22'</li><li>'CCL3'</li><li>'CCL4'</li><li>'CCL5'</li><li>'CCND2'</li><li>'CCR1'</li><li>'CCR10'</li><li>'CCR2'</li><li>'CCR3'</li><li>'CCR4'</li><li>'CCR5'</li><li>'CCR7'</li><li>'CCR8'</li><li>'CCR9'</li><li>'CD14'</li><li>'CD160'</li><li>'CD163'</li><li>'CD1B'</li><li>'CD1C'</li><li>'CD2'</li><li>'CD200'</li><li>'CD209'</li><li>'CD22'</li><li>'CD24'</li><li>'CD244'</li><li>'CD247'</li><li>'CD27'</li><li>'CD274'</li><li>'CD28'</li><li>'CD300A'</li><li>'CD33'</li><li>'CD34'</li><li>'CD36'</li><li>'CD37'</li><li>'CD38'</li><li>'CD3D'</li><li>'CD3E'</li><li>'CD3G'</li><li>'CD4'</li><li>'CD40'</li><li>'CD44'</li><li>'CD48'</li><li>'CD5'</li><li>'CD52'</li><li>'CD6'</li><li>'CD63'</li><li>'CD69'</li><li>'CD7'</li><li>'CD70'</li><li>'CD72'</li><li>'CD74'</li><li>'CD79A'</li><li>'CD79B'</li><li>'CD80'</li><li>'CD86'</li><li>'CD8A'</li><li>'CD8B'</li><li>'CD9'</li><li>'CEACAM8'</li><li>'CHI3L1'</li><li>'CHI3L2'</li><li>'CLC'</li><li>'CLEC4D'</li><li>'CLEC4E'</li><li>'CMKLR1'</li><li>'CMTM2'</li><li>'CNOT2'</li><li>'CNTNAP3'</li><li>'CR2'</li><li>'CSF2'</li><li>'CSF3'</li><li>'CST7'</li><li>'CTLA4'</li><li>'CTSD'</li><li>'CTSG'</li><li>'CTSW'</li><li>'CX3CR1'</li><li>'CXCL10'</li><li>'CXCL13'</li><li>'CXCL16'</li><li>'CXCL2'</li><li>'CXCL3'</li><li>'CXCL5'</li><li>'CXCL8'</li><li>'CXCR1'</li><li>'CXCR2'</li><li>'CXCR3'</li><li>'CXCR4'</li><li>'CXCR5'</li><li>'CXCR6'</li><li>'DOCK8'</li><li>'DPP4'</li><li>'DUSP1'</li><li>'DUSP2'</li><li>'DUSP4'</li><li>'EBF1'</li><li>'EGR1'</li><li>'EGR3'</li><li>'ELANE'</li><li>'ENTPD1'</li><li>'EOMES'</li><li>'F13A1'</li><li>'F5'</li><li>'FAS'</li><li>'FASLG'</li><li>'FCER1G'</li><li>'FCER2'</li><li>'FCGR3A'</li><li>'FCN1'</li><li>'FLT3'</li><li>'FN1'</li><li>'FOSB'</li><li>'FOSL1'</li><li>'FOXO1'</li><li>'FOXP1'</li><li>'FOXP3'</li><li>'FTH1'</li><li>'FUT4'</li><li>'FYN'</li><li>'GAB2'</li><li>'GAPDH'</li><li>'GIMAP2'</li><li>'GIMAP5'</li><li>'GNAI2'</li><li>'GNLY'</li><li>'GZMA'</li><li>'GZMB'</li><li>'GZMH'</li><li>'GZMK'</li><li>'HAVCR2'</li><li>'HLA-A'</li><li>'HLA-DMA'</li><li>'HLA-DPA1'</li><li>'HLA-DQB1'</li><li>'HLA-DRA'</li><li>'HMMR'</li><li>'ICAM1'</li><li>'ICOS'</li><li>'IER3'</li><li>'IFITM2'</li><li>'IFITM3'</li><li>'IFNG'</li><li>'IFNGR1'</li><li>'IGBP1'</li><li>'IGHD'</li><li>'IGHE'</li><li>'IGHG1'</li><li>'IGHG2'</li><li>'IGHG4'</li><li>'IGHM'</li><li>'IGKC'</li><li>'IKZF1'</li><li>'IKZF2'</li><li>'IL12A'</li><li>'IL12RB1'</li><li>'IL12RB2'</li><li>'IL13'</li><li>'IL15'</li><li>'IL15RA'</li><li>'IL18'</li><li>'IL18R1'</li><li>'IL18RAP'</li><li>'IL1B'</li><li>'IL1R2'</li><li>'IL1RL1'</li><li>'IL1RN'</li><li>'IL2'</li><li>'IL21'</li><li>'IL22'</li><li>'IL23R'</li><li>'IL2RA'</li><li>'IL2RB'</li><li>'IL32'</li><li>'IL33'</li><li>'IL3RA'</li><li>'IL4'</li><li>'IL4R'</li><li>'IL5'</li><li>'IL6'</li><li>'IL7R'</li><li>'IL9'</li><li>'IRF4'</li><li>'IRF8'</li><li>'ITGA4'</li><li>'ITGAE'</li><li>'ITGAM'</li><li>'ITGAX'</li><li>'ITGB2'</li><li>'JCHAIN'</li><li>'JUN'</li><li>'JUNB'</li><li>'KCNE3'</li><li>'KDELR1'</li><li>'KIR2DL1'</li><li>'KIT'</li><li>'KLRB1'</li><li>'KLRC1'</li><li>'KLRC3'</li><li>'KLRC4'</li><li>'KLRF1'</li><li>'KLRG1'</li><li>'KLRK1'</li><li>'LAG3'</li><li>'LAIR2'</li><li>'LAMP1'</li><li>'LAMP3'</li><li>'LAP3'</li><li>'LAT'</li><li>'LAT2'</li><li>'LCK'</li><li>'LEF1'</li><li>'LGALS1'</li><li>'LGALS3'</li><li>'LGALS9'</li><li>'LIF'</li><li>'LILRB4'</li><li>'LIPA'</li><li>'LRRC32'</li><li>'LTA'</li><li>'LTB'</li><li>'LY86'</li><li>'LYN'</li><li>'MCM2'</li><li>'MCM4'</li><li>'MGST1'</li><li>'MITF'</li><li>'MKI67'</li><li>'MME'</li><li>'MMP12'</li><li>'MMP9'</li><li>'MS4A1'</li><li>'MYC'</li><li>'MZB1'</li><li>'NAMPT'</li><li>'NCAM1'</li><li>'NCR3'</li><li>'NINJ2'</li><li>'NKG7'</li><li>'NRP1'</li><li>'NT5E'</li><li>'PASK'</li><li>'PCNA'</li><li>'PDCD1'</li><li>'PDIA4'</li><li>'PDIA6'</li><li>'PIK3AP1'</li><li>'PIK3IP1'</li><li>'PMCH'</li><li>'POU2AF1'</li><li>'PRDM1'</li><li>'PRF1'</li><li>'PSEN1'</li><li>'PTGDR2'</li><li>'PTPRC'</li><li>'PTTG2'</li><li>'QPCT'</li><li>'RGS1'</li><li>'RNASE2'</li><li>'RNASE6'</li><li>'RORA'</li><li>'RORC'</li><li>'RPN2'</li><li>'RUNX3'</li><li>'S100A10'</li><li>'S100A12'</li><li>'S100A9'</li><li>'SELL'</li><li>'SELPLG'</li><li>'SLC25A37'</li><li>'SLC7A7'</li><li>'SNCA'</li><li>'STAT1'</li><li>'STAT3'</li><li>'STAT4'</li><li>'STAT5A'</li><li>'STAT6'</li><li>'TBX21'</li><li>'TCF4'</li><li>'TCF7'</li><li>'TGFB1'</li><li>'TGFB3'</li><li>'TGFBI'</li><li>'THBD'</li><li>'THBS1'</li><li>'TIAF1'</li><li>'TIGIT'</li><li>'TLR2'</li><li>'TLR7'</li><li>'TLR8'</li><li>'TLR9'</li><li>'TMEM97'</li><li>'TNF'</li><li>'TNFRSF13C'</li><li>'TNFRSF25'</li><li>'TNFRSF4'</li><li>'TNFRSF8'</li><li>'TNFRSF9'</li><li>'TNFSF10'</li><li>'TNFSF13'</li><li>'TNFSF13B'</li><li>'TNFSF14'</li><li>'TNFSF8'</li><li>'TOP2A'</li><li>'TPSAB1'</li><li>'TRAC'</li><li>'TRAT1'</li><li>'TRBC2'</li><li>'TRDC'</li><li>'TREM1'</li><li>'TRIB2'</li><li>'TSPAN32'</li><li>'TXK'</li><li>'TYMS'</li><li>'UBE2C'</li><li>'VEGFA'</li><li>'VMO1'</li><li>'VNN2'</li><li>'VPS28'</li><li>'VSIG4'</li><li>'XBP1'</li><li>'YBX3'</li><li>'ZAP70'</li><li>'ZBED2'</li><li>'ZBTB16'</li><li>'ZNF683'</li><li>'CCL13'</li><li>'CLEC10A'</li><li>'EPX'</li><li>'IGLC3'</li><li>'SPP1'</li><li>'IGHA1'</li><li>'CXCL11'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'ADA'\n",
       "\\item 'ADGRE1'\n",
       "\\item 'ADGRG3'\n",
       "\\item 'AIM2'\n",
       "\\item 'ALAS2'\n",
       "\\item 'ANXA5'\n",
       "\\item 'AOC3'\n",
       "\\item 'APOBEC3G'\n",
       "\\item 'APOE'\n",
       "\\item 'AQP9'\n",
       "\\item 'ARG1'\n",
       "\\item 'ARL4C'\n",
       "\\item 'ATF6B'\n",
       "\\item 'AURKB'\n",
       "\\item 'AZU1'\n",
       "\\item 'B3GAT1'\n",
       "\\item 'BACH2'\n",
       "\\item 'BAX'\n",
       "\\item 'BCL11B'\n",
       "\\item 'BCL2'\n",
       "\\item 'BCL2A1'\n",
       "\\item 'BCL6'\n",
       "\\item 'BIN2'\n",
       "\\item 'BIRC3'\n",
       "\\item 'BLK'\n",
       "\\item 'BLNK'\n",
       "\\item 'BPI'\n",
       "\\item 'BTG1'\n",
       "\\item 'BTLA'\n",
       "\\item 'C1QA'\n",
       "\\item 'CASP5'\n",
       "\\item 'CBLB'\n",
       "\\item 'CCL1'\n",
       "\\item 'CCL17'\n",
       "\\item 'CCL19'\n",
       "\\item 'CCL20'\n",
       "\\item 'CCL22'\n",
       "\\item 'CCL3'\n",
       "\\item 'CCL4'\n",
       "\\item 'CCL5'\n",
       "\\item 'CCND2'\n",
       "\\item 'CCR1'\n",
       "\\item 'CCR10'\n",
       "\\item 'CCR2'\n",
       "\\item 'CCR3'\n",
       "\\item 'CCR4'\n",
       "\\item 'CCR5'\n",
       "\\item 'CCR7'\n",
       "\\item 'CCR8'\n",
       "\\item 'CCR9'\n",
       "\\item 'CD14'\n",
       "\\item 'CD160'\n",
       "\\item 'CD163'\n",
       "\\item 'CD1B'\n",
       "\\item 'CD1C'\n",
       "\\item 'CD2'\n",
       "\\item 'CD200'\n",
       "\\item 'CD209'\n",
       "\\item 'CD22'\n",
       "\\item 'CD24'\n",
       "\\item 'CD244'\n",
       "\\item 'CD247'\n",
       "\\item 'CD27'\n",
       "\\item 'CD274'\n",
       "\\item 'CD28'\n",
       "\\item 'CD300A'\n",
       "\\item 'CD33'\n",
       "\\item 'CD34'\n",
       "\\item 'CD36'\n",
       "\\item 'CD37'\n",
       "\\item 'CD38'\n",
       "\\item 'CD3D'\n",
       "\\item 'CD3E'\n",
       "\\item 'CD3G'\n",
       "\\item 'CD4'\n",
       "\\item 'CD40'\n",
       "\\item 'CD44'\n",
       "\\item 'CD48'\n",
       "\\item 'CD5'\n",
       "\\item 'CD52'\n",
       "\\item 'CD6'\n",
       "\\item 'CD63'\n",
       "\\item 'CD69'\n",
       "\\item 'CD7'\n",
       "\\item 'CD70'\n",
       "\\item 'CD72'\n",
       "\\item 'CD74'\n",
       "\\item 'CD79A'\n",
       "\\item 'CD79B'\n",
       "\\item 'CD80'\n",
       "\\item 'CD86'\n",
       "\\item 'CD8A'\n",
       "\\item 'CD8B'\n",
       "\\item 'CD9'\n",
       "\\item 'CEACAM8'\n",
       "\\item 'CHI3L1'\n",
       "\\item 'CHI3L2'\n",
       "\\item 'CLC'\n",
       "\\item 'CLEC4D'\n",
       "\\item 'CLEC4E'\n",
       "\\item 'CMKLR1'\n",
       "\\item 'CMTM2'\n",
       "\\item 'CNOT2'\n",
       "\\item 'CNTNAP3'\n",
       "\\item 'CR2'\n",
       "\\item 'CSF2'\n",
       "\\item 'CSF3'\n",
       "\\item 'CST7'\n",
       "\\item 'CTLA4'\n",
       "\\item 'CTSD'\n",
       "\\item 'CTSG'\n",
       "\\item 'CTSW'\n",
       "\\item 'CX3CR1'\n",
       "\\item 'CXCL10'\n",
       "\\item 'CXCL13'\n",
       "\\item 'CXCL16'\n",
       "\\item 'CXCL2'\n",
       "\\item 'CXCL3'\n",
       "\\item 'CXCL5'\n",
       "\\item 'CXCL8'\n",
       "\\item 'CXCR1'\n",
       "\\item 'CXCR2'\n",
       "\\item 'CXCR3'\n",
       "\\item 'CXCR4'\n",
       "\\item 'CXCR5'\n",
       "\\item 'CXCR6'\n",
       "\\item 'DOCK8'\n",
       "\\item 'DPP4'\n",
       "\\item 'DUSP1'\n",
       "\\item 'DUSP2'\n",
       "\\item 'DUSP4'\n",
       "\\item 'EBF1'\n",
       "\\item 'EGR1'\n",
       "\\item 'EGR3'\n",
       "\\item 'ELANE'\n",
       "\\item 'ENTPD1'\n",
       "\\item 'EOMES'\n",
       "\\item 'F13A1'\n",
       "\\item 'F5'\n",
       "\\item 'FAS'\n",
       "\\item 'FASLG'\n",
       "\\item 'FCER1G'\n",
       "\\item 'FCER2'\n",
       "\\item 'FCGR3A'\n",
       "\\item 'FCN1'\n",
       "\\item 'FLT3'\n",
       "\\item 'FN1'\n",
       "\\item 'FOSB'\n",
       "\\item 'FOSL1'\n",
       "\\item 'FOXO1'\n",
       "\\item 'FOXP1'\n",
       "\\item 'FOXP3'\n",
       "\\item 'FTH1'\n",
       "\\item 'FUT4'\n",
       "\\item 'FYN'\n",
       "\\item 'GAB2'\n",
       "\\item 'GAPDH'\n",
       "\\item 'GIMAP2'\n",
       "\\item 'GIMAP5'\n",
       "\\item 'GNAI2'\n",
       "\\item 'GNLY'\n",
       "\\item 'GZMA'\n",
       "\\item 'GZMB'\n",
       "\\item 'GZMH'\n",
       "\\item 'GZMK'\n",
       "\\item 'HAVCR2'\n",
       "\\item 'HLA-A'\n",
       "\\item 'HLA-DMA'\n",
       "\\item 'HLA-DPA1'\n",
       "\\item 'HLA-DQB1'\n",
       "\\item 'HLA-DRA'\n",
       "\\item 'HMMR'\n",
       "\\item 'ICAM1'\n",
       "\\item 'ICOS'\n",
       "\\item 'IER3'\n",
       "\\item 'IFITM2'\n",
       "\\item 'IFITM3'\n",
       "\\item 'IFNG'\n",
       "\\item 'IFNGR1'\n",
       "\\item 'IGBP1'\n",
       "\\item 'IGHD'\n",
       "\\item 'IGHE'\n",
       "\\item 'IGHG1'\n",
       "\\item 'IGHG2'\n",
       "\\item 'IGHG4'\n",
       "\\item 'IGHM'\n",
       "\\item 'IGKC'\n",
       "\\item 'IKZF1'\n",
       "\\item 'IKZF2'\n",
       "\\item 'IL12A'\n",
       "\\item 'IL12RB1'\n",
       "\\item 'IL12RB2'\n",
       "\\item 'IL13'\n",
       "\\item 'IL15'\n",
       "\\item 'IL15RA'\n",
       "\\item 'IL18'\n",
       "\\item 'IL18R1'\n",
       "\\item 'IL18RAP'\n",
       "\\item 'IL1B'\n",
       "\\item 'IL1R2'\n",
       "\\item 'IL1RL1'\n",
       "\\item 'IL1RN'\n",
       "\\item 'IL2'\n",
       "\\item 'IL21'\n",
       "\\item 'IL22'\n",
       "\\item 'IL23R'\n",
       "\\item 'IL2RA'\n",
       "\\item 'IL2RB'\n",
       "\\item 'IL32'\n",
       "\\item 'IL33'\n",
       "\\item 'IL3RA'\n",
       "\\item 'IL4'\n",
       "\\item 'IL4R'\n",
       "\\item 'IL5'\n",
       "\\item 'IL6'\n",
       "\\item 'IL7R'\n",
       "\\item 'IL9'\n",
       "\\item 'IRF4'\n",
       "\\item 'IRF8'\n",
       "\\item 'ITGA4'\n",
       "\\item 'ITGAE'\n",
       "\\item 'ITGAM'\n",
       "\\item 'ITGAX'\n",
       "\\item 'ITGB2'\n",
       "\\item 'JCHAIN'\n",
       "\\item 'JUN'\n",
       "\\item 'JUNB'\n",
       "\\item 'KCNE3'\n",
       "\\item 'KDELR1'\n",
       "\\item 'KIR2DL1'\n",
       "\\item 'KIT'\n",
       "\\item 'KLRB1'\n",
       "\\item 'KLRC1'\n",
       "\\item 'KLRC3'\n",
       "\\item 'KLRC4'\n",
       "\\item 'KLRF1'\n",
       "\\item 'KLRG1'\n",
       "\\item 'KLRK1'\n",
       "\\item 'LAG3'\n",
       "\\item 'LAIR2'\n",
       "\\item 'LAMP1'\n",
       "\\item 'LAMP3'\n",
       "\\item 'LAP3'\n",
       "\\item 'LAT'\n",
       "\\item 'LAT2'\n",
       "\\item 'LCK'\n",
       "\\item 'LEF1'\n",
       "\\item 'LGALS1'\n",
       "\\item 'LGALS3'\n",
       "\\item 'LGALS9'\n",
       "\\item 'LIF'\n",
       "\\item 'LILRB4'\n",
       "\\item 'LIPA'\n",
       "\\item 'LRRC32'\n",
       "\\item 'LTA'\n",
       "\\item 'LTB'\n",
       "\\item 'LY86'\n",
       "\\item 'LYN'\n",
       "\\item 'MCM2'\n",
       "\\item 'MCM4'\n",
       "\\item 'MGST1'\n",
       "\\item 'MITF'\n",
       "\\item 'MKI67'\n",
       "\\item 'MME'\n",
       "\\item 'MMP12'\n",
       "\\item 'MMP9'\n",
       "\\item 'MS4A1'\n",
       "\\item 'MYC'\n",
       "\\item 'MZB1'\n",
       "\\item 'NAMPT'\n",
       "\\item 'NCAM1'\n",
       "\\item 'NCR3'\n",
       "\\item 'NINJ2'\n",
       "\\item 'NKG7'\n",
       "\\item 'NRP1'\n",
       "\\item 'NT5E'\n",
       "\\item 'PASK'\n",
       "\\item 'PCNA'\n",
       "\\item 'PDCD1'\n",
       "\\item 'PDIA4'\n",
       "\\item 'PDIA6'\n",
       "\\item 'PIK3AP1'\n",
       "\\item 'PIK3IP1'\n",
       "\\item 'PMCH'\n",
       "\\item 'POU2AF1'\n",
       "\\item 'PRDM1'\n",
       "\\item 'PRF1'\n",
       "\\item 'PSEN1'\n",
       "\\item 'PTGDR2'\n",
       "\\item 'PTPRC'\n",
       "\\item 'PTTG2'\n",
       "\\item 'QPCT'\n",
       "\\item 'RGS1'\n",
       "\\item 'RNASE2'\n",
       "\\item 'RNASE6'\n",
       "\\item 'RORA'\n",
       "\\item 'RORC'\n",
       "\\item 'RPN2'\n",
       "\\item 'RUNX3'\n",
       "\\item 'S100A10'\n",
       "\\item 'S100A12'\n",
       "\\item 'S100A9'\n",
       "\\item 'SELL'\n",
       "\\item 'SELPLG'\n",
       "\\item 'SLC25A37'\n",
       "\\item 'SLC7A7'\n",
       "\\item 'SNCA'\n",
       "\\item 'STAT1'\n",
       "\\item 'STAT3'\n",
       "\\item 'STAT4'\n",
       "\\item 'STAT5A'\n",
       "\\item 'STAT6'\n",
       "\\item 'TBX21'\n",
       "\\item 'TCF4'\n",
       "\\item 'TCF7'\n",
       "\\item 'TGFB1'\n",
       "\\item 'TGFB3'\n",
       "\\item 'TGFBI'\n",
       "\\item 'THBD'\n",
       "\\item 'THBS1'\n",
       "\\item 'TIAF1'\n",
       "\\item 'TIGIT'\n",
       "\\item 'TLR2'\n",
       "\\item 'TLR7'\n",
       "\\item 'TLR8'\n",
       "\\item 'TLR9'\n",
       "\\item 'TMEM97'\n",
       "\\item 'TNF'\n",
       "\\item 'TNFRSF13C'\n",
       "\\item 'TNFRSF25'\n",
       "\\item 'TNFRSF4'\n",
       "\\item 'TNFRSF8'\n",
       "\\item 'TNFRSF9'\n",
       "\\item 'TNFSF10'\n",
       "\\item 'TNFSF13'\n",
       "\\item 'TNFSF13B'\n",
       "\\item 'TNFSF14'\n",
       "\\item 'TNFSF8'\n",
       "\\item 'TOP2A'\n",
       "\\item 'TPSAB1'\n",
       "\\item 'TRAC'\n",
       "\\item 'TRAT1'\n",
       "\\item 'TRBC2'\n",
       "\\item 'TRDC'\n",
       "\\item 'TREM1'\n",
       "\\item 'TRIB2'\n",
       "\\item 'TSPAN32'\n",
       "\\item 'TXK'\n",
       "\\item 'TYMS'\n",
       "\\item 'UBE2C'\n",
       "\\item 'VEGFA'\n",
       "\\item 'VMO1'\n",
       "\\item 'VNN2'\n",
       "\\item 'VPS28'\n",
       "\\item 'VSIG4'\n",
       "\\item 'XBP1'\n",
       "\\item 'YBX3'\n",
       "\\item 'ZAP70'\n",
       "\\item 'ZBED2'\n",
       "\\item 'ZBTB16'\n",
       "\\item 'ZNF683'\n",
       "\\item 'CCL13'\n",
       "\\item 'CLEC10A'\n",
       "\\item 'EPX'\n",
       "\\item 'IGLC3'\n",
       "\\item 'SPP1'\n",
       "\\item 'IGHA1'\n",
       "\\item 'CXCL11'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'ADA'\n",
       "2. 'ADGRE1'\n",
       "3. 'ADGRG3'\n",
       "4. 'AIM2'\n",
       "5. 'ALAS2'\n",
       "6. 'ANXA5'\n",
       "7. 'AOC3'\n",
       "8. 'APOBEC3G'\n",
       "9. 'APOE'\n",
       "10. 'AQP9'\n",
       "11. 'ARG1'\n",
       "12. 'ARL4C'\n",
       "13. 'ATF6B'\n",
       "14. 'AURKB'\n",
       "15. 'AZU1'\n",
       "16. 'B3GAT1'\n",
       "17. 'BACH2'\n",
       "18. 'BAX'\n",
       "19. 'BCL11B'\n",
       "20. 'BCL2'\n",
       "21. 'BCL2A1'\n",
       "22. 'BCL6'\n",
       "23. 'BIN2'\n",
       "24. 'BIRC3'\n",
       "25. 'BLK'\n",
       "26. 'BLNK'\n",
       "27. 'BPI'\n",
       "28. 'BTG1'\n",
       "29. 'BTLA'\n",
       "30. 'C1QA'\n",
       "31. 'CASP5'\n",
       "32. 'CBLB'\n",
       "33. 'CCL1'\n",
       "34. 'CCL17'\n",
       "35. 'CCL19'\n",
       "36. 'CCL20'\n",
       "37. 'CCL22'\n",
       "38. 'CCL3'\n",
       "39. 'CCL4'\n",
       "40. 'CCL5'\n",
       "41. 'CCND2'\n",
       "42. 'CCR1'\n",
       "43. 'CCR10'\n",
       "44. 'CCR2'\n",
       "45. 'CCR3'\n",
       "46. 'CCR4'\n",
       "47. 'CCR5'\n",
       "48. 'CCR7'\n",
       "49. 'CCR8'\n",
       "50. 'CCR9'\n",
       "51. 'CD14'\n",
       "52. 'CD160'\n",
       "53. 'CD163'\n",
       "54. 'CD1B'\n",
       "55. 'CD1C'\n",
       "56. 'CD2'\n",
       "57. 'CD200'\n",
       "58. 'CD209'\n",
       "59. 'CD22'\n",
       "60. 'CD24'\n",
       "61. 'CD244'\n",
       "62. 'CD247'\n",
       "63. 'CD27'\n",
       "64. 'CD274'\n",
       "65. 'CD28'\n",
       "66. 'CD300A'\n",
       "67. 'CD33'\n",
       "68. 'CD34'\n",
       "69. 'CD36'\n",
       "70. 'CD37'\n",
       "71. 'CD38'\n",
       "72. 'CD3D'\n",
       "73. 'CD3E'\n",
       "74. 'CD3G'\n",
       "75. 'CD4'\n",
       "76. 'CD40'\n",
       "77. 'CD44'\n",
       "78. 'CD48'\n",
       "79. 'CD5'\n",
       "80. 'CD52'\n",
       "81. 'CD6'\n",
       "82. 'CD63'\n",
       "83. 'CD69'\n",
       "84. 'CD7'\n",
       "85. 'CD70'\n",
       "86. 'CD72'\n",
       "87. 'CD74'\n",
       "88. 'CD79A'\n",
       "89. 'CD79B'\n",
       "90. 'CD80'\n",
       "91. 'CD86'\n",
       "92. 'CD8A'\n",
       "93. 'CD8B'\n",
       "94. 'CD9'\n",
       "95. 'CEACAM8'\n",
       "96. 'CHI3L1'\n",
       "97. 'CHI3L2'\n",
       "98. 'CLC'\n",
       "99. 'CLEC4D'\n",
       "100. 'CLEC4E'\n",
       "101. 'CMKLR1'\n",
       "102. 'CMTM2'\n",
       "103. 'CNOT2'\n",
       "104. 'CNTNAP3'\n",
       "105. 'CR2'\n",
       "106. 'CSF2'\n",
       "107. 'CSF3'\n",
       "108. 'CST7'\n",
       "109. 'CTLA4'\n",
       "110. 'CTSD'\n",
       "111. 'CTSG'\n",
       "112. 'CTSW'\n",
       "113. 'CX3CR1'\n",
       "114. 'CXCL10'\n",
       "115. 'CXCL13'\n",
       "116. 'CXCL16'\n",
       "117. 'CXCL2'\n",
       "118. 'CXCL3'\n",
       "119. 'CXCL5'\n",
       "120. 'CXCL8'\n",
       "121. 'CXCR1'\n",
       "122. 'CXCR2'\n",
       "123. 'CXCR3'\n",
       "124. 'CXCR4'\n",
       "125. 'CXCR5'\n",
       "126. 'CXCR6'\n",
       "127. 'DOCK8'\n",
       "128. 'DPP4'\n",
       "129. 'DUSP1'\n",
       "130. 'DUSP2'\n",
       "131. 'DUSP4'\n",
       "132. 'EBF1'\n",
       "133. 'EGR1'\n",
       "134. 'EGR3'\n",
       "135. 'ELANE'\n",
       "136. 'ENTPD1'\n",
       "137. 'EOMES'\n",
       "138. 'F13A1'\n",
       "139. 'F5'\n",
       "140. 'FAS'\n",
       "141. 'FASLG'\n",
       "142. 'FCER1G'\n",
       "143. 'FCER2'\n",
       "144. 'FCGR3A'\n",
       "145. 'FCN1'\n",
       "146. 'FLT3'\n",
       "147. 'FN1'\n",
       "148. 'FOSB'\n",
       "149. 'FOSL1'\n",
       "150. 'FOXO1'\n",
       "151. 'FOXP1'\n",
       "152. 'FOXP3'\n",
       "153. 'FTH1'\n",
       "154. 'FUT4'\n",
       "155. 'FYN'\n",
       "156. 'GAB2'\n",
       "157. 'GAPDH'\n",
       "158. 'GIMAP2'\n",
       "159. 'GIMAP5'\n",
       "160. 'GNAI2'\n",
       "161. 'GNLY'\n",
       "162. 'GZMA'\n",
       "163. 'GZMB'\n",
       "164. 'GZMH'\n",
       "165. 'GZMK'\n",
       "166. 'HAVCR2'\n",
       "167. 'HLA-A'\n",
       "168. 'HLA-DMA'\n",
       "169. 'HLA-DPA1'\n",
       "170. 'HLA-DQB1'\n",
       "171. 'HLA-DRA'\n",
       "172. 'HMMR'\n",
       "173. 'ICAM1'\n",
       "174. 'ICOS'\n",
       "175. 'IER3'\n",
       "176. 'IFITM2'\n",
       "177. 'IFITM3'\n",
       "178. 'IFNG'\n",
       "179. 'IFNGR1'\n",
       "180. 'IGBP1'\n",
       "181. 'IGHD'\n",
       "182. 'IGHE'\n",
       "183. 'IGHG1'\n",
       "184. 'IGHG2'\n",
       "185. 'IGHG4'\n",
       "186. 'IGHM'\n",
       "187. 'IGKC'\n",
       "188. 'IKZF1'\n",
       "189. 'IKZF2'\n",
       "190. 'IL12A'\n",
       "191. 'IL12RB1'\n",
       "192. 'IL12RB2'\n",
       "193. 'IL13'\n",
       "194. 'IL15'\n",
       "195. 'IL15RA'\n",
       "196. 'IL18'\n",
       "197. 'IL18R1'\n",
       "198. 'IL18RAP'\n",
       "199. 'IL1B'\n",
       "200. 'IL1R2'\n",
       "201. 'IL1RL1'\n",
       "202. 'IL1RN'\n",
       "203. 'IL2'\n",
       "204. 'IL21'\n",
       "205. 'IL22'\n",
       "206. 'IL23R'\n",
       "207. 'IL2RA'\n",
       "208. 'IL2RB'\n",
       "209. 'IL32'\n",
       "210. 'IL33'\n",
       "211. 'IL3RA'\n",
       "212. 'IL4'\n",
       "213. 'IL4R'\n",
       "214. 'IL5'\n",
       "215. 'IL6'\n",
       "216. 'IL7R'\n",
       "217. 'IL9'\n",
       "218. 'IRF4'\n",
       "219. 'IRF8'\n",
       "220. 'ITGA4'\n",
       "221. 'ITGAE'\n",
       "222. 'ITGAM'\n",
       "223. 'ITGAX'\n",
       "224. 'ITGB2'\n",
       "225. 'JCHAIN'\n",
       "226. 'JUN'\n",
       "227. 'JUNB'\n",
       "228. 'KCNE3'\n",
       "229. 'KDELR1'\n",
       "230. 'KIR2DL1'\n",
       "231. 'KIT'\n",
       "232. 'KLRB1'\n",
       "233. 'KLRC1'\n",
       "234. 'KLRC3'\n",
       "235. 'KLRC4'\n",
       "236. 'KLRF1'\n",
       "237. 'KLRG1'\n",
       "238. 'KLRK1'\n",
       "239. 'LAG3'\n",
       "240. 'LAIR2'\n",
       "241. 'LAMP1'\n",
       "242. 'LAMP3'\n",
       "243. 'LAP3'\n",
       "244. 'LAT'\n",
       "245. 'LAT2'\n",
       "246. 'LCK'\n",
       "247. 'LEF1'\n",
       "248. 'LGALS1'\n",
       "249. 'LGALS3'\n",
       "250. 'LGALS9'\n",
       "251. 'LIF'\n",
       "252. 'LILRB4'\n",
       "253. 'LIPA'\n",
       "254. 'LRRC32'\n",
       "255. 'LTA'\n",
       "256. 'LTB'\n",
       "257. 'LY86'\n",
       "258. 'LYN'\n",
       "259. 'MCM2'\n",
       "260. 'MCM4'\n",
       "261. 'MGST1'\n",
       "262. 'MITF'\n",
       "263. 'MKI67'\n",
       "264. 'MME'\n",
       "265. 'MMP12'\n",
       "266. 'MMP9'\n",
       "267. 'MS4A1'\n",
       "268. 'MYC'\n",
       "269. 'MZB1'\n",
       "270. 'NAMPT'\n",
       "271. 'NCAM1'\n",
       "272. 'NCR3'\n",
       "273. 'NINJ2'\n",
       "274. 'NKG7'\n",
       "275. 'NRP1'\n",
       "276. 'NT5E'\n",
       "277. 'PASK'\n",
       "278. 'PCNA'\n",
       "279. 'PDCD1'\n",
       "280. 'PDIA4'\n",
       "281. 'PDIA6'\n",
       "282. 'PIK3AP1'\n",
       "283. 'PIK3IP1'\n",
       "284. 'PMCH'\n",
       "285. 'POU2AF1'\n",
       "286. 'PRDM1'\n",
       "287. 'PRF1'\n",
       "288. 'PSEN1'\n",
       "289. 'PTGDR2'\n",
       "290. 'PTPRC'\n",
       "291. 'PTTG2'\n",
       "292. 'QPCT'\n",
       "293. 'RGS1'\n",
       "294. 'RNASE2'\n",
       "295. 'RNASE6'\n",
       "296. 'RORA'\n",
       "297. 'RORC'\n",
       "298. 'RPN2'\n",
       "299. 'RUNX3'\n",
       "300. 'S100A10'\n",
       "301. 'S100A12'\n",
       "302. 'S100A9'\n",
       "303. 'SELL'\n",
       "304. 'SELPLG'\n",
       "305. 'SLC25A37'\n",
       "306. 'SLC7A7'\n",
       "307. 'SNCA'\n",
       "308. 'STAT1'\n",
       "309. 'STAT3'\n",
       "310. 'STAT4'\n",
       "311. 'STAT5A'\n",
       "312. 'STAT6'\n",
       "313. 'TBX21'\n",
       "314. 'TCF4'\n",
       "315. 'TCF7'\n",
       "316. 'TGFB1'\n",
       "317. 'TGFB3'\n",
       "318. 'TGFBI'\n",
       "319. 'THBD'\n",
       "320. 'THBS1'\n",
       "321. 'TIAF1'\n",
       "322. 'TIGIT'\n",
       "323. 'TLR2'\n",
       "324. 'TLR7'\n",
       "325. 'TLR8'\n",
       "326. 'TLR9'\n",
       "327. 'TMEM97'\n",
       "328. 'TNF'\n",
       "329. 'TNFRSF13C'\n",
       "330. 'TNFRSF25'\n",
       "331. 'TNFRSF4'\n",
       "332. 'TNFRSF8'\n",
       "333. 'TNFRSF9'\n",
       "334. 'TNFSF10'\n",
       "335. 'TNFSF13'\n",
       "336. 'TNFSF13B'\n",
       "337. 'TNFSF14'\n",
       "338. 'TNFSF8'\n",
       "339. 'TOP2A'\n",
       "340. 'TPSAB1'\n",
       "341. 'TRAC'\n",
       "342. 'TRAT1'\n",
       "343. 'TRBC2'\n",
       "344. 'TRDC'\n",
       "345. 'TREM1'\n",
       "346. 'TRIB2'\n",
       "347. 'TSPAN32'\n",
       "348. 'TXK'\n",
       "349. 'TYMS'\n",
       "350. 'UBE2C'\n",
       "351. 'VEGFA'\n",
       "352. 'VMO1'\n",
       "353. 'VNN2'\n",
       "354. 'VPS28'\n",
       "355. 'VSIG4'\n",
       "356. 'XBP1'\n",
       "357. 'YBX3'\n",
       "358. 'ZAP70'\n",
       "359. 'ZBED2'\n",
       "360. 'ZBTB16'\n",
       "361. 'ZNF683'\n",
       "362. 'CCL13'\n",
       "363. 'CLEC10A'\n",
       "364. 'EPX'\n",
       "365. 'IGLC3'\n",
       "366. 'SPP1'\n",
       "367. 'IGHA1'\n",
       "368. 'CXCL11'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"ADA\"       \"ADGRE1\"    \"ADGRG3\"    \"AIM2\"      \"ALAS2\"     \"ANXA5\"    \n",
       "  [7] \"AOC3\"      \"APOBEC3G\"  \"APOE\"      \"AQP9\"      \"ARG1\"      \"ARL4C\"    \n",
       " [13] \"ATF6B\"     \"AURKB\"     \"AZU1\"      \"B3GAT1\"    \"BACH2\"     \"BAX\"      \n",
       " [19] \"BCL11B\"    \"BCL2\"      \"BCL2A1\"    \"BCL6\"      \"BIN2\"      \"BIRC3\"    \n",
       " [25] \"BLK\"       \"BLNK\"      \"BPI\"       \"BTG1\"      \"BTLA\"      \"C1QA\"     \n",
       " [31] \"CASP5\"     \"CBLB\"      \"CCL1\"      \"CCL17\"     \"CCL19\"     \"CCL20\"    \n",
       " [37] \"CCL22\"     \"CCL3\"      \"CCL4\"      \"CCL5\"      \"CCND2\"     \"CCR1\"     \n",
       " [43] \"CCR10\"     \"CCR2\"      \"CCR3\"      \"CCR4\"      \"CCR5\"      \"CCR7\"     \n",
       " [49] \"CCR8\"      \"CCR9\"      \"CD14\"      \"CD160\"     \"CD163\"     \"CD1B\"     \n",
       " [55] \"CD1C\"      \"CD2\"       \"CD200\"     \"CD209\"     \"CD22\"      \"CD24\"     \n",
       " [61] \"CD244\"     \"CD247\"     \"CD27\"      \"CD274\"     \"CD28\"      \"CD300A\"   \n",
       " [67] \"CD33\"      \"CD34\"      \"CD36\"      \"CD37\"      \"CD38\"      \"CD3D\"     \n",
       " [73] \"CD3E\"      \"CD3G\"      \"CD4\"       \"CD40\"      \"CD44\"      \"CD48\"     \n",
       " [79] \"CD5\"       \"CD52\"      \"CD6\"       \"CD63\"      \"CD69\"      \"CD7\"      \n",
       " [85] \"CD70\"      \"CD72\"      \"CD74\"      \"CD79A\"     \"CD79B\"     \"CD80\"     \n",
       " [91] \"CD86\"      \"CD8A\"      \"CD8B\"      \"CD9\"       \"CEACAM8\"   \"CHI3L1\"   \n",
       " [97] \"CHI3L2\"    \"CLC\"       \"CLEC4D\"    \"CLEC4E\"    \"CMKLR1\"    \"CMTM2\"    \n",
       "[103] \"CNOT2\"     \"CNTNAP3\"   \"CR2\"       \"CSF2\"      \"CSF3\"      \"CST7\"     \n",
       "[109] \"CTLA4\"     \"CTSD\"      \"CTSG\"      \"CTSW\"      \"CX3CR1\"    \"CXCL10\"   \n",
       "[115] \"CXCL13\"    \"CXCL16\"    \"CXCL2\"     \"CXCL3\"     \"CXCL5\"     \"CXCL8\"    \n",
       "[121] \"CXCR1\"     \"CXCR2\"     \"CXCR3\"     \"CXCR4\"     \"CXCR5\"     \"CXCR6\"    \n",
       "[127] \"DOCK8\"     \"DPP4\"      \"DUSP1\"     \"DUSP2\"     \"DUSP4\"     \"EBF1\"     \n",
       "[133] \"EGR1\"      \"EGR3\"      \"ELANE\"     \"ENTPD1\"    \"EOMES\"     \"F13A1\"    \n",
       "[139] \"F5\"        \"FAS\"       \"FASLG\"     \"FCER1G\"    \"FCER2\"     \"FCGR3A\"   \n",
       "[145] \"FCN1\"      \"FLT3\"      \"FN1\"       \"FOSB\"      \"FOSL1\"     \"FOXO1\"    \n",
       "[151] \"FOXP1\"     \"FOXP3\"     \"FTH1\"      \"FUT4\"      \"FYN\"       \"GAB2\"     \n",
       "[157] \"GAPDH\"     \"GIMAP2\"    \"GIMAP5\"    \"GNAI2\"     \"GNLY\"      \"GZMA\"     \n",
       "[163] \"GZMB\"      \"GZMH\"      \"GZMK\"      \"HAVCR2\"    \"HLA-A\"     \"HLA-DMA\"  \n",
       "[169] \"HLA-DPA1\"  \"HLA-DQB1\"  \"HLA-DRA\"   \"HMMR\"      \"ICAM1\"     \"ICOS\"     \n",
       "[175] \"IER3\"      \"IFITM2\"    \"IFITM3\"    \"IFNG\"      \"IFNGR1\"    \"IGBP1\"    \n",
       "[181] \"IGHD\"      \"IGHE\"      \"IGHG1\"     \"IGHG2\"     \"IGHG4\"     \"IGHM\"     \n",
       "[187] \"IGKC\"      \"IKZF1\"     \"IKZF2\"     \"IL12A\"     \"IL12RB1\"   \"IL12RB2\"  \n",
       "[193] \"IL13\"      \"IL15\"      \"IL15RA\"    \"IL18\"      \"IL18R1\"    \"IL18RAP\"  \n",
       "[199] \"IL1B\"      \"IL1R2\"     \"IL1RL1\"    \"IL1RN\"     \"IL2\"       \"IL21\"     \n",
       "[205] \"IL22\"      \"IL23R\"     \"IL2RA\"     \"IL2RB\"     \"IL32\"      \"IL33\"     \n",
       "[211] \"IL3RA\"     \"IL4\"       \"IL4R\"      \"IL5\"       \"IL6\"       \"IL7R\"     \n",
       "[217] \"IL9\"       \"IRF4\"      \"IRF8\"      \"ITGA4\"     \"ITGAE\"     \"ITGAM\"    \n",
       "[223] \"ITGAX\"     \"ITGB2\"     \"JCHAIN\"    \"JUN\"       \"JUNB\"      \"KCNE3\"    \n",
       "[229] \"KDELR1\"    \"KIR2DL1\"   \"KIT\"       \"KLRB1\"     \"KLRC1\"     \"KLRC3\"    \n",
       "[235] \"KLRC4\"     \"KLRF1\"     \"KLRG1\"     \"KLRK1\"     \"LAG3\"      \"LAIR2\"    \n",
       "[241] \"LAMP1\"     \"LAMP3\"     \"LAP3\"      \"LAT\"       \"LAT2\"      \"LCK\"      \n",
       "[247] \"LEF1\"      \"LGALS1\"    \"LGALS3\"    \"LGALS9\"    \"LIF\"       \"LILRB4\"   \n",
       "[253] \"LIPA\"      \"LRRC32\"    \"LTA\"       \"LTB\"       \"LY86\"      \"LYN\"      \n",
       "[259] \"MCM2\"      \"MCM4\"      \"MGST1\"     \"MITF\"      \"MKI67\"     \"MME\"      \n",
       "[265] \"MMP12\"     \"MMP9\"      \"MS4A1\"     \"MYC\"       \"MZB1\"      \"NAMPT\"    \n",
       "[271] \"NCAM1\"     \"NCR3\"      \"NINJ2\"     \"NKG7\"      \"NRP1\"      \"NT5E\"     \n",
       "[277] \"PASK\"      \"PCNA\"      \"PDCD1\"     \"PDIA4\"     \"PDIA6\"     \"PIK3AP1\"  \n",
       "[283] \"PIK3IP1\"   \"PMCH\"      \"POU2AF1\"   \"PRDM1\"     \"PRF1\"      \"PSEN1\"    \n",
       "[289] \"PTGDR2\"    \"PTPRC\"     \"PTTG2\"     \"QPCT\"      \"RGS1\"      \"RNASE2\"   \n",
       "[295] \"RNASE6\"    \"RORA\"      \"RORC\"      \"RPN2\"      \"RUNX3\"     \"S100A10\"  \n",
       "[301] \"S100A12\"   \"S100A9\"    \"SELL\"      \"SELPLG\"    \"SLC25A37\"  \"SLC7A7\"   \n",
       "[307] \"SNCA\"      \"STAT1\"     \"STAT3\"     \"STAT4\"     \"STAT5A\"    \"STAT6\"    \n",
       "[313] \"TBX21\"     \"TCF4\"      \"TCF7\"      \"TGFB1\"     \"TGFB3\"     \"TGFBI\"    \n",
       "[319] \"THBD\"      \"THBS1\"     \"TIAF1\"     \"TIGIT\"     \"TLR2\"      \"TLR7\"     \n",
       "[325] \"TLR8\"      \"TLR9\"      \"TMEM97\"    \"TNF\"       \"TNFRSF13C\" \"TNFRSF25\" \n",
       "[331] \"TNFRSF4\"   \"TNFRSF8\"   \"TNFRSF9\"   \"TNFSF10\"   \"TNFSF13\"   \"TNFSF13B\" \n",
       "[337] \"TNFSF14\"   \"TNFSF8\"    \"TOP2A\"     \"TPSAB1\"    \"TRAC\"      \"TRAT1\"    \n",
       "[343] \"TRBC2\"     \"TRDC\"      \"TREM1\"     \"TRIB2\"     \"TSPAN32\"   \"TXK\"      \n",
       "[349] \"TYMS\"      \"UBE2C\"     \"VEGFA\"     \"VMO1\"      \"VNN2\"      \"VPS28\"    \n",
       "[355] \"VSIG4\"     \"XBP1\"      \"YBX3\"      \"ZAP70\"     \"ZBED2\"     \"ZBTB16\"   \n",
       "[361] \"ZNF683\"    \"CCL13\"     \"CLEC10A\"   \"EPX\"       \"IGLC3\"     \"SPP1\"     \n",
       "[367] \"IGHA1\"     \"CXCL11\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scaling features for provided objects\n",
      "\n",
      "Computing within dataset neighborhoods\n",
      "\n",
      "Finding all pairwise anchors\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Projecting new data onto SVD\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4558 anchors\n",
      "\n",
      "Merging dataset 1 into 2\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Warning message:\n",
      "“Keys should be one or more alphanumeric characters followed by an underscore, setting key from rna.rpca_ to rnarpca_”\n",
      "Warning message:\n",
      "“Keys should be one or more alphanumeric characters followed by an underscore, setting key from rna.rpca_ to rnarpca_”\n",
      "Warning message:\n",
      "“Keys should be one or more alphanumeric characters followed by an underscore, setting key from rna.rpca_ to rnarpca_”\n",
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 4 features requested have zero variance (running reduction without them): AQP9, CCL17, IL33, EPX”\n",
      "PC_ 1 \n",
      "Positive:  GZMH, GZMB, FCGR3A, GNLY, NKG7, LGALS1, CX3CR1, CST7, PRF1, ZNF683 \n",
      "\t   ITGAM, ITGB2, CCL5, GZMA, IFNG, B3GAT1, CD63, APOBEC3G, RUNX3, CD300A \n",
      "\t   KLRC3, CTSW, LAIR2, KIR2DL1, CTSD, KLRK1, CCL4, CD3G, CD8B, HLA-DRA \n",
      "Negative:  IL7R, LTB, GZMK, TCF7, KLRB1, TNFRSF25, CD27, MYC, RORC, ZBTB16 \n",
      "\t   TRAT1, CCR7, DPP4, JUNB, PIK3IP1, DUSP2, HLA-A, PASK, IL18RAP, CCR2 \n",
      "\t   CXCR4, DUSP1, CD69, CXCR6, LEF1, CD28, NT5E, BIRC3, NCR3, BTG1 \n",
      "PC_ 2 \n",
      "Positive:  GNLY, CCL5, NKG7, IGLC3, S100A12, GZMB, ZNF683, CD1C, ALAS2, LIF \n",
      "\t   F13A1, SNCA, CD80, TLR2, IL5, IGHG2, MMP12, CXCL8, ADGRG3, CTSG \n",
      "\t   FCER2, CD33, CXCL2, CD14, CLEC4D, GZMH, CXCR2, TRDC, ZBED2, CNTNAP3 \n",
      "Negative:  CD27, DUSP1, LCK, CD69, GAPDH, CD74, JUNB, RGS1, FOSB, DUSP2 \n",
      "\t   CD3E, BTG1, CD48, HLA-DRA, CD6, CD2, GIMAP2, CXCR3, CD44, FYN \n",
      "\t   GNAI2, SELL, PTPRC, CD3G, ITGA4, CD37, BIN2, ADA, CD5, LAT \n",
      "PC_ 3 \n",
      "Positive:  KLRB1, NKG7, GZMA, RORC, ZBTB16, NCR3, IL18RAP, CST7, GZMK, KLRG1 \n",
      "\t   CCL5, DPP4, CCR2, CTSW, PRF1, IL7R, ARL4C, PIK3AP1, CXCR6, KLRF1 \n",
      "\t   TNFRSF25, IL18R1, PRDM1, RORA, EOMES, BCL2, DUSP1, RUNX3, IFNGR1, CD247 \n",
      "Negative:  LEF1, SELL, PASK, CCR7, CD8B, YBX3, TOP2A, TYMS, CD79A, CXCR3 \n",
      "\t   CD27, KLRK1, AURKB, UBE2C, HLA-DRA, IGHM, LGALS1, LGALS9, ICOS, BTLA \n",
      "\t   CTLA4, ZNF683, MKI67, MCM2, MCM4, HLA-DMA, CD38, TIGIT, FOSB, RGS1 \n",
      "PC_ 4 \n",
      "Positive:  TYMS, TOP2A, AURKB, UBE2C, MCM2, MCM4, CCR5, GZMA, CCL5, HLA-DRA \n",
      "\t   CD38, MKI67, GZMK, CTLA4, HLA-DMA, PCNA, NKG7, CCR2, CXCR6, ENTPD1 \n",
      "\t   CST7, ADA, HLA-DQB1, KLRB1, CD27, HAVCR2, PDCD1, CHI3L2, ANXA5, CD70 \n",
      "Negative:  DUSP1, JUNB, FOSB, CXCR4, BTG1, JUN, LEF1, CCR7, TCF7, KLRC3 \n",
      "\t   PIK3IP1, ITGAM, CD69, IFITM2, KLRK1, KLRF1, CD37, KIR2DL1, RUNX3, CD3E \n",
      "\t   NT5E, HLA-A, CD300A, CD48, ITGAX, CD8B, CTSW, B3GAT1, CD7, FCGR3A \n",
      "PC_ 5 \n",
      "Positive:  DUSP2, RGS1, CCL5, CXCR4, CD74, CST7, HLA-DMA, FOSB, DUSP1, NKG7 \n",
      "\t   GZMK, HLA-DRA, CD160, TIGIT, YBX3, CD69, GZMA, MS4A1, JUN, JUNB \n",
      "\t   HLA-DPA1, CCL4, CHI3L2, IFNG, BTG1, HLA-DQB1, LAIR2, TGFBI, CCL3, HLA-A \n",
      "Negative:  ZNF683, CD52, GNLY, LTB, LGALS3, SELPLG, IFITM2, TNFRSF25, SELL, KIR2DL1 \n",
      "\t   NCR3, IL7R, LEF1, IL18RAP, IL32, DPP4, MYC, RORC, KLRC1, GZMB \n",
      "\t   PRF1, ZBTB16, TXK, S100A10, ANXA5, CD300A, CX3CR1, KLRC3, CCR2, CD247 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 3\n",
      "  .. ..$ RNA     :Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:3268808] 7 11 12 13 17 22 23 27 37 38 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:53195] 0 75 134 207 277 319 375 431 475 557 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 368 53194\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:3268808] 353 361 111 1 252 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:3268808] 7 11 12 13 17 22 23 27 37 38 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:53195] 0 75 134 207 277 319 375 431 475 557 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 368 53194\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:3268808] 9.13 9.15 7.97 3.3 8.79 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data   : num[0 , 0 ] \n",
      "  .. .. .. ..@ key          : chr \"rna_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : logi(0) \n",
      "  .. .. .. ..@ meta.features:'data.frame':\t368 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : list()\n",
      "  .. ..$ ADT     :Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:1117072] 0 1 2 3 4 6 7 8 9 10 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:53195] 0 27 50 71 95 117 139 166 183 208 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 30 53194\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:30] \"CCR7\" \"CD11c\" \"CD127\" \"CD134\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:1117072] 4 9 37 7 10 368 3 43 1 5 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         : num [1:30, 1:53194] 0.15 0.31 0.914 0.249 0.339 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:30] \"CCR7\" \"CD11c\" \"CD127\" \"CD134\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. ..@ scale.data   : num[0 , 0 ] \n",
      "  .. .. .. ..@ key          : chr \"adt_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : logi(0) \n",
      "  .. .. .. ..@ meta.features:'data.frame':\t30 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : list()\n",
      "  .. ..$ RNA.RPCA:Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int(0) \n",
      "  .. .. .. .. .. ..@ p       : int 0\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 0 0\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..@ x       : num(0) \n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:5408320] 0 5 6 7 10 11 12 13 15 16 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:53195] 0 239 473 716 963 1176 1402 1638 1859 2122 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 368 53194\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:5408320] -0.251 0.954 -0.106 9.171 -0.33 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data   : num [1:368, 1:53194] -0.64417 -0.03503 -0.025 -0.07869 0.00447 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. ..@ key          : chr \"rnarpca_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. ..@ meta.features:'data.frame':\t368 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : NULL\n",
      "  ..@ meta.data   :'data.frame':\t53194 obs. of  7 variables:\n",
      "  .. ..$ orig.ident  : chr [1:53194] \"TP3_ID2\" \"TP3_ID2\" \"TP3_ID2\" \"TP3_ID2\" ...\n",
      "  .. ..$ nCount_RNA  : num [1:53194] 38417 13181 21478 30520 6272 ...\n",
      "  .. ..$ nFeature_RNA: int [1:53194] 75 59 73 70 42 56 56 44 82 89 ...\n",
      "  .. ..$ nCount_ADT  : num [1:53194] 14060 3504 4328 5057 4018 ...\n",
      "  .. ..$ nFeature_ADT: int [1:53194] 27 23 21 24 22 22 27 17 25 27 ...\n",
      "  .. ..$ batch       : chr [1:53194] \"Batch1-ID2\" \"Batch1-ID2\" \"Batch1-ID2\" \"Batch1-ID2\" ...\n",
      "  .. ..$ BC          : chr [1:53194] \"28\" \"80\" \"185\" \"229\" ...\n",
      "  ..@ active.assay: chr \"RNA.RPCA\"\n",
      "  ..@ active.ident: Factor w/ 11 levels \"humanSampleTag1\",..: 11 11 11 11 11 11 11 11 11 11 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  :List of 1\n",
      "  .. ..$ pca:Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:53194, 1:50] 2.45 -1.93 1.88 -3.77 -1.98 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:53194] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings          : num [1:364, 1:50] 0.03507 -0.00532 -0.00424 -0.00164 -0.00427 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:364] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ global                    : logi FALSE\n",
      "  .. .. .. ..@ stdev                     : num [1:50] 2.94 2.57 1.91 1.74 1.59 ...\n",
      "  .. .. .. ..@ key                       : chr \"PC_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      :List of 1\n",
      "  .. .. .. .. ..$ total.variance: num 254\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"SeuratProject\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 4 0 4\n",
      "  ..@ commands    :List of 4\n",
      "  .. ..$ FindIntegrationAnchors:Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindIntegrationAnchors\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:30:45\"\n",
      "  .. .. .. ..@ assay.used : chr [1:2] \"RNA\" \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"FindIntegrationAnchors(object.list = seu.list, anchor.features = features, \" \"    reduction = \\\"rpca\\\", max.features = 50)\"\n",
      "  .. .. .. ..@ params     :List of 15\n",
      "  .. .. .. .. ..$ assay               : chr [1:2] \"RNA\" \"RNA\"\n",
      "  .. .. .. .. ..$ anchor.features     : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. ..$ scale               : logi TRUE\n",
      "  .. .. .. .. ..$ normalization.method: chr \"LogNormalize\"\n",
      "  .. .. .. .. ..$ reduction           : chr \"pca\"\n",
      "  .. .. .. .. ..$ l2.norm             : logi TRUE\n",
      "  .. .. .. .. ..$ dims                : int [1:30] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ k.anchor            : num 5\n",
      "  .. .. .. .. ..$ k.filter            : logi NA\n",
      "  .. .. .. .. ..$ k.score             : num 30\n",
      "  .. .. .. .. ..$ max.features        : num 50\n",
      "  .. .. .. .. ..$ nn.method           : chr \"annoy\"\n",
      "  .. .. .. .. ..$ n.trees             : num 50\n",
      "  .. .. .. .. ..$ eps                 : num 0\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ withCallingHandlers   :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"withCallingHandlers\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:31:07\"\n",
      "  .. .. .. ..@ assay.used : NULL\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"withCallingHandlers(expr, warning = function(w) if (inherits(w, \" \"    classes)) tryInvokeRestart(\\\"muffleWarning\\\"))\"\n",
      "  .. .. .. ..@ params     :List of 9\n",
      "  .. .. .. .. ..$ new.assay.name      : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ normalization.method: chr \"LogNormalize\"\n",
      "  .. .. .. .. ..$ features            : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. ..$ dims                : int [1:30] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ k.weight            : num 100\n",
      "  .. .. .. .. ..$ sd.weight           : num 1\n",
      "  .. .. .. .. ..$ preserve.order      : logi FALSE\n",
      "  .. .. .. .. ..$ eps                 : num 0\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ ScaleData.RNA.RPCA    :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"ScaleData.RNA.RPCA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:31:07\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ call.string: chr \"ScaleData(seu, verbose = FALSE)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ features          : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. ..$ assay             : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ model.use         : chr \"linear\"\n",
      "  .. .. .. .. ..$ use.umi           : logi FALSE\n",
      "  .. .. .. .. ..$ do.scale          : logi TRUE\n",
      "  .. .. .. .. ..$ do.center         : logi TRUE\n",
      "  .. .. .. .. ..$ scale.max         : num 10\n",
      "  .. .. .. .. ..$ block.size        : num 1000\n",
      "  .. .. .. .. ..$ min.cells.to.block: num 3000\n",
      "  .. .. .. .. ..$ verbose           : logi FALSE\n",
      "  .. ..$ RunPCA.RNA.RPCA       :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunPCA.RNA.RPCA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:31:32\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ call.string: chr \"RunPCA(object = seu, npcs = 50)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ assay          : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ npcs           : num 50\n",
      "  .. .. .. .. ..$ rev.pca        : logi FALSE\n",
      "  .. .. .. .. ..$ weight.by.var  : logi TRUE\n",
      "  .. .. .. .. ..$ verbose        : logi TRUE\n",
      "  .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5\n",
      "  .. .. .. .. ..$ nfeatures.print: num 30\n",
      "  .. .. .. .. ..$ reduction.name : chr \"pca\"\n",
      "  .. .. .. .. ..$ reduction.key  : chr \"PC_\"\n",
      "  .. .. .. .. ..$ seed.use       : num 42\n",
      "  ..@ tools       :List of 1\n",
      "  .. ..$ Integration:Formal class 'IntegrationData' [package \"Seurat\"] with 7 slots\n",
      "  .. .. .. ..@ neighbors         : NULL\n",
      "  .. .. .. ..@ weights           : NULL\n",
      "  .. .. .. ..@ integration.matrix: NULL\n",
      "  .. .. .. ..@ anchors           :'data.frame':\t9116 obs. of  5 variables:\n",
      "  .. .. .. .. ..$ cell1   : num [1:9116] 1 12 15 19 21 25 30 31 33 41 ...\n",
      "  .. .. .. .. ..$ cell2   : num [1:9116] 23464 21622 38181 16222 1807 ...\n",
      "  .. .. .. .. ..$ score   : num [1:9116] 0.25 0.438 0.562 0.875 0.688 ...\n",
      "  .. .. .. .. ..$ dataset1: int [1:9116] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ dataset2: int [1:9116] 2 2 2 2 2 2 2 2 2 2 ...\n",
      "  .. .. .. ..@ offsets           : NULL\n",
      "  .. .. .. ..@ objects.ncell     : NULL\n",
      "  .. .. .. ..@ sample.tree       : num [1, 1:2] -2 -1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 16232 rows containing missing values (`geom_point()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li><li>10</li><li>11</li><li>12</li><li>13</li><li>14</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 4\n",
       "\\item 5\n",
       "\\item 6\n",
       "\\item 7\n",
       "\\item 8\n",
       "\\item 9\n",
       "\\item 10\n",
       "\\item 11\n",
       "\\item 12\n",
       "\\item 13\n",
       "\\item 14\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 3\n",
       "4. 4\n",
       "5. 5\n",
       "6. 6\n",
       "7. 7\n",
       "8. 8\n",
       "9. 9\n",
       "10. 10\n",
       "11. 11\n",
       "12. 12\n",
       "13. 13\n",
       "14. 14\n",
       "15. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing across cells\n",
      "\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mUsing an external vector in selections was deprecated in tidyselect 1.1.0.\n",
      "\u001b[36mℹ\u001b[39m Please use `all_of()` or `any_of()` instead.\n",
      "  # Was:\n",
      "  data %>% select(genes_ADT)\n",
      "\n",
      "  # Now:\n",
      "  data %>% select(all_of(genes_ADT))\n",
      "\n",
      "See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ADTnorm will process all the ADT markers from the ADT matrix:CD27, CD28, CD56, CD62L, CD161\"\n",
      "[1] \"CD27\"\n",
      "                [,1]     [,2]\n",
      "Batch1-ID2  2.087341 6.055210\n",
      "Batch1-ID27 1.893811 5.716496\n",
      "Batch2-1    1.651837 6.393909\n",
      "Batch2-2    1.216360 5.377795\n",
      "Progress:  Each dot is a curve\n",
      "....\n",
      "[1] \"CD28\"\n",
      "                [,1]     [,2]\n",
      "Batch1-ID2  2.216365 5.089162\n",
      "Batch1-ID27 1.783462 4.302122\n",
      "Batch2-1    1.271949 5.285975\n",
      "Batch2-2    1.035795 4.223399\n",
      "Progress:  Each dot is a curve\n",
      "....\n",
      "[1] \"CD56\"\n",
      "                 [,1]     [,2]\n",
      "Batch1-ID2  0.9694961 4.864753\n",
      "Batch1-ID27 0.9694968 4.543956\n",
      "Batch2-1    1.4735885 5.918807\n",
      "Batch2-2    1.1069948 4.864791\n",
      "Progress:  Each dot is a curve\n",
      "....\n",
      "[1] \"CD62L\"\n",
      "                [,1]     [,2]\n",
      "Batch1-ID2  1.711528 6.372667\n",
      "Batch1-ID27 2.180741       NA\n",
      "Batch2-1    1.763647 6.143015\n",
      "Batch2-2    1.242320 5.609762\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 rows containing missing values (`geom_segment()`).”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress:  Each dot is a curve\n",
      "....\n",
      "[1] \"CD161\"\n",
      "                 [,1]     [,2]\n",
      "Batch1-ID2  1.2456451 5.246698\n",
      "Batch1-ID27 1.0007316 4.593460\n",
      "Batch2-1    1.1639998 5.532452\n",
      "Batch2-2    0.9598868 4.470966\n",
      "Progress:  Each dot is a curve\n",
      "....\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“You're computing too large a percentage of total singular values, use a standard svd instead.”\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“did not converge--results might be invalid!; try increasing work or maxit”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "PC_ 1 \n",
      "Positive:  CD45RA, CD11c, HLA-DR, Tim3, CD56, CD278, GITR, CD137, CD62L, CD134 \n",
      "Negative:  CD28, CD127, CD27, CD161, CD196, CXCR6, CD25, CCR7, CD279, CD183 \n",
      "PC_ 2 \n",
      "Positive:  CD161, CD56, CD196, CXCR6, CD45RA, CD11c, GITR, CD127, Tim3, CD134 \n",
      "Negative:  CD272, CD183, CCR7, CD279, CD278, CD27, HLA-DR, CD25, CD137, CD62L \n",
      "PC_ 3 \n",
      "Positive:  CD279, HLA-DR, CD278, CD137, CXCR6, CD161, Tim3, CD196, CD27, CD28 \n",
      "Negative:  CCR7, CD25, CD62L, CD183, CD56, CD127, CD45RA, CD272, GITR, CD11c \n",
      "PC_ 4 \n",
      "Positive:  CD62L, Tim3, CD27, CD11c, CD56, CD28, CD279, CCR7, GITR, HLA-DR \n",
      "Negative:  CD272, CD45RA, CD196, CD127, CD278, CD161, CXCR6, CD183, CD25, CD134 \n",
      "PC_ 5 \n",
      "Positive:  CD134, GITR, CD25, CD137, CD278, CD196, Tim3, CD279, CD161, CD272 \n",
      "Negative:  CD183, CCR7, CD56, HLA-DR, CD11c, CD45RA, CXCR6, CD62L, CD27, CD127 \n",
      "\n",
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: PC_), setting key to 'apca_'”\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 53194\n",
      "Number of edges: 1168338\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9273\n",
      "Number of communities: 108\n",
      "Elapsed time: 69 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "97 singletons identified. 11 final clusters.\n",
      "\n",
      "Warning message:\n",
      "“The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric\n",
      "To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'\n",
      "This message will be shown once per session”\n",
      "09:11:28 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "09:11:28 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "09:11:30 Initializing from normalized Laplacian + noise\n",
      "\n",
      "09:11:31 Commencing optimization for 200 epochs, with 1703272 positive edges\n",
      "\n",
      "09:11:53 Optimization finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Input layer\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "sample.name <- \"COVID_ADTnorm_RPCA.wnn\"\n",
    "dir.name=\"Seurat_plots\"\n",
    "resol <- 0.25 # resolution for clustering\n",
    "genes_CLR <- c(\"CCR7\", \"CD11c\", \"CD25\", \"CD45RA\", \"CD127\",\n",
    "              \"CD134\", \"CD137\", \"CD183\", \"CD196\", \"CD272\",\n",
    "              \"CD278\", \"CD279\", \"CXCR6\", \"GITR\", \"HLA-DR\", \"Tim3\")\n",
    "genes_ADT <- c(\"CD27\", \"CD28\", \"CD56\", \"CD62L\", \"CD161\")\n",
    "\n",
    "########################## Processing layer ###############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name)\n",
    "\n",
    "###Correct batch effect in mRNA using RPCA\n",
    "#Batch1 dataset\n",
    "seu.T <- subset(seu, subset = batch %in% c(\"Batch1-ID2\", \"Batch1-ID27\"))\n",
    "DefaultAssay(seu.T) <- 'RNA'\n",
    "seu.T <- NormalizeData(seu.T, scale.factor=1000000)\n",
    "VariableFeatures(seu.T) <- rownames(seu.T[[\"RNA\"]])\n",
    "#Batch2 dataset\n",
    "seu.L <- subset(seu, subset = batch %in% c(\"Batch2-1\", \"Batch2-2\"))\n",
    "DefaultAssay(seu.L) <- 'RNA'\n",
    "seu.L <- NormalizeData(seu.L, scale.factor=1000000)\n",
    "VariableFeatures(seu.L) <- rownames(seu.L[[\"RNA\"]])\n",
    "seu.list <- list(seu.T, seu.L)\n",
    "\n",
    "# select features that are repeatedly variable across datasets for integration run PCA on each\n",
    "# dataset using these features\n",
    "features <- SelectIntegrationFeatures(object.list = seu.list)\n",
    "features\n",
    "seu.list <- lapply(X = seu.list, FUN = function(x) {\n",
    "    x <- ScaleData(x, features = features, verbose = FALSE)\n",
    "    x <- RunPCA(x, features = features, verbose = FALSE)\n",
    "})\n",
    "\n",
    "###Perform integration\n",
    "immune.anchors <- FindIntegrationAnchors(object.list = seu.list,\n",
    "                                         anchor.features = features,\n",
    "                                         reduction = \"rpca\",\n",
    "                                         max.features = 50)\n",
    "# this command creates an 'RNA.RPCA' data assay\n",
    "seu <- IntegrateData(anchorset = immune.anchors,\n",
    "                     new.assay.name = \"RNA.RPCA\")\n",
    "DefaultAssay(seu) <- \"RNA.RPCA\"\n",
    "seu <- ScaleData(seu, verbose = FALSE)\n",
    "#Perform PCA\n",
    "VariableFeatures(seu) <- rownames(seu[[\"RNA.RPCA\"]])\n",
    "seu = RunPCA(object = seu, npcs = 50)\n",
    "\n",
    "#JackStraw\n",
    "seu = JackStraw(object = seu, reduction = \"pca\", num.replicate = 100, dims = 50)\n",
    "seu <- ScoreJackStraw(object = seu, reduction = \"pca\", dims = 1:50, score.thresh = 0.05)\n",
    "file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Jackstraw.png\", sep='.')\n",
    "png(file.name, width = 1250, height = 500)\n",
    "JackStrawPlot(object = seu, dims = 1:50)\n",
    "dev.off()\n",
    "#Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "#Extract PCs which fulfill the pvalue threshold\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "dims\n",
    "\n",
    "###Perform ADTnorm\n",
    "\n",
    "#Exclude genes unrelated with CD8 function\n",
    "#Excluded: CD14, CD16, CD19, CD3, CD4, CD8, CXCR5, IgD, IgM\n",
    "genes_analyze <- c(genes_CLR, genes_ADT)\n",
    "counts <- GetAssayData(seu, slot=\"counts\", assay=\"ADT\")\n",
    "#str(counts)\n",
    "counts.sub <- counts[genes_analyze,]\n",
    "#str(counts.sub)\n",
    "adt_assay <- CreateAssayObject(counts = counts.sub)\n",
    "seu[[\"ADT\"]] <- adt_assay\n",
    "\n",
    "##Process CLRnormalization\n",
    "DefaultAssay(seu) <- \"ADT\"\n",
    "seu <- NormalizeData(seu, normalization.method = \"CLR\", margin = 2, assay = \"ADT\")\n",
    "\n",
    "##Process ADTnorm\n",
    "#call coutn data and batch information\n",
    "cell_x_adt <- as.data.frame(t(as.matrix(seu@assays$ADT@counts)))\n",
    "cell_x_adt <- dplyr::select(cell_x_adt, genes_ADT)\n",
    "cell_x_feature <- seu@meta.data\n",
    "cell_x_feature$sample = factor(cell_x_feature$batch)\n",
    "cell_x_feature$batch = factor(cell_x_feature$batch)\n",
    "#Process\n",
    "cell_x_adt_norm = ADTnorm(\n",
    "    cell_x_adt = cell_x_adt, \n",
    "    cell_x_feature = cell_x_feature, \n",
    "    save_outpath = dir.name, \n",
    "    study_name = \"ADTnorm\",\n",
    "    peak_type = \"mode\",\n",
    "    marker_to_process = NULL, ## setting it to NULL by default will process all available markers in cell_x_adt.\n",
    "    bimodal_marker = NULL, ## setting it to NULL will trigger ADTnorm to try different settings to find biomodal peaks for all the markers.\n",
    "    brewer_palettes = \"Dark2\", ## colorbrewer palettes setting for the density plot\n",
    "    save_intermediate_fig = TRUE\n",
    ")\n",
    "\n",
    "##Merge CLR and ADTnorm\n",
    "aft.norm <- as.data.frame(t(as.matrix(seu@assays$ADT@data)))\n",
    "aft.norm <- dplyr::select(aft.norm, -genes_ADT)\n",
    "aft.norm <- cbind(aft.norm, cell_x_adt_norm)\n",
    "seu@assays$ADT@data <- t(aft.norm)\n",
    "\n",
    "##Perform wnn analysis\n",
    "##For Abseq\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Normalizing and scaling data\n",
    "VariableFeatures(seu) <- rownames(seu[[\"ADT\"]])\n",
    "seu = ScaleData(object = seu) \n",
    "#Perform PCA\n",
    "seu = RunPCA(object = seu, reduction.name = 'apca')\n",
    "    \n",
    "#Clustering Weighted Nearest Neighbor\n",
    "seu <- FindMultiModalNeighbors(seu, reduction.list = list(\"pca\", \"apca\"), \n",
    "                                dims.list = list(dims, dims))\n",
    "seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resol)\n",
    "\n",
    "seu <- RunUMAP(seu, nn.name = \"weighted.nn\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "#Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"res\", resol, \"rda\", sep='.')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33a1da0",
   "metadata": {},
   "source": [
    "# Check the result of 1st Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "be5aff2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: DimPlot for cluster distribution\n",
    "##Input: seurat object\n",
    "##Output: DimplotStraw Plot\n",
    "\n",
    "DimOrigin <- function(seu, dir.name, sample.name, red.use, resol){\n",
    "    p1 = DimPlot(object = seu, reduction = red.use, label = TRUE, label.size = 10, pt.size = 0.5) +\n",
    "      theme(axis.title.x = element_text(size=10, family = \"Arial\"), \n",
    "            axis.title.y = element_text(size=10, family = \"Arial\"), \n",
    "            axis.text.x = element_text(size=10, colour = 1, family = \"Arial\"), \n",
    "            axis.text.y = element_text(size = 10, colour = 1, family = \"Arial\")) +\n",
    "      theme(panel.border = element_rect(fill = NA, size = 1)) \n",
    "\n",
    "    legend1 <- cowplot::get_legend(p1)\n",
    "    p1 = p1 + theme(legend.position = 'none')\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(red.use, \"reso\", resol, \"png\", sep='.')\n",
    "    save_plot(file = file.name, plot_grid(p1, legend1, ncol=2, nrow=1), device=\"png\", \n",
    "              units=\"in\", dpi = 600, base_width = 10, base_height = 5, limitsize=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e65526ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: Find marker genes with marker heatmap\n",
    "##Input: seurat object\n",
    "##Output: Marker gene table and marker gene heatmap plot\n",
    "\n",
    "MarkerHeatmap <- function(seu, dir.name, sample.name, resol){\n",
    "    seu.markers = FindAllMarkers(seu, verbose = TRUE, test.use=\"wilcox\", only.pos=TRUE, min.pct=0.1, features.use = NULL, return.thresh=0.05)\n",
    "\n",
    "    #Create heatmap with top10 marker genes\n",
    "    top10 = seu.markers %>% group_by(cluster) %>% top_n(10, avg_log2FC)\n",
    "    top10 = as.data.frame(top10)\n",
    "    top10  = top10 [!duplicated(top10$gene),]\n",
    "    top10 = top10 %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    top10 = as.data.frame(top10)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"marker_res\", resol, \"png\", sep='.')\n",
    "    p <- DoHeatmap(seu, features = top10$gene, disp.min = -2.5, disp.max = 2.5, size = 8)\n",
    "    ggsave(file = file.name, plot = p, device=\"png\", units=\"in\", dpi = 300,\n",
    "           width = 20, height = 20, limitsize=FALSE)\n",
    "\n",
    "    seu.markers$cluster = as.numeric(seu.markers$cluster)\n",
    "    seu.markers = seu.markers %>% arrange(desc(avg_log2FC))  %>% arrange(cluster)\n",
    "    seu.markers = as.data.frame(seu.markers)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"ALLmarkers_minpct0.1_Adj_p0.05.txt\", sep='')\n",
    "    fwrite(seu.markers, file.name, row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "    file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"ALLmarkers_minpct0.1_Adj_p0.05.rda\", sep='')\n",
    "    save(seu.markers, file=file.name)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7b0cb632",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Plot clones in dimentional reduction plot (DimPlot)\n",
    "#seu: Seurat object / tmp2: identifier of plotting cells\n",
    "\n",
    "PlotClones = function(seu, tmp2, name.output, meta.data, reduction, width, height, size){\n",
    "    #Exctract names for cells in plot\n",
    "    tmp_raw <- as.vector(row.names(meta.data))\n",
    "    #Extract colors for plot\n",
    "    p2 = Seurat::DimPlot(object = seu, reduction = reduction, pt.size = 0.1)\n",
    "    orig_ident_build = ggplot2::ggplot_build(p2)\n",
    "    orig_ident_build = orig_ident_build$data[[1]]\n",
    "    orig_ident_build =  orig_ident_build[order(orig_ident_build$group), ]\n",
    "    ident.cols = unique(orig_ident_build$colour) # Get a vector of unique colors\n",
    "    names(ident.cols)=c(0:(max(as.numeric(seu@active.ident)-1)))\n",
    "\n",
    "    #Extract x- and y- axis information for DimPlot\n",
    "    xmin <- min(orig_ident_build$x)-0.5\n",
    "    xmax <- max(orig_ident_build$x)+0.5\n",
    "    ymin <- min(orig_ident_build$y)-0.5\n",
    "    ymax <- max(orig_ident_build$y)+0.5\n",
    "    x_label_min <- floor(xmin/10)*10\n",
    "    x_label_max <- ceiling(xmax/10)*10\n",
    "    y_label_min <- floor(ymin/10)*10\n",
    "    y_label_max <- ceiling(ymax/10)*10\n",
    "\n",
    "    #Dimplot\n",
    "    p1 <- DimPlot(object = seu, label = FALSE,\n",
    "                    cells = tmp2,\n",
    "                    reduction = reduction,\n",
    "                    cols = ident.cols,\n",
    "                    pt.size = size) + \n",
    "        NoLegend() +\n",
    "        scale_x_continuous(limits = c(xmin,xmax), breaks= seq(x_label_min,x_label_max,10)) +\n",
    "        scale_y_continuous(limits = c(ymin,ymax), breaks= seq(y_label_min,y_label_max,10)) +\n",
    "        theme(axis.title = element_blank(), \n",
    "              axis.text = element_blank(),\n",
    "             axis.line = element_line(size = 0.2))\n",
    "    \n",
    "      save_plot(file = name.output, plot_grid(p1, ncol=1, nrow=1), device=\"tiff\", \n",
    "                units=\"in\", dpi = 300, base_width = width, base_height = height, limitsize=FALSE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5b684c2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Plot bar graph for cluster distribution\n",
    "Cluster_bar <- function(seu, red.use, data, dir.name, file.name, width, height){\n",
    "    #Extract colors for plot\n",
    "    p2 = Seurat::DimPlot(object = seu, reduction = red.use, pt.size = 0.1)\n",
    "    orig_ident_build = ggplot2::ggplot_build(p2)\n",
    "    orig_ident_build = orig_ident_build$data[[1]]\n",
    "    orig_ident_build =  orig_ident_build[order(orig_ident_build$group), ]\n",
    "    ident.cols = unique(orig_ident_build$colour) # Get a vector of unique colors\n",
    "    names(ident.cols)=c(0:(max(as.numeric(seu@active.ident)-1)))\n",
    "\n",
    "    #Plot bar graph\n",
    "    ppi <- 600\n",
    "     g <- ggplot(data.table, aes(x = Treat, y = Number, fill = clust)) +\n",
    "        scale_fill_manual(values=ident.cols) +\n",
    "        geom_bar(stat = \"identity\", position = \"fill\") +\n",
    "        labs(y=\"Prop. of cluster\") +\n",
    "        scale_y_continuous(labels = percent) +\n",
    "        theme_classic() + \n",
    "        theme(plot.title = element_blank(), \n",
    "              panel.background = element_rect(fill = \"transparent\",color = NA),\n",
    "              plot.background = element_rect(fill = \"transparent\",color = NA),\n",
    "                axis.title.x = element_blank(), \n",
    "                axis.title.y = element_text(size=8, colour = 1, family = \"Arial\"), \n",
    "                axis.text.x = element_text(size = 6, colour = 1, family = \"Arial\", angle = 45), \n",
    "                axis.text.y = element_text(size = 6, colour = 1, family = \"Arial\")) +\n",
    "        guides(fill = \"none\") \n",
    "    save_plot(file = file.name, plot(g), device=\"tiff\", \n",
    "              units=\"in\", dpi = 600, base_width = width, base_height = height, limitsize=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ae0cd58e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Function: Create Scatter and Violin plot for selected genes\n",
    "##Input: seurat object, gene list\n",
    "##Output: Scatter plot, Violin plot\n",
    "\n",
    "ScatterViolin <- function(seu, dir.name, sample.name, red.use, tmps, tmp_names){\n",
    "    for (i in 1:length(tmp_names)){\n",
    "        #Call gene list\n",
    "        tmp <- tmps[[i]]\n",
    "        tmp_name <- tmp_names[i]\n",
    "  \n",
    "        #Scatter Plot\n",
    "        file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Scatter\", tmp_name, \"png\", sep='.')\n",
    "        png(file.name, width = 1536, height = 240)\n",
    "        p <- FeaturePlot(seu, features = tmps, ncol = 6, order = TRUE,\n",
    "                        reduction = red.use, dims=c(1,2), cols = c(\"grey\", \"red\"), pt.size = 0.2)\n",
    "        plot(p)\n",
    "        dev.off()\n",
    "        \n",
    "        #Violin Plot\n",
    "        file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Violin\", tmp_name, \"png\", sep='.')\n",
    "        png(file.name, width = 4608, height = 720)\n",
    "        p <- VlnPlot(seu, features = tmps, ncol = 6, pt.size = 0.1)\n",
    "        plot(p)\n",
    "        dev.off()\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "93eebfd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/Seurat_plots' already exists”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use the `linewidth` argument instead.”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)):\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Create Dim Plot\n",
    "##Check lineage marker gene expression\n",
    "\n",
    "###Input layer\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "sample.name <- \"COVID.ADTnorm_RPCA.wnn\"\n",
    "dir.name <- \"Seurat_plots\"\n",
    "red.use <- \"wnn.umap\"\n",
    "resol <- 0.25 # resolution for clustering\n",
    "tmp_rna = c(\"TRBC2\", \"TRDC\", \"CD14\", \"MS4A1\", \"KLRB1\", \"ZBTB16\")\n",
    "tmp_adt = c(\"CD62L\", \"CD45RA\", \"CD56\", \"CD196\", \"CD161\", \"CXCR6\")\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name)\n",
    "\n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "\n",
    "###RNA\n",
    "DefaultAssay(seu) <- 'RNA.RPCA'\n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"RNA\"), resol)\n",
    "##Check lineage marker gene expression\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"RNA\"), red.use, tmp_rna, \"lineage\")\n",
    "\n",
    "###ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"ADT\"), resol)\n",
    "##Check lineage marker gene expression\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"ADT\"), red.use, tmp_adt, \"lineage\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a967de",
   "metadata": {},
   "source": [
    "# 2nd analysis with removing gd T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e0c1b6f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'Seurat' [package \"SeuratObject\"] with 13 slots\n",
      "  ..@ assays      :List of 3\n",
      "  .. ..$ RNA     :Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:3185490] 7 11 12 13 17 22 23 27 37 38 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:51854] 0 75 134 207 277 319 375 431 475 557 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 368 51853\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:3185490] 353 361 111 1 252 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:3185490] 7 11 12 13 17 22 23 27 37 38 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:51854] 0 75 134 207 277 319 375 431 475 557 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 368 51853\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:3185490] 9.13 9.15 7.97 3.3 8.79 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data   : num[0 , 0 ] \n",
      "  .. .. .. ..@ key          : chr \"rna_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : chr(0) \n",
      "  .. .. .. ..@ meta.features:'data.frame':\t368 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : list()\n",
      "  .. ..$ ADT     :Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:806541] 0 1 2 3 4 5 6 7 8 9 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:51854] 0 21 36 52 71 88 105 124 136 155 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 21 51853\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:21] \"CCR7\" \"CD11c\" \"CD25\" \"CD45RA\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:806541] 4 9 11 3259 37 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         : num [1:21, 1:51853] 0.158 0.327 0.387 4.948 0.951 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:21] \"CCR7\" \"CD11c\" \"CD25\" \"CD45RA\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. ..@ scale.data   : num [1:21, 1:51853] -0.245 -0.528 0.472 1.234 0.075 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:21] \"CCR7\" \"CD11c\" \"CD25\" \"CD45RA\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. ..@ key          : chr \"adt_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : chr [1:21] \"CCR7\" \"CD11c\" \"CD25\" \"CD45RA\" ...\n",
      "  .. .. .. ..@ meta.features:'data.frame':\t21 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : list()\n",
      "  .. ..$ RNA.RPCA:Formal class 'Assay' [package \"SeuratObject\"] with 8 slots\n",
      "  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int(0) \n",
      "  .. .. .. .. .. ..@ p       : int 0\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 0 0\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. .. ..$ : NULL\n",
      "  .. .. .. .. .. ..@ x       : num(0) \n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package \"Matrix\"] with 6 slots\n",
      "  .. .. .. .. .. ..@ i       : int [1:5303363] 0 5 6 7 10 11 12 13 15 16 ...\n",
      "  .. .. .. .. .. ..@ p       : int [1:51854] 0 239 473 716 963 1176 1402 1638 1859 2122 ...\n",
      "  .. .. .. .. .. ..@ Dim     : int [1:2] 368 51853\n",
      "  .. .. .. .. .. ..@ Dimnames:List of 2\n",
      "  .. .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..@ x       : num [1:5303363] -0.251 0.954 -0.106 9.171 -0.33 ...\n",
      "  .. .. .. .. .. ..@ factors : list()\n",
      "  .. .. .. ..@ scale.data   : num [1:368, 1:51853] -0.64417 -0.03503 -0.025 -0.07869 0.00447 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. ..@ key          : chr \"rnarpca_\"\n",
      "  .. .. .. ..@ assay.orig   : NULL\n",
      "  .. .. .. ..@ var.features : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. ..@ meta.features:'data.frame':\t368 obs. of  0 variables\n",
      "  .. .. .. ..@ misc         : NULL\n",
      "  ..@ meta.data   :'data.frame':\t51853 obs. of  11 variables:\n",
      "  .. ..$ orig.ident     : chr [1:51853] \"TP3_ID2\" \"TP3_ID2\" \"TP3_ID2\" \"TP3_ID2\" ...\n",
      "  .. ..$ nCount_RNA     : num [1:51853] 38417 13181 21478 30520 6272 ...\n",
      "  .. ..$ nFeature_RNA   : int [1:51853] 75 59 73 70 42 56 56 44 82 89 ...\n",
      "  .. ..$ nCount_ADT     : num [1:51853] 4391 1798 3435 2204 1788 ...\n",
      "  .. ..$ nFeature_ADT   : int [1:51853] 21 15 16 19 17 17 19 12 19 21 ...\n",
      "  .. ..$ batch          : chr [1:51853] \"Batch1-ID2\" \"Batch1-ID2\" \"Batch1-ID2\" \"Batch1-ID2\" ...\n",
      "  .. ..$ BC             : chr [1:51853] \"28\" \"80\" \"185\" \"229\" ...\n",
      "  .. ..$ RNA.RPCA.weight: num [1:51853] 0.517 0.604 0.745 0.382 0.704 ...\n",
      "  .. ..$ ADT.weight     : num [1:51853] 0.483 0.396 0.255 0.618 0.296 ...\n",
      "  .. ..$ wsnn_res.0.25  : Factor w/ 11 levels \"0\",\"1\",\"10\",\"2\",..: 8 5 2 4 4 5 5 4 7 7 ...\n",
      "  .. ..$ seurat_clusters: Factor w/ 11 levels \"0\",\"1\",\"2\",\"3\",..: 7 4 2 3 3 4 4 3 6 6 ...\n",
      "  ..@ active.assay: chr \"ADT\"\n",
      "  ..@ active.ident: Factor w/ 10 levels \"0\",\"1\",\"2\",\"3\",..: 7 4 2 3 3 4 4 3 6 6 ...\n",
      "  .. ..- attr(*, \"names\")= chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  ..@ graphs      : list()\n",
      "  ..@ neighbors   : list()\n",
      "  ..@ reductions  :List of 3\n",
      "  .. ..$ pca     :Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:51853, 1:50] 2.45 -1.93 1.88 -3.77 -1.98 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings          : num [1:364, 1:50] 0.03507 -0.00532 -0.00424 -0.00164 -0.00427 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:364] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:50] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ global                    : logi FALSE\n",
      "  .. .. .. ..@ stdev                     : num [1:50] 2.94 2.57 1.91 1.74 1.59 ...\n",
      "  .. .. .. ..@ key                       : chr \"PC_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      :List of 1\n",
      "  .. .. .. .. ..$ total.variance: num 254\n",
      "  .. ..$ apca    :Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:51853, 1:20] 1.49 -1.42 2.54 -3.51 -4.77 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:20] \"apca_1\" \"apca_2\" \"apca_3\" \"apca_4\" ...\n",
      "  .. .. .. ..@ feature.loadings          : num [1:21, 1:20] -0.147 0.255 -0.159 0.345 -0.355 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:21] \"CCR7\" \"CD11c\" \"CD25\" \"CD45RA\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:20] \"PC_1\" \"PC_2\" \"PC_3\" \"PC_4\" ...\n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"ADT\"\n",
      "  .. .. .. ..@ global                    : logi FALSE\n",
      "  .. .. .. ..@ stdev                     : num [1:20] 2.19 1.65 1.37 1.16 1.04 ...\n",
      "  .. .. .. ..@ key                       : chr \"apca_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      :List of 1\n",
      "  .. .. .. .. ..$ total.variance: num 21\n",
      "  .. ..$ wnn.umap:Formal class 'DimReduc' [package \"SeuratObject\"] with 9 slots\n",
      "  .. .. .. ..@ cell.embeddings           : num [1:51853, 1:2] -1.137 0.201 -2.542 5.735 7.257 ...\n",
      "  .. .. .. .. ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. .. .. .. .. ..$ : chr [1:51853] \"Batch1-ID2_28\" \"Batch1-ID2_80\" \"Batch1-ID2_185\" \"Batch1-ID2_229\" ...\n",
      "  .. .. .. .. .. ..$ : chr [1:2] \"wnnUMAP_1\" \"wnnUMAP_2\"\n",
      "  .. .. .. ..@ feature.loadings          : num[0 , 0 ] \n",
      "  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] \n",
      "  .. .. .. ..@ assay.used                : chr \"ADT\"\n",
      "  .. .. .. ..@ global                    : logi TRUE\n",
      "  .. .. .. ..@ stdev                     : num(0) \n",
      "  .. .. .. ..@ key                       : chr \"wnnUMAP_\"\n",
      "  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package \"SeuratObject\"] with 4 slots\n",
      "  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] \n",
      "  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] \n",
      "  .. .. .. ..@ misc                      : list()\n",
      "  ..@ images      : list()\n",
      "  ..@ project.name: chr \"SeuratProject\"\n",
      "  ..@ misc        : list()\n",
      "  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1\n",
      "  .. ..$ : int [1:3] 4 0 4\n",
      "  ..@ commands    :List of 13\n",
      "  .. ..$ FindIntegrationAnchors :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindIntegrationAnchors\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:30:45\"\n",
      "  .. .. .. ..@ assay.used : chr [1:2] \"RNA\" \"RNA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"FindIntegrationAnchors(object.list = seu.list, anchor.features = features, \" \"    reduction = \\\"rpca\\\", max.features = 50)\"\n",
      "  .. .. .. ..@ params     :List of 15\n",
      "  .. .. .. .. ..$ assay               : chr [1:2] \"RNA\" \"RNA\"\n",
      "  .. .. .. .. ..$ anchor.features     : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. ..$ scale               : logi TRUE\n",
      "  .. .. .. .. ..$ normalization.method: chr \"LogNormalize\"\n",
      "  .. .. .. .. ..$ reduction           : chr \"pca\"\n",
      "  .. .. .. .. ..$ l2.norm             : logi TRUE\n",
      "  .. .. .. .. ..$ dims                : int [1:30] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ k.anchor            : num 5\n",
      "  .. .. .. .. ..$ k.filter            : logi NA\n",
      "  .. .. .. .. ..$ k.score             : num 30\n",
      "  .. .. .. .. ..$ max.features        : num 50\n",
      "  .. .. .. .. ..$ nn.method           : chr \"annoy\"\n",
      "  .. .. .. .. ..$ n.trees             : num 50\n",
      "  .. .. .. .. ..$ eps                 : num 0\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ withCallingHandlers    :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"withCallingHandlers\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:31:07\"\n",
      "  .. .. .. ..@ assay.used : NULL\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"withCallingHandlers(expr, warning = function(w) if (inherits(w, \" \"    classes)) tryInvokeRestart(\\\"muffleWarning\\\"))\"\n",
      "  .. .. .. ..@ params     :List of 9\n",
      "  .. .. .. .. ..$ new.assay.name      : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ normalization.method: chr \"LogNormalize\"\n",
      "  .. .. .. .. ..$ features            : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. ..$ dims                : int [1:30] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ k.weight            : num 100\n",
      "  .. .. .. .. ..$ sd.weight           : num 1\n",
      "  .. .. .. .. ..$ preserve.order      : logi FALSE\n",
      "  .. .. .. .. ..$ eps                 : num 0\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ ScaleData.RNA.RPCA     :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"ScaleData.RNA.RPCA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:31:07\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ call.string: chr \"ScaleData(seu, verbose = FALSE)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ features          : chr [1:368] \"ADA\" \"ADGRE1\" \"ADGRG3\" \"AIM2\" ...\n",
      "  .. .. .. .. ..$ assay             : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ model.use         : chr \"linear\"\n",
      "  .. .. .. .. ..$ use.umi           : logi FALSE\n",
      "  .. .. .. .. ..$ do.scale          : logi TRUE\n",
      "  .. .. .. .. ..$ do.center         : logi TRUE\n",
      "  .. .. .. .. ..$ scale.max         : num 10\n",
      "  .. .. .. .. ..$ block.size        : num 1000\n",
      "  .. .. .. .. ..$ min.cells.to.block: num 3000\n",
      "  .. .. .. .. ..$ verbose           : logi FALSE\n",
      "  .. ..$ RunPCA.RNA.RPCA        :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunPCA.RNA.RPCA\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 08:31:32\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ call.string: chr \"RunPCA(object = seu, npcs = 50)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ assay          : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ npcs           : num 50\n",
      "  .. .. .. .. ..$ rev.pca        : logi FALSE\n",
      "  .. .. .. .. ..$ weight.by.var  : logi TRUE\n",
      "  .. .. .. .. ..$ verbose        : logi TRUE\n",
      "  .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5\n",
      "  .. .. .. .. ..$ nfeatures.print: num 30\n",
      "  .. .. .. .. ..$ reduction.name : chr \"pca\"\n",
      "  .. .. .. .. ..$ reduction.key  : chr \"PC_\"\n",
      "  .. .. .. .. ..$ seed.use       : num 42\n",
      "  .. ..$ JackStraw.RNA.RPCA.pca :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"JackStraw.RNA.RPCA.pca\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:07:23\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"JackStraw(object = seu, reduction = \\\"pca\\\", num.replicate = 100, \" \"    dims = 50)\"\n",
      "  .. .. .. ..@ params     :List of 7\n",
      "  .. .. .. .. ..$ reduction    : chr \"pca\"\n",
      "  .. .. .. .. ..$ assay        : chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..$ dims         : num 50\n",
      "  .. .. .. .. ..$ num.replicate: num 100\n",
      "  .. .. .. .. ..$ prop.freq    : num 0.01\n",
      "  .. .. .. .. ..$ verbose      : logi TRUE\n",
      "  .. .. .. .. ..$ maxit        : num 1000\n",
      "  .. ..$ ScoreJackStraw         :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"ScoreJackStraw\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:07:23\"\n",
      "  .. .. .. ..@ assay.used : chr \"RNA.RPCA\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"ScoreJackStraw(object = seu, reduction = \\\"pca\\\", dims = 1:50, \" \"    score.thresh = 0.05)\"\n",
      "  .. .. .. ..@ params     :List of 4\n",
      "  .. .. .. .. ..$ reduction   : chr \"pca\"\n",
      "  .. .. .. .. ..$ dims        : int [1:50] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ score.thresh: num 0.05\n",
      "  .. .. .. .. ..$ do.plot     : logi FALSE\n",
      "  .. ..$ NormalizeData.ADT      :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"NormalizeData.ADT\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:07:25\"\n",
      "  .. .. .. ..@ assay.used : chr \"ADT\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"NormalizeData(seu, normalization.method = \\\"CLR\\\", margin = 2, \" \"    assay = \\\"ADT\\\")\"\n",
      "  .. .. .. ..@ params     :List of 5\n",
      "  .. .. .. .. ..$ assay               : chr \"ADT\"\n",
      "  .. .. .. .. ..$ normalization.method: chr \"CLR\"\n",
      "  .. .. .. .. ..$ scale.factor        : num 10000\n",
      "  .. .. .. .. ..$ margin              : num 2\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ ScaleData.ADT          :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"ScaleData.ADT\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:07:41\"\n",
      "  .. .. .. ..@ assay.used : chr \"ADT\"\n",
      "  .. .. .. ..@ call.string: chr \"ScaleData(object = seu)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ features          : chr [1:21] \"CCR7\" \"CD11c\" \"CD25\" \"CD45RA\" ...\n",
      "  .. .. .. .. ..$ assay             : chr \"ADT\"\n",
      "  .. .. .. .. ..$ model.use         : chr \"linear\"\n",
      "  .. .. .. .. ..$ use.umi           : logi FALSE\n",
      "  .. .. .. .. ..$ do.scale          : logi TRUE\n",
      "  .. .. .. .. ..$ do.center         : logi TRUE\n",
      "  .. .. .. .. ..$ scale.max         : num 10\n",
      "  .. .. .. .. ..$ block.size        : num 1000\n",
      "  .. .. .. .. ..$ min.cells.to.block: num 3000\n",
      "  .. .. .. .. ..$ verbose           : logi TRUE\n",
      "  .. ..$ RunPCA.ADT             :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunPCA.ADT\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:07:55\"\n",
      "  .. .. .. ..@ assay.used : chr \"ADT\"\n",
      "  .. .. .. ..@ call.string: chr \"RunPCA(object = seu, reduction.name = \\\"apca\\\")\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ assay          : chr \"ADT\"\n",
      "  .. .. .. .. ..$ npcs           : num 50\n",
      "  .. .. .. .. ..$ rev.pca        : logi FALSE\n",
      "  .. .. .. .. ..$ weight.by.var  : logi TRUE\n",
      "  .. .. .. .. ..$ verbose        : logi TRUE\n",
      "  .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5\n",
      "  .. .. .. .. ..$ nfeatures.print: num 30\n",
      "  .. .. .. .. ..$ reduction.name : chr \"apca\"\n",
      "  .. .. .. .. ..$ reduction.key  : chr \"PC_\"\n",
      "  .. .. .. .. ..$ seed.use       : num 42\n",
      "  .. ..$ FindModalityWeights    :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindModalityWeights\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:08:27\"\n",
      "  .. .. .. ..@ assay.used : Named chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..- attr(*, \"names\")= chr \"pca\"\n",
      "  .. .. .. ..@ call.string: chr [1:4] \"FindModalityWeights(object = object, reduction.list = reduction.list, \" \"    dims.list = dims.list, k.nn = k.nn, sd.scale = sd.scale, \" \"    l2.norm = l2.norm, cross.contant.list = cross.contant.list, \" \"    smooth = smooth, verbose = verbose)\"\n",
      "  .. .. .. ..@ params     :List of 12\n",
      "  .. .. .. .. ..$ reduction.list    : Named chr [1:2] \"pca\" \"apca\"\n",
      "  .. .. .. .. .. ..- attr(*, \"names\")= chr [1:2] \"pca\" \"apca\"\n",
      "  .. .. .. .. ..$ dims.list         : Named int [1:30] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. .. ..- attr(*, \"names\")= chr [1:30] \"pca1\" \"pca2\" \"pca3\" \"pca4\" ...\n",
      "  .. .. .. .. ..$ k.nn              : num 20\n",
      "  .. .. .. .. ..$ snn.far.nn        : logi TRUE\n",
      "  .. .. .. .. ..$ s.nn              : num 20\n",
      "  .. .. .. .. ..$ prune.SNN         : num 0\n",
      "  .. .. .. .. ..$ l2.norm           : logi TRUE\n",
      "  .. .. .. .. ..$ sd.scale          : num 1\n",
      "  .. .. .. .. ..$ cross.contant.list: Named num [1:2] 1e-04 1e-04\n",
      "  .. .. .. .. .. ..- attr(*, \"names\")= chr [1:2] \"pca\" \"apca\"\n",
      "  .. .. .. .. ..$ sigma.idx         : num 20\n",
      "  .. .. .. .. ..$ smooth            : logi FALSE\n",
      "  .. .. .. .. ..$ verbose           : logi TRUE\n",
      "  .. ..$ FindMultiModalNeighbors:Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindMultiModalNeighbors\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:10:12\"\n",
      "  .. .. .. ..@ assay.used : Named chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..- attr(*, \"names\")= chr \"pca\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"FindMultiModalNeighbors(seu, reduction.list = list(\\\"pca\\\", \\\"apca\\\"), \" \"    dims.list = list(dims, dims))\"\n",
      "  .. .. .. ..@ params     :List of 15\n",
      "  .. .. .. .. ..$ reduction.list      :List of 2\n",
      "  .. .. .. .. .. ..$ : chr \"pca\"\n",
      "  .. .. .. .. .. ..$ : chr \"apca\"\n",
      "  .. .. .. .. ..$ dims.list           :List of 2\n",
      "  .. .. .. .. .. ..$ : int [1:15] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. .. ..$ : int [1:15] 1 2 3 4 5 6 7 8 9 10 ...\n",
      "  .. .. .. .. ..$ k.nn                : num 20\n",
      "  .. .. .. .. ..$ l2.norm             : logi TRUE\n",
      "  .. .. .. .. ..$ knn.graph.name      : chr \"wknn\"\n",
      "  .. .. .. .. ..$ snn.graph.name      : chr \"wsnn\"\n",
      "  .. .. .. .. ..$ weighted.nn.name    : chr \"weighted.nn\"\n",
      "  .. .. .. .. ..$ modality.weight.name: chr [1:2] \"RNA.RPCA.weight\" \"ADT.weight\"\n",
      "  .. .. .. .. ..$ knn.range           : num 200\n",
      "  .. .. .. .. ..$ prune.SNN           : num 0.0667\n",
      "  .. .. .. .. ..$ sd.scale            : num 1\n",
      "  .. .. .. .. ..$ cross.contant.list  :List of 2\n",
      "  .. .. .. .. .. ..$ : num 1e-04\n",
      "  .. .. .. .. .. ..$ : num 1e-04\n",
      "  .. .. .. .. ..$ smooth              : logi FALSE\n",
      "  .. .. .. .. ..$ return.intermediate : logi FALSE\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. ..$ FindClusters           :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"FindClusters\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:11:28\"\n",
      "  .. .. .. ..@ assay.used : Named chr \"RNA.RPCA\"\n",
      "  .. .. .. .. ..- attr(*, \"names\")= chr \"pca\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"FindClusters(object = seu, graph.name = \\\"wsnn\\\", algorithm = 3, \" \"    resolution = resol)\"\n",
      "  .. .. .. ..@ params     :List of 10\n",
      "  .. .. .. .. ..$ graph.name      : chr \"wsnn\"\n",
      "  .. .. .. .. ..$ modularity.fxn  : num 1\n",
      "  .. .. .. .. ..$ resolution      : num 0.25\n",
      "  .. .. .. .. ..$ method          : chr \"matrix\"\n",
      "  .. .. .. .. ..$ algorithm       : num 3\n",
      "  .. .. .. .. ..$ n.start         : num 10\n",
      "  .. .. .. .. ..$ n.iter          : num 10\n",
      "  .. .. .. .. ..$ random.seed     : num 0\n",
      "  .. .. .. .. ..$ group.singletons: logi TRUE\n",
      "  .. .. .. .. ..$ verbose         : logi TRUE\n",
      "  .. ..$ RunUMAP.ADT.pca        :Formal class 'SeuratCommand' [package \"SeuratObject\"] with 5 slots\n",
      "  .. .. .. ..@ name       : chr \"RunUMAP.ADT.pca\"\n",
      "  .. .. .. ..@ time.stamp : POSIXct[1:1], format: \"2023-04-02 09:11:53\"\n",
      "  .. .. .. ..@ assay.used : chr \"ADT\"\n",
      "  .. .. .. ..@ call.string: chr [1:2] \"RunUMAP(seu, nn.name = \\\"weighted.nn\\\", reduction.name = \\\"wnn.umap\\\", \" \"    reduction.key = \\\"wnnUMAP_\\\")\"\n",
      "  .. .. .. ..@ params     :List of 26\n",
      "  .. .. .. .. ..$ reduction           : chr \"pca\"\n",
      "  .. .. .. .. ..$ assay               : chr \"ADT\"\n",
      "  .. .. .. .. ..$ nn.name             : chr \"weighted.nn\"\n",
      "  .. .. .. .. ..$ slot                : chr \"data\"\n",
      "  .. .. .. .. ..$ umap.method         : chr \"uwot\"\n",
      "  .. .. .. .. ..$ return.model        : logi FALSE\n",
      "  .. .. .. .. ..$ n.neighbors         : int 30\n",
      "  .. .. .. .. ..$ n.components        : int 2\n",
      "  .. .. .. .. ..$ metric              : chr \"cosine\"\n",
      "  .. .. .. .. ..$ learning.rate       : num 1\n",
      "  .. .. .. .. ..$ min.dist            : num 0.3\n",
      "  .. .. .. .. ..$ spread              : num 1\n",
      "  .. .. .. .. ..$ set.op.mix.ratio    : num 1\n",
      "  .. .. .. .. ..$ local.connectivity  : int 1\n",
      "  .. .. .. .. ..$ repulsion.strength  : num 1\n",
      "  .. .. .. .. ..$ negative.sample.rate: int 5\n",
      "  .. .. .. .. ..$ uwot.sgd            : logi FALSE\n",
      "  .. .. .. .. ..$ seed.use            : int 42\n",
      "  .. .. .. .. ..$ angular.rp.forest   : logi FALSE\n",
      "  .. .. .. .. ..$ densmap             : logi FALSE\n",
      "  .. .. .. .. ..$ dens.lambda         : num 2\n",
      "  .. .. .. .. ..$ dens.frac           : num 0.3\n",
      "  .. .. .. .. ..$ dens.var.shift      : num 0.1\n",
      "  .. .. .. .. ..$ verbose             : logi TRUE\n",
      "  .. .. .. .. ..$ reduction.name      : chr \"wnn.umap\"\n",
      "  .. .. .. .. ..$ reduction.key       : chr \"wnnUMAP_\"\n",
      "  ..@ tools       :List of 1\n",
      "  .. ..$ Integration:Formal class 'IntegrationData' [package \"Seurat\"] with 7 slots\n",
      "  .. .. .. ..@ neighbors         : NULL\n",
      "  .. .. .. ..@ weights           : NULL\n",
      "  .. .. .. ..@ integration.matrix: NULL\n",
      "  .. .. .. ..@ anchors           :'data.frame':\t9116 obs. of  5 variables:\n",
      "  .. .. .. .. ..$ cell1   : num [1:9116] 1 12 15 19 21 25 30 31 33 41 ...\n",
      "  .. .. .. .. ..$ cell2   : num [1:9116] 23464 21622 38181 16222 1807 ...\n",
      "  .. .. .. .. ..$ score   : num [1:9116] 0.25 0.438 0.562 0.875 0.688 ...\n",
      "  .. .. .. .. ..$ dataset1: int [1:9116] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..$ dataset2: int [1:9116] 2 2 2 2 2 2 2 2 2 2 ...\n",
      "  .. .. .. ..@ offsets           : NULL\n",
      "  .. .. .. ..@ objects.ncell     : NULL\n",
      "  .. .. .. ..@ sample.tree       : num [1, 1:2] -2 -1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 4 features requested have zero variance (running reduction without them): AQP9, CCL17, IL33, EPX”\n",
      "PC_ 1 \n",
      "Positive:  GZMH, GZMB, FCGR3A, GNLY, NKG7, LGALS1, CX3CR1, PRF1, CST7, ZNF683 \n",
      "\t   ITGAM, ITGB2, CCL5, IFNG, GZMA, APOBEC3G, B3GAT1, CD63, RUNX3, CD300A \n",
      "\t   KLRC3, LAIR2, CTSW, KIR2DL1, CD8B, CTSD, KLRK1, CD8A, CD3G, CCL4 \n",
      "Negative:  IL7R, LTB, GZMK, TCF7, KLRB1, TNFRSF25, CD27, MYC, ZBTB16, RORC \n",
      "\t   TRAT1, CCR7, DPP4, JUNB, DUSP2, PIK3IP1, HLA-A, PASK, IL18RAP, CCR2 \n",
      "\t   CXCR4, DUSP1, CD69, CXCR6, LEF1, NCR3, NT5E, CD28, BIRC3, BTG1 \n",
      "PC_ 2 \n",
      "Positive:  GNLY, CCL5, NKG7, IGLC3, GZMB, ZNF683, S100A12, CD1C, ALAS2, F13A1 \n",
      "\t   LIF, SNCA, CD80, IL5, TLR2, IGHG2, GZMH, MMP12, CXCL8, ADGRG3 \n",
      "\t   CTSG, FCER2, CD33, CXCL2, CXCR2, CD14, CLEC4D, ZBED2, CNTNAP3, MME \n",
      "Negative:  CD27, DUSP1, CD69, LCK, GAPDH, CD74, JUNB, RGS1, FOSB, DUSP2 \n",
      "\t   BTG1, CD3E, CD48, HLA-DRA, CD6, CXCR3, GIMAP2, FYN, CD44, CD2 \n",
      "\t   SELL, GNAI2, PTPRC, CD37, ITGA4, CD3G, ADA, BIN2, CD5, GZMK \n",
      "PC_ 3 \n",
      "Positive:  KLRB1, NKG7, GZMA, RORC, ZBTB16, NCR3, IL18RAP, CST7, GZMK, KLRG1 \n",
      "\t   CCL5, DPP4, CCR2, PRF1, CTSW, IL7R, ARL4C, PIK3AP1, CXCR6, TNFRSF25 \n",
      "\t   IL18R1, RORA, PRDM1, EOMES, BCL2, KLRF1, CCR5, IFNGR1, RUNX3, BLK \n",
      "Negative:  LEF1, PASK, SELL, CCR7, CD8B, YBX3, TOP2A, CD79A, TYMS, CXCR3 \n",
      "\t   CD27, KLRK1, LGALS9, AURKB, IGHM, HLA-DRA, UBE2C, LGALS1, ICOS, BTLA \n",
      "\t   CTLA4, FOSB, RGS1, TIGIT, MKI67, HLA-DMA, MCM2, CD38, MCM4, ZNF683 \n",
      "PC_ 4 \n",
      "Positive:  TYMS, TOP2A, AURKB, UBE2C, MCM2, MCM4, CCR5, HLA-DRA, CCL5, GZMA \n",
      "\t   CD38, MKI67, GZMK, CTLA4, HLA-DMA, PCNA, NKG7, ENTPD1, CST7, HLA-DQB1 \n",
      "\t   CCR2, CXCR6, ADA, CD27, KLRB1, HAVCR2, CHI3L2, PDCD1, PTTG2, ICOS \n",
      "Negative:  DUSP1, JUNB, FOSB, CXCR4, BTG1, LEF1, JUN, KLRC3, ITGAM, TCF7 \n",
      "\t   CCR7, PIK3IP1, IFITM2, KIR2DL1, CD69, CD37, KLRK1, CD3E, CD300A, NT5E \n",
      "\t   RUNX3, KLRF1, HLA-A, CD48, CD8B, ZNF683, B3GAT1, CTSW, ITGAX, CD7 \n",
      "PC_ 5 \n",
      "Positive:  ZNF683, CD52, GNLY, LTB, LGALS3, IFITM2, SELL, KIR2DL1, TNFRSF25, NCR3 \n",
      "\t   SELPLG, LEF1, IL18RAP, IL7R, DPP4, MYC, KLRC1, RORC, TXK, CD300A \n",
      "\t   KLRC3, ITGAX, ZBTB16, IL32, ANXA5, TYMS, GZMB, S100A10, PRF1, CD7 \n",
      "Negative:  DUSP2, RGS1, CCL5, CXCR4, CD74, CST7, NKG7, FOSB, HLA-DMA, DUSP1 \n",
      "\t   GZMK, GZMA, YBX3, MS4A1, TIGIT, CD160, CD69, HLA-DRA, JUN, JUNB \n",
      "\t   CCL4, HLA-DPA1, CHI3L2, TGFBI, BTG1, IFNG, FYN, CD8A, LAIR2, CD8B \n",
      "\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 16215 rows containing missing values (`geom_point()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li><li>10</li><li>11</li><li>12</li><li>13</li><li>14</li><li>15</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 4\n",
       "\\item 5\n",
       "\\item 6\n",
       "\\item 7\n",
       "\\item 8\n",
       "\\item 9\n",
       "\\item 10\n",
       "\\item 11\n",
       "\\item 12\n",
       "\\item 13\n",
       "\\item 14\n",
       "\\item 15\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 3\n",
       "4. 4\n",
       "5. 5\n",
       "6. 6\n",
       "7. 7\n",
       "8. 8\n",
       "9. 9\n",
       "10. 10\n",
       "11. 11\n",
       "12. 12\n",
       "13. 13\n",
       "14. 14\n",
       "15. 15\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“You're computing too large a percentage of total singular values, use a standard svd instead.”\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“did not converge--results might be invalid!; try increasing work or maxit”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "PC_ 1 \n",
      "Positive:  CD45RA, CD11c, HLA-DR, Tim3, CD56, CD278, GITR, CD137, CD62L, CD272 \n",
      "Negative:  CD28, CD127, CD27, CD161, CD196, CXCR6, CD25, CCR7, CD279, CD183 \n",
      "PC_ 2 \n",
      "Positive:  CD56, CD161, CD196, CXCR6, CD45RA, CD11c, GITR, CD127, Tim3, CD134 \n",
      "Negative:  CD272, CCR7, CD183, CD279, CD278, CD27, HLA-DR, CD25, CD137, CD28 \n",
      "PC_ 3 \n",
      "Positive:  CD279, HLA-DR, CD278, CD137, CXCR6, CD161, Tim3, CD196, CD27, CD28 \n",
      "Negative:  CCR7, CD25, CD62L, CD183, CD56, CD127, CD45RA, CD272, GITR, CD11c \n",
      "PC_ 4 \n",
      "Positive:  CD62L, Tim3, CD11c, CD27, CD56, CD28, CD279, CCR7, GITR, HLA-DR \n",
      "Negative:  CD272, CD45RA, CD196, CD127, CD278, CD161, CXCR6, CD25, CD183, CD134 \n",
      "PC_ 5 \n",
      "Positive:  CD134, GITR, CD25, CD137, Tim3, CD278, CD196, CD279, CD161, CD272 \n",
      "Negative:  CD183, CCR7, CD56, CXCR6, HLA-DR, CD45RA, CD11c, CD62L, CD27, CD127 \n",
      "\n",
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: PC_) setting key to original value 'apca_'”\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "09:54:00 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "09:54:00 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "09:54:02 Initializing from normalized Laplacian + noise\n",
      "\n",
      "09:54:03 Commencing optimization for 200 epochs, with 1597036 positive edges\n",
      "\n",
      "09:54:22 Optimization finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Input layer\n",
    "seu <- subset(seu, idents = c(0, 1, 2, 3, 4, 5, 6, 8, 9, 10))#7: gd T cells\n",
    "str(seu)\n",
    "\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "sample.name <- \"COVID_ADTnorm_RPCA.wnn\"\n",
    "dir.name <- \"Seurat_plots_2nd\"\n",
    "red.use <- \"wnn.umap\"\n",
    "resol <- 0.25 # resolution for clustering\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "###RNA\n",
    "DefaultAssay(seu) <- \"RNA.RPCA\"\n",
    "#Scale data and Perform PCA\n",
    "VariableFeatures(seu) <- rownames(seu[[\"RNA.RPCA\"]])\n",
    "seu <- ScaleData(seu, verbose = FALSE)\n",
    "seu = RunPCA(object = seu, npcs = 50)\n",
    "#JackStraw\n",
    "seu = JackStraw(object = seu, reduction = \"pca\", num.replicate = 100, dims = 50)\n",
    "seu <- ScoreJackStraw(object = seu, reduction = \"pca\", dims = 1:50, score.thresh = 0.05)\n",
    "file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Jackstraw.png\", sep='.')\n",
    "png(file.name, width = 1250, height = 500)\n",
    "JackStrawPlot(object = seu, dims = 1:50)\n",
    "dev.off()\n",
    "#Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "#Extract PCs which fulfill the pvalue threshold\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "dims\n",
    "\n",
    "###ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Scale data and Perform PCA\n",
    "VariableFeatures(seu) <- rownames(seu[[\"ADT\"]])\n",
    "seu = ScaleData(object = seu) \n",
    "seu = RunPCA(object = seu, reduction.name = 'apca')\n",
    "    \n",
    "###Clustering Weighted Nearest Neighbor\n",
    "seu <- FindMultiModalNeighbors(seu, reduction.list = list(\"pca\", \"apca\"), \n",
    "                                dims.list = list(dims, c(1:10)))\n",
    "\n",
    "seu <- RunUMAP(seu, nn.name = \"weighted.nn\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "###Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"_2nd.rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "29300426",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name, recursive = TRUE):\n",
      "“'result/intermediate/3_SCT/Seurat_plots_2nd' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 51853\n",
      "Number of edges: 1297731\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9168\n",
      "Number of communities: 73\n",
      "Elapsed time: 84 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "62 singletons identified. 11 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 9\n",
      "\n",
      "Calculating cluster 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Check clustering resolution / Merker genes for each cluster\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "sample.name <- \"COVID.ADTnorm_RPCA.wnn\"\n",
    "dir.name <- \"Seurat_plots_2nd\"\n",
    "red.use <- \"wnn.umap\"\n",
    "tmp_rna = c(\"TRBC2\", \"TRDC\", \"CD14\", \"MS4A1\", \"KLRB1\", \"ZBTB16\")\n",
    "tmp_adt = c(\"CD62L\", \"CD45RA\", \"CD56\", \"CD196\", \"CD161\", \"CXCR6\")\n",
    "resol = 0.28\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "#Find Clusters\n",
    "seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resol)\n",
    "\n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "    \n",
    "###RNA\n",
    "DefaultAssay(seu) <- 'RNA.RPCA'\n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"RNA\"), resol)\n",
    "##Check lineage marker gene expression\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"RNA\"), red.use, tmp_rna, \"lineage\")\n",
    "\n",
    "###ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"ADT\"), resol)\n",
    "##Check lineage marker gene expression\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"ADT\"), red.use, tmp_adt, \"lineage\")\n",
    "\n",
    "###Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"2nd.res\", resol, \"rda\", sep='.')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b742086",
   "metadata": {},
   "source": [
    "# 3rd analysis with removing MAIT cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "001dd0fe",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in PrepDR(object = object, features = features, verbose = verbose):\n",
      "“The following 5 features requested have zero variance (running reduction without them): AQP9, CCL17, IL33, EPX, SPP1”\n",
      "PC_ 1 \n",
      "Positive:  GZMH, NKG7, GNLY, GZMB, CST7, FCGR3A, PRF1, GZMA, CCL5, CX3CR1 \n",
      "\t   LGALS1, ZNF683, CTSW, ITGAM, ITGB2, CD63, CD300A, B3GAT1, RUNX3, APOBEC3G \n",
      "\t   KIR2DL1, KLRC3, KLRF1, CCL4, LAIR2, CTSD, IFNG, CD244, CD247, ITGAX \n",
      "Negative:  TCF7, IL7R, LTB, CD27, CCR7, GZMK, PASK, LEF1, JUNB, PIK3IP1 \n",
      "\t   TNFRSF25, TRAT1, MYC, DUSP2, FOSB, CD69, SELL, YBX3, DUSP1, CXCR4 \n",
      "\t   HLA-A, CXCR3, NT5E, CD28, RGS1, CD79A, BTG1, JUN, BIRC3, CD5 \n",
      "PC_ 2 \n",
      "Positive:  IGLC3, ALAS2, S100A12, CD1C, F13A1, GNLY, SNCA, ADGRG3, CNTNAP3, IGHG2 \n",
      "\t   TLR2, NT5E, IL5, IL9, CD80, CXCL8, TREM1, LIF, CCL5, CD14 \n",
      "\t   CXCL2, CTSG, IL4, TNFRSF4, CLEC4D, MMP12, APOE, JCHAIN, MME, CCL20 \n",
      "Negative:  GAPDH, CD74, HLA-DRA, LCK, CD3G, ITGB2, ADA, GNAI2, TYMS, APOBEC3G \n",
      "\t   BIN2, IL2RB, CD8A, TOP2A, ANXA5, PTPRC, CD3D, RGS1, CD2, CD3E \n",
      "\t   FYN, CTSD, HLA-DMA, CCR5, ITGA4, CD27, ARL4C, MCM2, PRF1, DUSP1 \n",
      "PC_ 3 \n",
      "Positive:  DUSP1, JUNB, CXCR4, FOSB, BTG1, CD3E, RUNX3, JUN, CD48, TCF7 \n",
      "\t   KLRF1, IL7R, ITGAM, KLRC3, CTSW, CD37, PIK3IP1, IFITM2, ARL4C, CD69 \n",
      "\t   HLA-A, CD300A, KIR2DL1, CD8A, LTB, IL2RB, DUSP2, PTPRC, CD7, KLRK1 \n",
      "Negative:  TYMS, TOP2A, UBE2C, AURKB, MCM2, MCM4, MKI67, HLA-DRA, CD38, CCR5 \n",
      "\t   CTLA4, PCNA, HLA-DMA, ENTPD1, CD27, CCL5, HLA-DQB1, CCR2, ADA, ICOS \n",
      "\t   BTLA, CXCR6, PTTG2, HAVCR2, CXCR3, ANXA5, GZMK, PDCD1, CD70, CHI3L2 \n",
      "PC_ 4 \n",
      "Positive:  ZNF683, SELL, CD52, LEF1, GNLY, IFITM2, LTB, KIR2DL1, CCR7, GZMB \n",
      "\t   LGALS1, PASK, LGALS3, SELPLG, CD79A, KLRC3, TNFRSF25, ITGAX, TXK, S100A10 \n",
      "\t   CD300A, KLRC1, IGHM, ANXA5, TYMS, MYC, CD7, IFITM3, FCGR3A, CX3CR1 \n",
      "Negative:  GZMK, DUSP2, CCL5, CST7, GZMA, NKG7, CD74, RGS1, CXCR4, MS4A1 \n",
      "\t   KLRB1, KLRG1, HLA-DMA, EOMES, DUSP1, CCL4, CD160, FYN, CCR5, PDCD1 \n",
      "\t   TIGIT, CHI3L2, TGFBI, RUNX3, TNF, HLA-DRA, HLA-DPA1, FOSB, CD69, HLA-A \n",
      "PC_ 5 \n",
      "Positive:  TOP2A, TYMS, KLRC3, UBE2C, CXCR4, DUSP1, TRDC, AURKB, ITGAX, KLRF1 \n",
      "\t   MCM2, PCNA, FOSB, NCR3, LAT2, JUN, MCM4, CD300A, IKZF2, JUNB \n",
      "\t   CTSW, LGALS3, CD160, DUSP2, IFITM2, PIK3AP1, BTG1, KIR2DL1, CD63, IL18RAP \n",
      "Negative:  CD2, CD8B, CD6, CX3CR1, CD8A, CD5, IL32, LAT, TRAC, CD27 \n",
      "\t   CD52, SELPLG, GIMAP5, GIMAP2, PDCD1, PTPRC, TRBC2, CD3G, LAIR2, APOBEC3G \n",
      "\t   IGBP1, GZMB, GZMH, CCR5, CD3D, BIN2, CXCR6, TNFRSF25, PRF1, CCR7 \n",
      "\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 16135 rows containing missing values (`geom_point()`).”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li><li>10</li><li>11</li><li>12</li><li>13</li><li>14</li><li>15</li><li>16</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1\n",
       "\\item 2\n",
       "\\item 3\n",
       "\\item 4\n",
       "\\item 5\n",
       "\\item 6\n",
       "\\item 7\n",
       "\\item 8\n",
       "\\item 9\n",
       "\\item 10\n",
       "\\item 11\n",
       "\\item 12\n",
       "\\item 13\n",
       "\\item 14\n",
       "\\item 15\n",
       "\\item 16\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1\n",
       "2. 2\n",
       "3. 3\n",
       "4. 4\n",
       "5. 5\n",
       "6. 6\n",
       "7. 7\n",
       "8. 8\n",
       "9. 9\n",
       "10. 10\n",
       "11. 11\n",
       "12. 12\n",
       "13. 13\n",
       "14. 14\n",
       "15. 15\n",
       "16. 16\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“You're computing too large a percentage of total singular values, use a standard svd instead.”\n",
      "Warning message in irlba(A = t(x = object), nv = npcs, ...):\n",
      "“did not converge--results might be invalid!; try increasing work or maxit”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "Warning message:\n",
      "“Requested number is larger than the number of available items (21). Setting to 21.”\n",
      "PC_ 1 \n",
      "Positive:  CD27, CD28, CD127, CCR7, CD279, CD25, CD183, CD272, CXCR6, CD161 \n",
      "Negative:  CD45RA, CD56, CD11c, Tim3, GITR, HLA-DR, CD137, CD62L, CD196, CD278 \n",
      "PC_ 2 \n",
      "Positive:  CD127, CD25, CCR7, CD56, CD62L, CD45RA, CD183, CD196, GITR, CD11c \n",
      "Negative:  CD279, HLA-DR, CD278, CD137, CXCR6, Tim3, CD161, CD27, CD28, CD134 \n",
      "PC_ 3 \n",
      "Positive:  CD272, CD45RA, CD278, CD183, CD196, CD137, CD127, HLA-DR, CD25, CD134 \n",
      "Negative:  CD62L, Tim3, CD11c, CD56, CD27, CD28, CD279, CXCR6, GITR, CD161 \n",
      "PC_ 4 \n",
      "Positive:  CD62L, HLA-DR, CD183, CD278, CCR7, CD272, CD137, CD11c, CD27, CD28 \n",
      "Negative:  CD161, CD196, CXCR6, CD25, CD134, CD279, CD127, CD45RA, GITR, CD56 \n",
      "PC_ 5 \n",
      "Positive:  CD25, CD134, Tim3, CD278, CD196, CD137, GITR, CD62L, CD272, CCR7 \n",
      "Negative:  CXCR6, CD183, CD56, CD45RA, CD161, CD27, HLA-DR, CD279, CD11c, CD127 \n",
      "\n",
      "Warning message:\n",
      "“Cannot add objects with duplicate keys (offending key: PC_) setting key to original value 'apca_'”\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "10:37:09 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "10:37:10 Commencing smooth kNN distance calibration using 1 thread\n",
      "\n",
      "10:37:11 Initializing from normalized Laplacian + noise\n",
      "\n",
      "10:37:11 Commencing optimization for 200 epochs, with 1270556 positive edges\n",
      "\n",
      "10:37:27 Optimization finished\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Input layer \n",
    "seu <- subset(seu, idents = c(0, 1, 3, 4, 5, 6, 8, 10)) #2,9: MAIT cells\n",
    "\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "sample.name <- \"COVID_ADTnorm_RPCA.wnn\"\n",
    "dir.name <- \"Seurat_plots_3rd\"\n",
    "red.use <- \"wnn.umap\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "###RNA\n",
    "DefaultAssay(seu) <- \"RNA.RPCA\"\n",
    "#Scale data and Perform PCA\n",
    "VariableFeatures(seu) <- rownames(seu[[\"RNA.RPCA\"]])\n",
    "seu <- ScaleData(seu, verbose = FALSE)\n",
    "seu = RunPCA(object = seu, npcs = 50)\n",
    "#JackStraw\n",
    "seu = JackStraw(object = seu, reduction = \"pca\", num.replicate = 100, dims = 50)\n",
    "seu <- ScoreJackStraw(object = seu, reduction = \"pca\", dims = 1:50, score.thresh = 0.05)\n",
    "file.name <- str_c(dir.name, sample.name, sep='/') %>% str_c(\"Jackstraw.png\", sep='.')\n",
    "png(file.name, width = 1250, height = 500)\n",
    "JackStrawPlot(object = seu, dims = 1:50)\n",
    "dev.off()\n",
    "#Determine PCs used for clustering/tSNE analysis (dims.use)\n",
    "#Extract PCs which fulfill the pvalue threshold\n",
    "tmp = as.data.frame(seu@reductions$pca@jackstraw@overall.p.values)\n",
    "tmp1 = tmp[tmp$Score>0.05,1]\n",
    "dims= c(1:(min(tmp1)-1))\n",
    "dims\n",
    "\n",
    "###ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Scale data and Perform PCA\n",
    "VariableFeatures(seu) <- rownames(seu[[\"ADT\"]])\n",
    "seu = ScaleData(object = seu) \n",
    "seu = RunPCA(object = seu, reduction.name = 'apca')\n",
    "    \n",
    "###Clustering Weighted Nearest Neighbor\n",
    "seu <- FindMultiModalNeighbors(seu, reduction.list = list(\"pca\", \"apca\"), \n",
    "                                dims.list = list(dims, c(1:10)))\n",
    "\n",
    "seu <- RunUMAP(seu, nn.name = \"weighted.nn\", reduction.name = \"wnn.umap\", reduction.key = \"wnnUMAP_\")\n",
    "\n",
    "###Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"_3rd.rda\", sep='')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4b3fbe04",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name, recursive = TRUE):\n",
      "“'result/intermediate/3_SCT/Seurat_plots_3rd' already exists”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck\n",
      "\n",
      "Number of nodes: 41852\n",
      "Number of edges: 1053523\n",
      "\n",
      "Running smart local moving algorithm...\n",
      "Maximum modularity in 10 random starts: 0.9174\n",
      "Number of communities: 73\n",
      "Elapsed time: 57 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "64 singletons identified. 9 final clusters.\n",
      "\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Calculating cluster 8\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 6\n",
      "\n",
      "Calculating cluster 7\n",
      "\n",
      "Warning message in FindMarkers.default(object = data.use, slot = data.slot, counts = counts, :\n",
      "“No features pass logfc.threshold threshold; returning empty data.frame”\n",
      "Calculating cluster 8\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Check clustering resolution / Merker genes for each cluster\n",
    "sample.name <- \"COVID\"\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "dir.name <- \"Seurat_plots_3rd\"\n",
    "red.use <- \"wnn.umap\"\n",
    "tmp_rna = c(\"TRBC2\", \"TRDC\", \"CD14\", \"MS4A1\", \"KLRB1\", \"ZBTB16\")\n",
    "tmp_adt = c(\"CD62L\", \"CD45RA\", \"CD56\", \"CD196\", \"CD161\", \"CXCR6\")\n",
    "resol = 0.2\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "#Find Clusters\n",
    "seu <- FindClusters(object = seu, graph.name = \"wsnn\", algorithm = 3, resolution =resol)\n",
    "\n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "    \n",
    "###RNA\n",
    "DefaultAssay(seu) <- 'RNA.RPCA'\n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"RNA\"), resol)\n",
    "##Check lineage marker gene expression\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"RNA\"), red.use, tmp_rna, \"lineage\")\n",
    "\n",
    "###ADT\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "#Marker gene extraction and create marker gene heatmap\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"ADT\"), resol)\n",
    "##Check lineage marker gene expression\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"ADT\"), red.use, tmp_adt, \"lineage\")\n",
    "\n",
    "###Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"3rd.res\", resol, \"rda\", sep='.')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a2b306e",
   "metadata": {},
   "source": [
    "# Removing minor clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "198a4841",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "“font family 'Arial' not found in PostScript font database”\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n",
      "Calculating cluster 0\n",
      "\n",
      "Calculating cluster 1\n",
      "\n",
      "Calculating cluster 2\n",
      "\n",
      "Calculating cluster 3\n",
      "\n",
      "Calculating cluster 4\n",
      "\n",
      "Calculating cluster 5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "seu <- subset(seu, idents = c(0, 1, 2, 3, 4, 5))\n",
    "\n",
    "###Visualize clusters\n",
    "sample.name <- \"COVID_ADTnorm_RPCA.wnn\"\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "dir.name <- \"Seurat_plots_3rd_womin\"\n",
    "red.use <- \"wnn.umap\"\n",
    "tmp_rna = c(\"TRBC2\", \"TRDC\", \"CD14\", \"MS4A1\", \"KLRB1\", \"ZBTB16\")\n",
    "tmp_adt = c(\"CD62L\", \"CD45RA\", \"CD56\", \"CD196\", \"CD161\", \"CXCR6\")\n",
    "resol = 0.2\n",
    "\n",
    "########################## Processing layer #############################\n",
    "dir.name <- str_c(dir.SCT, dir.name, sep = \"/\")\n",
    "dir.create(dir.name, recursive = TRUE)\n",
    "\n",
    "#DimPlot with sample origin\n",
    "DimOrigin(seu, dir.name, sample.name, red.use, resol)\n",
    "    \n",
    "#Marker gene extraction and create marker gene heatmap (RNA)\n",
    "DefaultAssay(seu) <- 'RNA.RPCA'\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"RNA\"), resol)\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "MarkerHeatmap(seu, dir.name, str_c(sample.name, \"ADT\"), resol)\n",
    "\n",
    "##Check lineage marker gene expression\n",
    "DefaultAssay(seu) <- 'RNA.RPCA'\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"RNA\"), red.use, tmp_rna, \"lineage\")\n",
    "DefaultAssay(seu) <- 'ADT'\n",
    "ScatterViolin(seu, dir.name, str_c(sample.name, \"ADT\"), red.use, tmp_adt, \"lineage\")    \n",
    "\n",
    "#Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, sample.name, sep = \"/\") %>% str_c(\"3rd_womin.res\", resol, \"rda\", sep='.')\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a3d1ae7",
   "metadata": {},
   "source": [
    "# Integration of scTCRseq results into SCT Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "87004d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "#Create histogram was showing the total number of TCR reads detected in each cell (top left) \n",
    "#and the percentage of TCR reads that were the most common among them (bottom left).\n",
    "#Scatter plot representing these two metrices of cells were also generated\n",
    "Histogram <- function(d, file.name, sample.name, dir.output){\n",
    "\n",
    "  d_count_total <- as.numeric(d$count.total)\n",
    "  \n",
    "  #Count histogram\n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"histogram.count.tiff\", sep = '.') \n",
    "  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)\n",
    "  p <- ggplot(NULL, aes(x=d_count_total)) +\n",
    "    geom_histogram(binwidth=0.1) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    labs(x = \"Read counts of TCR\") +\n",
    "    theme(\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) +\n",
    "    scale_x_continuous(trans=scales::log2_trans(),\n",
    "                   breaks=scales::trans_breaks(\"log2\",function(x) 2^x),\n",
    "                   labels=scales::trans_format(\"log2\",scales::math_format(2^.x)))\n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  #Frequency histogram  \n",
    "  d_freq <- as.numeric(d$freq)\n",
    "  med_f <- 0.3\n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"histogram.freq.tiff\", sep = '.')\n",
    "  tiff(image.file, width=1.2*ppi, height=0.8*ppi, res=ppi)\n",
    "  p <- ggplot(NULL, aes(x=d_freq)) +\n",
    "    geom_histogram(binwidth=0.05) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    labs(x = \"Proportion of the largest TCR read\") +\n",
    "    theme(\n",
    "      axis.title.y=element_blank(),\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) \n",
    "  print(p)\n",
    "  dev.off()\n",
    "\n",
    "  df <- data.frame(d_count_total, d_freq)\n",
    "  \n",
    "  #Scatter plot    \n",
    "  total <- nrow(subset(df, df$d_freq >= 0))\n",
    "  lb <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_total < 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  rb <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_total < 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  lt <- paste(round(100*nrow( subset(df, df$d_freq < 0.6 & df$d_count_total >= 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  rt <- paste(round(100*nrow( subset(df, df$d_freq >= 0.6 & df$d_count_total >= 2^5 )) / total, digits = 1), \"%\", sep=\"\" ) \n",
    "  \n",
    "  ppi <- 300\n",
    "  image.file <- str_c(dir.output, file.name, sep = \"/\") %>% str_c(sample.name, \"Scatter.tiff\", sep = '.')                     \n",
    "  tiff(image.file, width=1.2*ppi, height=1.2*ppi, res=ppi)\n",
    "  p <- ggplot(d, aes(x=d_freq, y=d_count_total)) +  \n",
    "    stat_bin2d(bins=60) +\n",
    "    scale_fill_gradient(low=\"lightblue\", high=\"red\") +\n",
    "    theme_bw(base_size = 6) +\n",
    "    geom_vline(aes(xintercept = 0.6), size=0.25, colour=\"black\") +\n",
    "    geom_hline(aes(yintercept = 2^5), size=0.25, colour=\"black\") +\n",
    "    labs(x = \"Proportion of the largest TCR read\", y = \"Read counts of TCR\") +\n",
    "    theme(\n",
    "      axis.text.x = element_text(family=\"Arial\"),\n",
    "      axis.text.y = element_text(family=\"Arial\"),\n",
    "      axis.title=element_text(size=4)) +\n",
    "    guides(fill=FALSE) +\n",
    "    scale_y_continuous(trans=scales::log2_trans(),\n",
    "                   breaks=scales::trans_breaks(\"log2\",function(x) 2^x),\n",
    "                   labels=scales::trans_format(\"log2\",scales::math_format(2^.x)))\n",
    "    p <- p + #xlim(0,1) +\n",
    "    annotate(\"text\", x=-Inf, y=0, hjust=-0.1, vjust=-0.4, label=lb,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=Inf, y=0, hjust=1.1, vjust=-0.4, label=rb,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=-Inf, y=Inf, hjust=-0.1, vjust=1.3, label=lt,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) +\n",
    "    annotate(\"text\", x=Inf, y=Inf, hjust=1.1, vjust=1.3, label=rt,\n",
    "             family=\"Arial\",colour=\"black\",size=1.5) \n",
    "  print(p)\n",
    "  dev.off()  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b5df2770",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "#Create TCRa/TCRb combine table for cell barcodes\n",
    "CombineTable <- function(dir.input, sample.name, dir.output, count_th, freq_th, clonotype, cores){\n",
    "    combined.tables <- data.frame()\n",
    "    \n",
    "    ###Define function\n",
    "    tableread_fast = function(i, header=TRUE, quote=\"\", sep=\",\"){\n",
    "      tmp = fread(i, header=header, sep=sep, quote=quote, nThread=32)\n",
    "      tmp = as.data.frame(tmp)\n",
    "      return(tmp)\n",
    "    }\n",
    "\n",
    "    #Convert MiXCR output into VDJtools format\n",
    "    Convert <- function(file.name, name.TRA){\n",
    "        name.input <- str_c(name.TRA, file.name, sep = \"/\")\n",
    "        data <- tableread_fast(name.input, header = TRUE, sep = '\\t')\n",
    "\n",
    "        #Extract TCR information from mixcr output\n",
    "        count.total<-data$cloneCount / data$cloneFraction\n",
    "        freq<-data$cloneFraction\n",
    "        cdr3nt<-data$nSeqImputedCDR3\n",
    "        cdr3aa<-data$aaSeqImputedCDR3\n",
    "        v<-str_sub(data$bestVHit, end=-4)\n",
    "        d<-str_sub(data$bestDHit, end=-4)\n",
    "        j<-str_sub(data$bestJHit, end=-4)\n",
    "        data3 <- rbind(count.total,freq,cdr3nt,cdr3aa,v,d,j)\n",
    "        data3 <- as.data.frame(t(data3))\n",
    "        names(data3) <- c(\"count.total\",\"freq\",\"cdr3nt\",\"cdr3aa\",\"v\",\"d\",\"j\")\n",
    "\n",
    "        #Extract largest clone in cell barcode and append cell barcode information\n",
    "        name_out <- str_split(file.name, \"_\")\n",
    "        CB <- name_out[[1]][[4]]\n",
    "        CB_out <- str_split(CB, \"\\\\.\")\n",
    "        CB <- CB_out[[1]][[1]]\n",
    "        d_out <- data3[1,]\n",
    "        d_out$CB <- CB\n",
    "\n",
    "        return(d_out)\n",
    "    }\n",
    "    \n",
    "    #Convert MiXCR output into VDJtools format\n",
    "    name.TRA <- str_c(sample.name, \"TRAC1_mixcr\", sep = \"_\")\n",
    "    files  <- list.files(name.TRA, pattern=\".txt\")\n",
    "    cl <- makeCluster(cores)\n",
    "    registerDoParallel(cl)\n",
    "    TRA.table <- invisible(foreach(file.name = files,\n",
    "            .combine = rbind, .packages=c(\"ggplot2\", \"extrafont\", \"stringr\", \"dplyr\", \"data.table\")) %dopar% {Convert(file.name, name.TRA)})\n",
    "    stopCluster(cl)\n",
    "    unlink(name.TRA, recursive=TRUE)\n",
    "\n",
    "    name.TRB <- str_c(sample.name, \"TRBC1_mixcr\", sep = \"_\")\n",
    "    files  <- list.files(name.TRB, pattern=\".txt\")\n",
    "    cl <- makeCluster(cores)\n",
    "    registerDoParallel(cl)\n",
    "    TRB.table <- invisible(foreach(file.name = files,\n",
    "            .combine = rbind, .packages=c(\"ggplot2\", \"extrafont\", \"stringr\", \"dplyr\", \"data.table\")) %dopar% {Convert(file.name, name.TRB)})\n",
    "    stopCluster(cl)\n",
    "    unlink(name.TRB, recursive=TRUE)\n",
    "\n",
    "    #Output histogram and scatter plot for summarizing scTCR status\n",
    "    Histogram(TRA.table, \"TRA\", sample.name, dir.output)\n",
    "    Histogram(TRB.table, \"TRB\", sample.name, dir.output)\n",
    "\n",
    "    #thresholding\n",
    "    TRA.table$count.total <- as.numeric(TRA.table$count.total)\n",
    "    TRA.table$freq <- as.numeric(TRA.table$freq)\n",
    "    TRB.table$count.total <- as.numeric(TRB.table$count.total)\n",
    "    TRB.table$freq <- as.numeric(TRB.table$freq)\n",
    "    TCRa_th <- subset(TRA.table, count.total >= 2^(count_th) & freq >= freq_th)\n",
    "    TCRb_th <- subset(TRB.table, count.total >= 2^(count_th) & freq >= freq_th)\n",
    "\n",
    "    #Paring TCRa and TCRb by cell barcode\n",
    "    names(TCRa_th) <- c(\"count.total.A\",\"freq.A\",\"cdr3nt.A\",\"cdr3aa.A\",\"v.A\",\"d.A\",\"j.A\", \"CB\")\n",
    "    names(TCRb_th) <- c(\"count.total.B\",\"freq.B\",\"cdr3nt.B\",\"cdr3aa.B\",\"v.B\",\"d.B\",\"j.B\", \"CB\")\n",
    "    combined <- merge(TCRa_th, TCRb_th, all=T, by =\"CB\")\n",
    "    combined$CB <- str_c(sample.name, combined$CB, sep = \"_\")\n",
    "    #Exclude cells in which neither TCRa nor TCRb sequence were detected\n",
    "    combined <- subset(combined, combined$cdr3nt.A != \"UD\" | combined$cdr3nt.B != \"UD\")\n",
    "\n",
    "    #Generate clone id for further analysis\n",
    "    #Definition of clones can be changed between \"ABnt\", \"ABaa\", \"Bnt\", and \"Baa\"\n",
    "    if(clonotype==\"nt\"){\n",
    "        combined$clone.id.TCRa <- str_c(combined$cdr3nt.A, combined$v.A, combined$j.A, sep=\"_\")\n",
    "        combined$clone.id.TCRb <- str_c(combined$cdr3nt.B, combined$v.B, combined$j.B, sep=\"_\")\n",
    "    }\n",
    "    if(clonotype==\"aa\"){\n",
    "        combined$clone.id.TCRa <- str_c(combined$cdr3aa.A, combined$v.A, combined$j.A, sep=\"_\")\n",
    "        combined$clone.id.TCRb <- str_c(combined$cdr3aa.B, combined$v.B, combined$j.B, sep=\"_\")\n",
    "    }\n",
    "    combined$clone.id.TCRab <- str_c(combined$clone.id.TCRa, combined$clone.id.TCRb, sep = \"_\")\n",
    "    name.output <- str_c(dir.output, sample.name, sep = \"/\") %>% str_c(clonotype, \"table\", \"count_th\", count_th, \"freq_th\", freq_th, clonotype, \"csv\", sep = \".\")\n",
    "    write.csv(combined, name.output, row.names = FALSE)\n",
    "    \n",
    "    combined.tables <- rbind(combined.tables, combined)\n",
    "    \n",
    "    return(combined.tables)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cf0f2a62",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2538 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2538 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mUsing `size` aesthetic for lines was deprecated in ggplot2 3.4.0.\n",
      "\u001b[36mℹ\u001b[39m Please use `linewidth` instead.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mThe `<scale>` argument of `guides()` cannot be `FALSE`. Use \"none\" instead as\n",
      "of ggplot2 3.3.4.”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2538 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2646 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2646 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 2646 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1893 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1893 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1893 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1454 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1454 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1454 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3360 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3360 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3360 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1942 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1942 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1942 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3612 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3612 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3612 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3289 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3289 rows containing non-finite values (`stat_bin()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mTransformation introduced infinite values in continuous y-axis”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3289 rows containing non-finite values (`stat_bin2d()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mUsing an external vector in selections was deprecated in tidyselect 1.1.0.\n",
      "\u001b[36mℹ\u001b[39m Please use `all_of()` or `any_of()` instead.\n",
      "  # Was:\n",
      "  data %>% select(clone.id)\n",
      "\n",
      "  # Now:\n",
      "  data %>% select(all_of(clone.id))\n",
      "\n",
      "See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.”\n"
     ]
    }
   ],
   "source": [
    "#Extract TCR sequence for each cell barcode\n",
    "#iteration process for dataset (batches)\n",
    "set.seed(seed = 42)\n",
    "sample.names <- c(\"ID02TCR\", \"ID28TCR\", \"TP38TCR1\", \"TP38TCR2\")\n",
    "dir.input <- \"data/original/SCT\"\n",
    "cores <- 12\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "dir.output <- \"scTCR_processing\"\n",
    "#Threshold for valid cell barcode with single TCR sequence\n",
    "count_th <- 5 #read count threshold for all TCR sequences per cell barcode\n",
    "freq_th <- 0.6 #proportion threshold for largest TCR\n",
    "clonotype <- \"nt\" #Definition of clones. nt: nucleotide sequence / aa: amino acid sequnece \n",
    "\n",
    "################################ Processing layer #################################################\n",
    "dir.output <- str_c(dir.SCT, dir.output, sep = \"/\")\n",
    "dir.create(dir.output)\n",
    "\n",
    "#Unzip files\n",
    "files <- list.files(dir.input, \".tar.bz2\")\n",
    "for(file in files){\n",
    "    name.input <- str_c(dir.input, file, sep = \"/\")\n",
    "    bunzip2(name.input, remove=FALSE)\n",
    "    name.input <- str_remove(name.input, pattern = \".bz2\")\n",
    "    untar(name.input)\n",
    "    file.remove(name.input)\n",
    "}\n",
    "\n",
    "combined.tables <- data.frame()\n",
    "\n",
    "for(sample.name in sample.names){\n",
    "    ##Create TCRa/TCRb combine table for cell barcodes\n",
    "    combined.table <- CombineTable(dir.input, sample.name, dir.output, count_th, freq_th, clonotype, cores)\n",
    "    combined.tables <- rbind(combined.tables, combined.table)\n",
    "}\n",
    "\n",
    "###Merge TCR info into meta.data\n",
    "combined.tables$CB <- str_replace(combined.tables$CB, \"ID02TCR\", \"Batch1-ID2\")\n",
    "combined.tables$CB <- str_replace(combined.tables$CB, \"ID28TCR\", \"Batch1-ID27\")\n",
    "combined.tables$CB <- str_replace(combined.tables$CB, \"TP38TCR1\", \"Batch2-1\")\n",
    "combined.tables$CB <- str_replace(combined.tables$CB, \"TP38TCR2\", \"Batch2-2\")\n",
    "\n",
    "#Extract meta.data and cell BC information\n",
    "meta.data <- seu@meta.data\n",
    "meta.data$CB <- row.names(meta.data)\n",
    "\n",
    "#Merge to Seurat object of SCT data\n",
    "meta.data <- merge(combined.tables, meta.data, all.y = T, by =\"CB\")\n",
    "clone.ids <- c(\"clone.id.TCRa\", \"clone.id.TCRb\", \"clone.id.TCRab\")\n",
    "for(clone.id in clone.ids){\n",
    "    clone.id.list <- dplyr::select(meta.data, clone.id)\n",
    "    row.names(clone.id.list)=as.character(meta.data$CB)\n",
    "    seu <- AddMetaData(object = seu, metadata = clone.id.list, col.name = clone.id)\n",
    "}\n",
    "\n",
    "#Assign TP and ID information\n",
    "ID_TP <- str_split(seu@meta.data$orig.ident, pattern = \"_\", simplify = TRUE)\n",
    "seu@meta.data$TP <- ID_TP[,1]\n",
    "seu@meta.data$ID <- ID_TP[,2]\n",
    "\n",
    "#Output Seurat object in initial analysis\n",
    "file.name <- str_c(dir.SCT, \"COVID_ADTnorm_RPCA.wnn_3rd.res.0.2.wo-min.scTCR.rda\", sep = \"/\")\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d72e5ff8",
   "metadata": {},
   "source": [
    "# Introducing Signature Scores from published SCT data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "09e812f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define function\n",
    "##Plotting Signature scores for clusters\n",
    "modify_vlnplot_score <- function(obj, \n",
    "                          feature, \n",
    "                          pt.size = 0, \n",
    "                          plot.margin = unit(c(-0.75, 0, -0.75, 0), \"cm\"),\n",
    "                          ...) {\n",
    "  \n",
    "  p<- VlnPlot(obj, features = feature, pt.size = pt.size)  + \n",
    "    xlab(\"\") + ylab(feature) + ggtitle(\"\") + \n",
    "    ggplot2::geom_boxplot(outlier.size = 0, size = 0.2) + \n",
    "    theme(legend.position = \"none\",\n",
    "          plot.title= element_blank(),\n",
    "          axis.title.x = element_blank(),\n",
    "          axis.text.x =  element_text(family = \"Arial\", size = 8, angle = 0, colour = \"black\", face = \"plain\"), \n",
    "          axis.ticks.x = element_blank(), \n",
    "          axis.title.y = element_blank(), \n",
    "          axis.text.y = element_text(family = \"Arial\", size = 6, angle = 0, colour = \"black\", face = \"plain\"), \n",
    "          plot.margin = plot.margin ) \n",
    "  return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0bf3ce4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Generate Gene list for Signature Scoring\n",
    "\n",
    "dir.input <- \"metadata\"\n",
    "name.input <- \"Score.Gene.list.txt\"\n",
    "name.reference <- \"BD_genes_human.txt\"\n",
    "dir.output <- \"result/intermediate/2_SCT/GeneScoreSource\"\n",
    "\n",
    "######################## Processing layer #########################\n",
    "dir.create(dir.output)\n",
    "\n",
    "d <- read.table(str_c(dir.input, name.input, sep = \"/\"), header = TRUE)\n",
    "d$clust <- str_c(d$Source, d$Name, sep = \".\")\n",
    "ref <- read.csv(str_c(dir.input, name.reference, sep = \"/\"), header = TRUE)\n",
    "name.clusts <- unique(d$clust)\n",
    "for(name.clust in name.clusts){\n",
    "    d_sub <- dplyr::filter(d, clust == name.clust)\n",
    "    g.ext <- unique(intersect(d_sub$gene, ref$Genesymbol))\n",
    "    d_out <- as.data.frame(g.ext)\n",
    "    names(d_out) <- \"Genes\"\n",
    "    name.output <- str_c(dir.output, \"Sig\", sep = \"/\") %>% str_c(name.clust, \"txt\", sep = \".\")\n",
    "    write.table(d_out, name.output, row.names = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d9295ba1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“The following features are not present in the object: KIAA0101, not searching for symbol synonyms”\n",
      "Warning message:\n",
      "“The following features are not present in the object: IFNA1, not searching for symbol synonyms”\n",
      "Warning message:\n",
      "“The following features are not present in the object: FAM65B, not searching for symbol synonyms”\n",
      "Warning message:\n",
      "“The following features are not present in the object: FYB, not searching for symbol synonyms”\n",
      "Warning message:\n",
      "“The following features are not present in the object: FYB, not searching for symbol synonyms”\n"
     ]
    }
   ],
   "source": [
    "# Introduce Gene Scores using AddModuleScore function\n",
    "\n",
    "#Put signature gene set into \"Signatures\" directory\n",
    "dir.input <- \"result/intermediate/2_SCT/GeneScoreSource\"\n",
    "dir.SCT <- \"result/intermediate/2_SCT\"\n",
    "\n",
    "########################## Processing layer #############################\n",
    "#load Signature data\n",
    "files  <- list.files(dir.input, pattern=\"Sig.\")\n",
    "\n",
    "for(i in files){\n",
    "    i <- str_c(dir.input, i, sep = \"/\")\n",
    "    gene_list <- read.table(i, header = TRUE)\n",
    "    gene_list <- list(as.vector(gene_list$Genes))\n",
    "    sig_name <- str_remove(i, dir.input)\n",
    "    sig_name <- str_remove(sig_name, \"/Sig.\")\n",
    "    sig_name <- str_remove(sig_name, \".txt\")\n",
    "    seu <- AddModuleScore(object = seu, \n",
    "                            ctrl=5,\n",
    "                            features = gene_list,\n",
    "                            assay=\"RNA.RPCA\",\n",
    "                            name=sig_name,\n",
    "                         )\n",
    "}\n",
    "\n",
    "#Output Seurat object\n",
    "file.name <- str_c(dir.SCT, \"COVID_ADTnorm_RPCA.wnn_3rd.res.0.2.wo-min.scTCR.Sig.rda\", sep = \"/\")\n",
    "save(seu, file=file.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eca848be",
   "metadata": {},
   "source": [
    "# Integration of Bulk TCRseq result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e611fc21",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Make clone table by assembling clones from scTCR results \n",
    "CloneSummary <- function(seu, dir.name, sample.name, clonotype){\n",
    "    #seu: Seurat object for analyze\n",
    "    #dir.name: directory name for output\n",
    "    dir.create(dir.name)\n",
    "\n",
    "    #load metadata\n",
    "    meta.data <- as.data.frame(seu@meta.data)\n",
    "    #extract cells only in the assigned sample\n",
    "    meta.data <- dplyr::filter(meta.data, orig.ident == sample.name)\n",
    "    \n",
    "    #Define clone by TCRa / TCRb/ TCRa&b\n",
    "    if(clonotype == \"TCRa\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRa \n",
    "    }\n",
    "    if(clonotype == \"TCRb\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRb \n",
    "    }\n",
    "    if(clonotype == \"TCRab\"){\n",
    "       meta.data$clone.id <- meta.data$clone.id.TCRab \n",
    "    } \n",
    "\n",
    "    #Summarize cluster distribution of each clone\n",
    "    tmp_out = table(meta.data$seurat_clusters, meta.data$clone.id)\n",
    "    tmp_out2 <- as.data.frame(tmp_out, row.names = NULL,\n",
    "                  responseName = \"Freq\", stringsAsFactors = TRUE,\n",
    "                  sep = \"\", base = list(LETTERS))\n",
    "    names(tmp_out2) <- c(\"Clust\", \"names\", \"Freq\")\n",
    "    tmp_out2 <- dcast(tmp_out2, names ~ Clust)\n",
    "\n",
    "    #Summarize the count and frequency of clones \n",
    "    tmp = meta.data %>% group_by(`clone.id`) %>%\n",
    "      dplyr::summarise(count = n()) %>%\n",
    "      dplyr::arrange(desc(count))\n",
    "    tmp = as.data.frame(tmp)\n",
    "    tmp <- tmp[!is.na(tmp$clone.id), ] #NAである細胞を除く\n",
    "    tmp[,3]=tmp[,2]/sum(tmp[,2])\n",
    "    colnames(tmp)=c(\"ntSeq_TRA_TRB_freq\", \"CloneCount\", \"CloneFreq\")\n",
    "    tmp <- tmp[order(tmp$ntSeq_TRA_TRB_freq),]\n",
    "    temp_count <- as.vector(tmp$CloneCount)\n",
    "    temp_freq <- as.vector(tmp$CloneFreq)\n",
    "    #Combine to the cluster distribution\n",
    "    tmp_out3 <- cbind(temp_count, temp_freq, tmp_out2)\n",
    "\n",
    "    #Assign ranks to clones\n",
    "    tmp_out3 <- tmp_out3[order(tmp_out3$temp_freq, decreasing=T),]\n",
    "    rank <-  1:nrow(tmp_out3)\n",
    "    tmp_out3 <- cbind(rank, tmp_out3)\n",
    "    tmp_out3 <- data.frame(tmp_out3)\n",
    "    tmp_out3$rank <- paste(\"Top\",tmp_out3$rank, sep=\"\")\n",
    "    \n",
    "    #Output\n",
    "    name.out <- str_c(dir.name, sample.name, sep = \"/\") %>% str_c(clonotype, \"clone_within_cluster.txt\", sep = \".\")\n",
    "    write.table(tmp_out3, name.out, row.names=F, col.names=T, sep=\"\\t\", quote=F) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8321dea6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n",
      "Warning message in dir.create(dir.name):\n",
      "“'result/intermediate/3_SCT/scTCR_analysis' already exists”\n",
      "Using Freq as value column: use value.var to override.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "##Make clone table by assembling clones from scTCR results \n",
    "\n",
    "dir.name <- \"result/intermediate/2_SCT/scTCR_analysis\"\n",
    "\n",
    "##################### Processing layer #############################################################\n",
    "dir.create(dir.name)\n",
    "\n",
    "sample.list <- unique(seu@meta.data$orig.ident)\n",
    "\n",
    "for(sample.name in sample.list){\n",
    "    #Make clone table by assembling clones from scTCR results \n",
    "    CloneSummary(seu, dir.name, sample.name, \"TCRb\")\n",
    "} \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b1d94dd",
   "metadata": {},
   "source": [
    "# Integration of Bulk TCRseq result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "281aa808",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Merge SCT metadata with WGCNA pattern / AIM positivity\n",
    "\n",
    "#Input layer\n",
    "dir.input.scTCR <- \"result/intermediate/2_SCT/scTCR_analysis\"\n",
    "dir.input.bulk <- \"result/intermediate/1_beta-binomial/JoinTP_DifAbund\"\n",
    "dir.output <- \"result/intermediate/2_SCT/scTCR_analysis\"\n",
    "\n",
    "sample.names <- c(\"002\", \"027\", \"010\", \"019\", \"022\", \"031\", \"035\")\n",
    "sample.scts <- c(\"ID2.TCR\", \"ID27.TCR\", \"ID10.TCR\", \"ID19.TCR\", \"ID22.TCR\", \"ID31.TCR\", \"ID35.TCR\")\n",
    "\n",
    "##################### Processing layer #############################################################\n",
    "for(i in 1:length(sample.names)){\n",
    "    #load data\n",
    "    sct.names <- list.files(dir.input.scTCR, sample.scts[i])\n",
    "    for(sct.name in sct.names){\n",
    "        sct <- read.table(str_c(dir.input.scTCR, sct.name, sep = \"/\"), header = TRUE)\n",
    "        bulk.name <- str_c(dir.input.bulk, \"NaraCOVID_CD8\", sep = \"/\") %>%\n",
    "            str_c(sample.names[i], \"TCR.DifAbund.csv\", sep = \"_\")\n",
    "        bulk <- read.csv(bulk.name, header = TRUE)\n",
    "        \n",
    "        #Merge bulk and sct dataset\n",
    "        d_output <- merge(sct, bulk, by.x = \"names\", by.y = \"ntvj\", all.x = T)\n",
    "        \n",
    "        #Output\n",
    "        name.out <- str_c(dir.output, sct.name, sep = \"/\") %>%\n",
    "                        str_replace(\"clone_within_cluster.txt\", \"clone_within_cluster.bulkinfo.csv\")\n",
    "        write.csv(d_output, name.out, row.names = FALSE)\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
